<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>202504_Workshop.knit.md</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="icon" type="image/png" href="assets/logo/DRSN_logo_smaller_filt.png">

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; } /* Alert */
code span.an { color: #008000; } /* Annotation */
code span.at { } /* Attribute */
code span.bu { } /* BuiltIn */
code span.cf { color: #0000ff; } /* ControlFlow */
code span.ch { color: #008080; } /* Char */
code span.cn { } /* Constant */
code span.co { color: #008000; } /* Comment */
code span.cv { color: #008000; } /* CommentVar */
code span.do { color: #008000; } /* Documentation */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.im { } /* Import */
code span.in { color: #008000; } /* Information */
code span.kw { color: #0000ff; } /* Keyword */
code span.op { } /* Operator */
code span.ot { color: #ff4000; } /* Other */
code span.pp { color: #ff4000; } /* Preprocessor */
code span.sc { color: #008080; } /* SpecialChar */
code span.ss { color: #008080; } /* SpecialString */
code span.st { color: #008080; } /* String */
code span.va { } /* Variable */
code span.vs { color: #008080; } /* VerbatimString */
code span.wa { color: #008000; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>


<link rel="stylesheet" href="adit/custom.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src='assets/logo/DRSN_logo_smaller.png' style='height:30px; vertical-align:middle; align-items:center; margin-right:10px;'></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="0_schedule.html">Schedule</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Introduction Series
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="202502_tidyverse1.html">Tidyverse 1: Intro to Plotting</a>
    </li>
    <li>
      <a href="202502_tidyverse2.html">Tidyverse 2: Intro to Manipulation</a>
    </li>
    <li>
      <a href="202503_tidyverse3.html">Tidyverse 3: Creating a Reproducible Report</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Workshops
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="202504.html">April: Flow Cytometry</a>
    </li>
    <li>
      <a href="202505.html">May: Spatial Transcriptomics</a>
    </li>
    <li>
      <a href="202506.html">June: Spatial Proteomics</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="0_contacts.html">Contacts</a>
</li>
<li>
  <a href="0_archive.html">Archive</a>
</li>
<li>
  <a href="0_resources.html">Extra Resources</a>
</li>
<li>
  <a href="https://github.com/DrThomasOneil/analysis-user-group">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
<li role="separator" class="divider"></li>
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>




</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-04-26
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong>
<code>DigitalResearchSkillsNetwork/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1337code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1337)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1337code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1337)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomDrThomasOneilDigitalResearchSkillsNetworktreee8a3722e8e15f677254f281365ef494fc582e958targetblanke8a3722a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/tree/e8a3722e8e15f677254f281365ef494fc582e958" target="_blank">e8a3722</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomDrThomasOneilDigitalResearchSkillsNetworktreee8a3722e8e15f677254f281365ef494fc582e958targetblanke8a3722a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/tree/e8a3722e8e15f677254f281365ef494fc582e958" target="_blank">e8a3722</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/.RData
    Ignored:    analysis/.Rhistory
    Ignored:    analysis/adit/.DS_Store

Untracked files:
    Untracked:  analysis/202505.Rmd
    Untracked:  analysis/202505_InFocus.Rmd
    Untracked:  analysis/202505_Seminar.Rmd
    Untracked:  analysis/adit/head_index.html
    Untracked:  plots/

Unstaged changes:
    Modified:   .gitignore
    Modified:   workflow.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/202504_Workshop.Rmd</code>) and
HTML (<code>docs/202504_Workshop.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/blob/e8a3722e8e15f677254f281365ef494fc582e958/analysis/202504_Workshop.Rmd" target="_blank">e8a3722</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-04-26
</td>
<td>
wflow_publish("analysis/202504_Workshop.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/938818ef38c5fbd79f7ed4d4059a7f675f9eeeb6/docs/202504_Workshop.html" target="_blank">938818e</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-04-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/ed690e68a3cc081c5e873976cef678ddb47b8b7b/docs/202504_Workshop.html" target="_blank">ed690e6</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-04-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/blob/1e8103ec4eeaa0d1fea219e9acc0b3953b1ae921/analysis/202504_Workshop.Rmd" target="_blank">1e8103e</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-04-14
</td>
<td>
wflow_publish("analysis/*.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/cb886efbb544b13e41d2b972b0c419456422faed/docs/202504_Workshop.html" target="_blank">cb886ef</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-04-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/blob/eb0a784637965e03fe89a40d92bc063cb6f64f8f/analysis/202504_Workshop.Rmd" target="_blank">eb0a784</a>
</td>
<td>
GitHub
</td>
<td>
2025-04-09
</td>
<td>
Update 202504_Workshop.Rmd
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/f6c0433939191a8e12e7efd41297bf2ab16a894f/docs/202504_Workshop.html" target="_blank">f6c0433</a>
</td>
<td>
YuchenLi
</td>
<td>
2025-04-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/blob/04953d6cbea0dd80abc73846d1e7dd1d459824b1/analysis/202504_Workshop.Rmd" target="_blank">04953d6</a>
</td>
<td>
YuchenLi
</td>
<td>
2025-04-09
</td>
<td>
wflow_publish("analysis/*.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/8d78b0f772e5fe892837c79676f027569bf77cd8/docs/202504_Workshop.html" target="_blank">8d78b0f</a>
</td>
<td>
YuchenLi
</td>
<td>
2025-04-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/7421f69d71654eed5edc9bd9d8d813ceab7f4567/docs/202504_Workshop.html" target="_blank">7421f69</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-04-08
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/f2d3385fb3cc4bf368b73563824c7b9e747c7175/docs/202504_Workshop.html" target="_blank">f2d3385</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-24
</td>
<td>
update
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/517dcf9e2aec3f21e4ed3a5067b6ae5575b96386/docs/202504_Workshop.html" target="_blank">517dcf9</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/6a4432bc64d1f2c738ddaeaede4c90607a475def/docs/202504_Workshop.html" target="_blank">6a4432b</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/1f218f33dd3dee770f017e0d06a62eed0fdcd1e7/docs/202504_Workshop.html" target="_blank">1f218f3</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/dca90e9664efab927f9c54dea0df460a000c85d1/docs/202504_Workshop.html" target="_blank">dca90e9</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/1d1c076545b00ed07ab6abc22c93e220776b23cb/docs/202504_Workshop.html" target="_blank">1d1c076</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/956ba9bc3838229756b2898719eb78404919210f/docs/202504_Workshop.html" target="_blank">956ba9b</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/blob/6b72e34e69f438c4eae125c8fb4f0f1bff58a700/analysis/202504_Workshop.Rmd" target="_blank">6b72e34</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-24
</td>
<td>
wflow_publish(c("analysis/*.Rmd"))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/d600337faa3f8497f6022cf81703bff0651f50f3/docs/202504_Workshop.html" target="_blank">d600337</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-18
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/b537f698f77268df329fd76f4c65d25329656918/docs/202504_Workshop.html" target="_blank">b537f69</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-18
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/b7982e9f3b91a038e2d2c4fae14739288542a24d/docs/202504_Workshop.html" target="_blank">b7982e9</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-16
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/fc5dd81aefeec7ff302b074072e4810c6f13238f/docs/202504_Workshop.html" target="_blank">fc5dd81</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-16
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/8e1c1a3dbe88e7d94dfec167b9f108964e3e7b5a/docs/202504_Workshop.html" target="_blank">8e1c1a3</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-16
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/blob/d75e5ae72ad4e99dd853ecdabd54aba2a06060e9/analysis/202504_Workshop.Rmd" target="_blank">d75e5ae</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-03-16
</td>
<td>
wflow_publish(c("analysis/*.Rmd"))
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="flow-cytometry-in-r-a-hands-on-introduction"
class="section level2">
<h2>Flow Cytometry in R: A Hands-On Introduction</h2>
<p>This beginner-friendly, hands-on workshop will guide users through a
complete flow cytometry analysis workflow in R, from loading data to
visualizing results. Through practical exercises, we’ll cover essential
steps like generating histograms and scatterplots, extracting summary
statistics, subsetting data, and performing dimension reduction and
clustering—all while demonstrating the advantages of reproducible
reports with RMarkdown.</p>
<p>Whether you’re new to R or just curious about coding your analyses,
this session will show you how accessible and powerful R can be for flow
cytometry.</p>
<p><strong>Your Demonstrators:</strong></p>
<ul>
<li>Thomas O’Neil<br />
</li>
<li>Yuchen Li (Online)</li>
</ul>
<p><strong>Author</strong>: Yuchen Li, Thomas O’Neil</p>
<p><a href="https://uni-sydney.zoom.us/s/88109360455">Zoom link</a>
(recorded)</p>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>In this workshop, we aim to give an overview of what is possible for
flow analysis in R. We will touch on two workflows:</p>
<ul>
<li><p>data visualisation and data collection</p></li>
<li><p>dimension reduction and clustering</p></li>
</ul>
<p>We’ll use a 25-colour flow cytometry dataset of innate lymphoid cells
(ILC), natural killer cells (NK), and innate-like T cells in human
intestinal tissues, after excluding dead cells, doublets, mast cells and
basophils, hematopoietic stem cells, myeloid cells and granulocytes, and
B cells.</p>
<p>The original data was published as is available in <a
href="https://pubmed.ncbi.nlm.nih.gov/35018731/">OMIP-082</a></p>
</div>
<div id="setup" class="section level1 tabset tabset-fade">
<h1 class="tabset tabset-fade">Setup</h1>
<p>We’ll take you through steps from the beginning, installing packages
and loading data.</p>
<div id="preparation-packages-and-directories" class="section level2">
<h2>Preparation: packages and directories</h2>
<p>Packages are collections of functions. Many are not in-built in the
base R but they do pretty cool things, enabling generating pretty plots,
performing data wrangling in an efficient way, etc. R also has a large
developer community, with many community-developed packages available on
platforms like <code>Bioconductor</code> or <code>GitHub</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Install packages used in this workshop.</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co">#install.packages(&quot;tidyverse&quot;)</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co">#install.packages(&quot;ggridges&quot;)</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co">#install.packages(&quot;GGally&quot;)</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co">#install.packages(&quot;cowplot&quot;)</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co">#install.packages(&quot;data.table&quot;)</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co">#install.packages(&quot;BiocManager&quot;)</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co">#BiocManager::install(&quot;FuseSOM&quot;)</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="co">#BiocManager::install(&quot;FlowSOM&quot;)</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co">#if(!require(&#39;remotes&#39;)) {install.packages(&#39;remotes&#39;)}</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="co">#remotes::install_github(repo = &quot;immunedynamics/spectre&quot;)</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="co"># Load packages</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a><span class="fu">library</span>(ggridges)</span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="fu">library</span>(Spectre)</span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a><span class="fu">library</span>(FuseSOM)</span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a><span class="fu">library</span>(FlowSOM)</span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a><span class="fu">library</span>(GGally)</span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a><span class="fu">library</span>(umap)</span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a><span class="fu">library</span>(Rtsne)</span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a></span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_classic</span>())</span></code></pre></div>
<p>Create some folders for plots and data.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">&quot;plots&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">&quot;data&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">&quot;https://raw.githubusercontent.com/DrThomasOneil/Digital-Research-Skills-Network/refs/heads/main/docs/assets/downloads/example_flowdata_omip082_20k.csv&quot;</span>, <span class="at">destfile =</span> <span class="st">&quot;data/example_flowdata_omip082_20k.csv&quot;</span>)</span></code></pre></div>
<details>
<summary>
Setting seed is crucial for reproducibility:
</summary>
<p>It’s a number of your choice, which pins down the “way” that how the
random process will be performed when you re-run these codes. (Easier
example: Say if you randomly take 10 numbers with a mean of 50, there
will be tonnes of ways to generate the numbers. But a seed ensures that
the report always returns the same set of numbers. It’s the similar idea
when you run other algorithms including randomness, e.g. clustering and
generating a UMAP).</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1337</span>)</span></code></pre></div>
</details>
</div>
<div id="load-flow-cytometry-data" class="section level2">
<h2>Load flow cytometry data</h2>
<p>The expression data is stored in a <code>.csv</code> file. We’ll
first load this in.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Read the .csv file in the working directory.</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>ILC <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;data/example_flowdata_omip082_20k.csv&quot;</span>)</span></code></pre></div>
<p>Have a look at the broad features of this dataset. In this dataset,
the channel names are stored as column names. Each row represents a
cell, storing expression values in each channel.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Dimension: number of cells x number of channels.</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">dim</span>(ILC)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co"># Check the column names.</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="fu">colnames</span>(ILC)</span></code></pre></div>
<p>The column names are wordy, with markers, fluorophores, and channels.
We just need the marker names to proceed. The first column ‘X’ is just
indexing the number of rows. It’s not necessary either.<br></p>
<p>Remove the first column.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># This keeps the 2-33 columns in the dataset, and store the change.</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>ILC <span class="ot">&lt;-</span> ILC[, <span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">colnames</span>(ILC)</span></code></pre></div>
<pre><code> [1] &quot;FSC.A&quot;                        &quot;FSC.H&quot;                       
 [3] &quot;FSC.W&quot;                        &quot;SSC.A&quot;                       
 [5] &quot;SSC.H&quot;                        &quot;SSC.W&quot;                       
 [7] &quot;NKp44_BB515.B515_30.A&quot;        &quot;Lineage_PerCP.Cy55.B710_50.A&quot;
 [9] &quot;CD45_BB790.B820_60.A&quot;         &quot;CD103_APC.R670_30.A&quot;         
[11] &quot;FVS_700.R710_40.A&quot;            &quot;CD4_APC_Cy7.R820_60.A&quot;       
[13] &quot;HLA.DR_BUV395.U379_28.A&quot;      &quot;CD16_BUV496.U515_30.A&quot;       
[15] &quot;CD3_BUV563.U586_15.A&quot;         &quot;CD94_BUV615.U610_20.A&quot;       
[17] &quot;CD39_BUV661.U670_30.A&quot;        &quot;CD56_BUV737.U740_35.A&quot;       
[19] &quot;TCRVA72_BUV805.U820_60.A&quot;     &quot;CKIT_BV421.V427_25.A&quot;        
[21] &quot;CD14_CD19_BV480.V474_25.A&quot;    &quot;CD49D_BV510.V525_50.A&quot;       
[23] &quot;CD244_SB600.V610_20.A&quot;        &quot;CD161_B650.V677_20.A&quot;        
[25] &quot;TCRgt_BV711.V710_50.A&quot;        &quot;CD69_BV750.V750_30.A&quot;        
[27] &quot;CD25_BV786.V820_60.A&quot;         &quot;ITAGB7_PE.Y586_15.A&quot;         
[29] &quot;CRTH2_PEDAZZLE594.Y610_20.A&quot;  &quot;NKp46_PECy5.Y670_30.A&quot;       
[31] &quot;CD127_PECy7.Y820_60.A&quot;        &quot;Time&quot;                        </code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Same with it, you can just enter the column indexes from 2-33.</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co"># ILC &lt;-ILC[,c(2:33)]</span></span></code></pre></div>
<p>Change the column names.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># Substitute everything after the first &#39;_&#39; with an empty &quot;&quot; in the column names that have the pattern &quot;_&quot; and something.</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">colnames</span>(ILC) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">&quot;_.*&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="fu">colnames</span>(ILC)) <span class="co">#&#39;_&#39; is the exact character in the names, the &#39;.&#39; means &#39;any character&#39;, &#39;*&#39; means &#39;any of the preceding things&#39;. </span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="fu">colnames</span>(ILC)</span></code></pre></div>
<pre><code> [1] &quot;FSC.A&quot;   &quot;FSC.H&quot;   &quot;FSC.W&quot;   &quot;SSC.A&quot;   &quot;SSC.H&quot;   &quot;SSC.W&quot;   &quot;NKp44&quot;  
 [8] &quot;Lineage&quot; &quot;CD45&quot;    &quot;CD103&quot;   &quot;FVS&quot;     &quot;CD4&quot;     &quot;HLA.DR&quot;  &quot;CD16&quot;   
[15] &quot;CD3&quot;     &quot;CD94&quot;    &quot;CD39&quot;    &quot;CD56&quot;    &quot;TCRVA72&quot; &quot;CKIT&quot;    &quot;CD14&quot;   
[22] &quot;CD49D&quot;   &quot;CD244&quot;   &quot;CD161&quot;   &quot;TCRgt&quot;   &quot;CD69&quot;    &quot;CD25&quot;    &quot;ITAGB7&quot; 
[29] &quot;CRTH2&quot;   &quot;NKp46&quot;   &quot;CD127&quot;   &quot;Time&quot;   </code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># Fix up the broken names FVS700 and CD14&amp;CD19</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co"># This tells R to: find which of the column equals &quot;FVS&quot; among all columns in the data, then change it to FVS700.</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="fu">colnames</span>(ILC)[<span class="fu">colnames</span>(ILC) <span class="sc">==</span> <span class="st">&quot;FVS&quot;</span>] <span class="ot">&lt;-</span> <span class="st">&quot;FVS700&quot;</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="fu">colnames</span>(ILC)[<span class="fu">colnames</span>(ILC) <span class="sc">==</span> <span class="st">&quot;CD14&quot;</span>] <span class="ot">&lt;-</span> <span class="st">&quot;CD14.CD19&quot;</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co"># Have a look</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="fu">colnames</span>(ILC)</span></code></pre></div>
<pre><code> [1] &quot;FSC.A&quot;     &quot;FSC.H&quot;     &quot;FSC.W&quot;     &quot;SSC.A&quot;     &quot;SSC.H&quot;     &quot;SSC.W&quot;    
 [7] &quot;NKp44&quot;     &quot;Lineage&quot;   &quot;CD45&quot;      &quot;CD103&quot;     &quot;FVS700&quot;    &quot;CD4&quot;      
[13] &quot;HLA.DR&quot;    &quot;CD16&quot;      &quot;CD3&quot;       &quot;CD94&quot;      &quot;CD39&quot;      &quot;CD56&quot;     
[19] &quot;TCRVA72&quot;   &quot;CKIT&quot;      &quot;CD14.CD19&quot; &quot;CD49D&quot;     &quot;CD244&quot;     &quot;CD161&quot;    
[25] &quot;TCRgt&quot;     &quot;CD69&quot;      &quot;CD25&quot;      &quot;ITAGB7&quot;    &quot;CRTH2&quot;     &quot;NKp46&quot;    
[31] &quot;CD127&quot;     &quot;Time&quot;     </code></pre>
</div>
<div id="time-gate-qc." class="section level2">
<h2>Time gate QC.</h2>
<p>The data is already preprocessed in FlowJo. But if you are
interested, you can view the Time gate, and filter the data based on
this in R.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">ggplot</span>(ILC, <span class="fu">aes</span>(<span class="at">x =</span> Time, <span class="at">y =</span> FSC.A))<span class="sc">+</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>  <span class="fu">geom_bin_2d</span>(<span class="at">bins =</span> <span class="dv">200</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>    <span class="fu">scale_fill_gradientn</span>(<span class="at">colours =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;green&quot;</span>, <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;red&quot;</span>))<span class="sc">+</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">4600000</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># Filter out the cells above the threshold.</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>ILC <span class="ot">&lt;-</span> ILC[ILC<span class="sc">$</span>Time <span class="sc">&lt;</span> <span class="dv">4600000</span>,]</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="fu">ggplot</span>(ILC, <span class="fu">aes</span>(<span class="at">x =</span> Time, <span class="at">y =</span> FSC.A))<span class="sc">+</span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>  <span class="fu">geom_bin_2d</span>(<span class="at">bins =</span> <span class="dv">200</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>    <span class="fu">scale_fill_gradientn</span>(<span class="at">colours =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;green&quot;</span>, <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;red&quot;</span>))</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-8-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="visualisations" class="section level1">
<h1>Visualisations</h1>
<div id="transformations" class="section level2">
<h2>Transformations</h2>
<p>Raw Flow Cytometry data looks bad!</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>ILC <span class="sc">%&gt;%</span> <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>CD4, <span class="at">y=</span>CD3))<span class="sc">+</span>        </span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>So we transform the data to make it look better. In FlowJo, when you
adjust the axes, you can see several transformations available. One is
arcsinh (the inverse of hyperbolic sin, if you want to get mathy).
Smaller values centre around zero, and larger values separate, giving
the appearance that we’re familiar with.</p>
<p>We are also able to adjust the level of transformation by dividing
the values by some cofactor.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co"># decide on a cofactor - try changing it and seeing how the data changes</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>cofactor <span class="ot">=</span> <span class="dv">500</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>ILC_transf <span class="ot">&lt;-</span> <span class="fu">asinh</span>(ILC[,<span class="dv">7</span><span class="sc">:</span><span class="dv">31</span>]<span class="sc">/</span>cofactor) <span class="co">#7:31 are the columns with markers.</span></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a>ILC_transf <span class="ot">&lt;-</span> <span class="fu">cbind</span>(ILC_transf, ILC[,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>,<span class="dv">32</span>)]) <span class="co"># Add the first 6 columns &amp; time back in.</span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span> <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>CD4, <span class="at">y=</span>CD3))<span class="sc">+</span></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a>  <span class="fu">geom_bin_2d</span>(<span class="at">bins =</span> <span class="dv">200</span>, <span class="at">show.legend=</span>T)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-10-1.png" width="960" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co"># Initialize an empty list to store the plots</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="co"># Define the x-axis marker for the plots</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>xaxis <span class="ot">=</span> <span class="st">&quot;CD39&quot;</span> <span class="co"># Choose an x-axis marker for comparison</span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a><span class="co"># Loop through each column in the dataset</span></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(ILC_transf)) {</span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a>  <span class="co"># Get the marker name for the current column</span></span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a>  Marker <span class="ot">&lt;-</span> <span class="fu">colnames</span>(ILC_transf)[i]</span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a>  </span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a>  <span class="co"># Generate a 2D bin plot for the current marker against the chosen x-axis marker</span></span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a>  p <span class="ot">&lt;-</span> ILC_transf <span class="sc">%&gt;%</span> </span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a>        <span class="fu">ggplot</span>(<span class="fu">aes_string</span>(<span class="at">x =</span> xaxis, <span class="at">y =</span> Marker)) <span class="sc">+</span>        </span>
<span id="cb17-14"><a href="#cb17-14" tabindex="-1"></a>        <span class="fu">geom_bin_2d</span>(<span class="at">bins =</span> <span class="dv">200</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span> <span class="co"># Create a heatmap-style scatter plot</span></span>
<span id="cb17-15"><a href="#cb17-15" tabindex="-1"></a>        <span class="fu">scale_fill_gradientn</span>(<span class="at">colors =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;red&quot;</span>)) <span class="co"># Define the color gradient</span></span>
<span id="cb17-16"><a href="#cb17-16" tabindex="-1"></a>  </span>
<span id="cb17-17"><a href="#cb17-17" tabindex="-1"></a>  <span class="co"># Add the generated plot to the list</span></span>
<span id="cb17-18"><a href="#cb17-18" tabindex="-1"></a>  plots[[i]] <span class="ot">&lt;-</span> p</span>
<span id="cb17-19"><a href="#cb17-19" tabindex="-1"></a>}</span>
<span id="cb17-20"><a href="#cb17-20" tabindex="-1"></a></span>
<span id="cb17-21"><a href="#cb17-21" tabindex="-1"></a><span class="co"># Combine all the plots into a grid layout and display them</span></span>
<span id="cb17-22"><a href="#cb17-22" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plots)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-11-1.png" width="960" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">rm</span>(plots)</span></code></pre></div>
<p>This code demonstrates how you would manually transform individual
columns in the dataset using different cofactors for each marker. It
creates a new dataset (<code>ILC_transf_indiv</code>) where specific
columns are transformed using the <code>asinh</code> function with their
respective cofactors. This allows for more granular control over
transformations.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>ILC_transf_indiv <span class="ot">&lt;-</span> ILC</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>ILC_transf_indiv[, <span class="st">&quot;CD3&quot;</span>] <span class="ot">&lt;-</span> <span class="fu">asinh</span>(ILC[,<span class="st">&quot;CD3&quot;</span>]<span class="sc">/</span><span class="dv">1000</span>)</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>ILC_transf_indiv[, <span class="st">&quot;CD4&quot;</span>] <span class="ot">&lt;-</span> <span class="fu">asinh</span>(ILC[,<span class="st">&quot;CD4&quot;</span>]<span class="sc">/</span><span class="dv">500</span>)</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="co"># and so on...</span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a><span class="co"># Visualize it</span></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a>ILC_transf_indiv <span class="sc">%&gt;%</span> <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>CD3, <span class="at">y=</span>CD4))<span class="sc">+</span>        </span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a>        <span class="fu">geom_bin_2d</span>(<span class="at">bins =</span> <span class="dv">200</span>, <span class="at">show.legend=</span>F) <span class="sc">+</span></span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a>        <span class="fu">scale_fill_gradientn</span>(<span class="at">colors =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>,  <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;red&quot;</span>))</span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a></span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a><span class="fu">rm</span>(ILC_transf_indiv)</span></code></pre></div>
</div>
<div id="histograms" class="section level2">
<h2>Histograms</h2>
<p>We can visualize the marker expressions via histograms and density
plots.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="co"># For example, this generates a histogram of CD3 expression.</span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD3))<span class="sc">+</span></span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a>  <span class="co"># Plot the histogram, and change the fill colour to lightblue, the line colour to black.</span></span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">35</span>, <span class="at">fill =</span> <span class="st">&quot;lightblue&quot;</span>, <span class="at">colour =</span><span class="st">&quot;black&quot;</span>)<span class="sc">+</span></span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a>  <span class="co"># Add labels</span></span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Frequency&quot;</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="co"># A density plot.</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD3))<span class="sc">+</span></span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">linewidth =</span> <span class="fl">0.7</span>, <span class="at">colour =</span> <span class="st">&quot;red4&quot;</span>, <span class="fu">aes</span>(<span class="at">fill =</span> <span class="st">&quot;red&quot;</span>), <span class="at">alpha =</span> <span class="fl">0.5</span>,<span class="at">show.legend =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-13-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>We can loop through the channels and create a density plot for all 25
markers.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="co"># Initialise an empty list to store the plots</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>plots<span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a><span class="co"># Loop through each column in the dataset</span></span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(ILC_transf)){</span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a></span>
<span id="cb22-7"><a href="#cb22-7" tabindex="-1"></a>  <span class="co"># Get the marker name for the current column</span></span>
<span id="cb22-8"><a href="#cb22-8" tabindex="-1"></a>  Marker <span class="ot">&lt;-</span> <span class="fu">colnames</span>(ILC_transf)[i]</span>
<span id="cb22-9"><a href="#cb22-9" tabindex="-1"></a></span>
<span id="cb22-10"><a href="#cb22-10" tabindex="-1"></a>  <span class="co"># Generate the histogram for the current marker</span></span>
<span id="cb22-11"><a href="#cb22-11" tabindex="-1"></a>  p <span class="ot">&lt;-</span> ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb22-12"><a href="#cb22-12" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> .data[[Marker]]))<span class="sc">+</span></span>
<span id="cb22-13"><a href="#cb22-13" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">linewidth =</span> <span class="fl">0.7</span>, <span class="at">colour =</span> <span class="st">&quot;red4&quot;</span>, <span class="fu">aes</span>(<span class="at">fill =</span> <span class="st">&quot;red&quot;</span>), <span class="at">alpha =</span> <span class="fl">0.5</span>,<span class="at">show.legend =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb22-14"><a href="#cb22-14" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x=</span>Marker,<span class="at">y=</span><span class="st">&quot;&quot;</span>)<span class="sc">+</span></span>
<span id="cb22-15"><a href="#cb22-15" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">1</span>))</span>
<span id="cb22-16"><a href="#cb22-16" tabindex="-1"></a></span>
<span id="cb22-17"><a href="#cb22-17" tabindex="-1"></a>  <span class="co"># Add the generated plot to the list</span></span>
<span id="cb22-18"><a href="#cb22-18" tabindex="-1"></a>  plots[[i]] <span class="ot">&lt;-</span> p</span>
<span id="cb22-19"><a href="#cb22-19" tabindex="-1"></a>}</span>
<span id="cb22-20"><a href="#cb22-20" tabindex="-1"></a></span>
<span id="cb22-21"><a href="#cb22-21" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plots[<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>])</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="fu">rm</span>(plots)</span></code></pre></div>
<p><br></p>
</div>
<div id="scatter-plot" class="section level2">
<h2>Scatter plot</h2>
<p>Like what we do in Flowjo, we can visualize 2 markers’ expressions,
We can generate scatterplots.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD4, <span class="at">y =</span> CD3))<span class="sc">+</span></span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color=</span><span class="st">&quot;blue&quot;</span>, <span class="at">alpha=</span><span class="fl">0.1</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="co"># Generate a scatterplot with a density overlay.</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span> <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>CD4, <span class="at">y=</span>CD3))<span class="sc">+</span>        </span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a>  <span class="fu">geom_bin_2d</span>(<span class="at">bins =</span> <span class="dv">200</span>, <span class="at">show.legend =</span> F)<span class="sc">+</span></span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>  <span class="fu">scale_fill_gradientn</span>(<span class="at">colors =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>,  <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;red&quot;</span>))</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-15-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="contour-plot" class="section level2">
<h2>Contour plot</h2>
<p>And a 2-marker countour plot. You can easily make it more stylish by
adding layers in ggplot!</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="co"># A contour plot of CD3 and CD4 expression values.</span></span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD3, <span class="at">y =</span> CD4))<span class="sc">+</span></span>
<span id="cb26-5"><a href="#cb26-5" tabindex="-1"></a>  <span class="fu">geom_density_2d</span>(<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, <span class="at">bins=</span><span class="dv">50</span>)<span class="sc">+</span></span>
<span id="cb26-6"><a href="#cb26-6" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;CD3 BUV563&quot;</span>)<span class="sc">+</span></span>
<span id="cb26-7"><a href="#cb26-7" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;CD4 APC-Cy7&quot;</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a></span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a><span class="co"># A contour plot of CD3 and CD4 expression over a scatter plot.</span></span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD3, <span class="at">y =</span> CD4))<span class="sc">+</span>        </span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">alpha=</span><span class="fl">0.5</span>)<span class="sc">+</span></span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a>  <span class="fu">geom_density_2d</span>(<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, <span class="at">bins=</span><span class="dv">50</span>, <span class="at">alpha=</span><span class="fl">0.5</span>)<span class="sc">+</span></span>
<span id="cb27-7"><a href="#cb27-7" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;CD3 BUV563&quot;</span>)<span class="sc">+</span></span>
<span id="cb27-8"><a href="#cb27-8" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;CD4 APC-Cy7&quot;</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-16-2.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a></span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a><span class="co"># A contour plot of CD3 and CD4 expression over a density plot.</span></span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb28-4"><a href="#cb28-4" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD3, <span class="at">y =</span> CD4))<span class="sc">+</span>        </span>
<span id="cb28-5"><a href="#cb28-5" tabindex="-1"></a>  <span class="fu">geom_bin_2d</span>(<span class="at">bins =</span> <span class="dv">200</span>)<span class="sc">+</span></span>
<span id="cb28-6"><a href="#cb28-6" tabindex="-1"></a>  <span class="fu">scale_fill_gradientn</span>(<span class="at">colors =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>,  <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;red&quot;</span>))<span class="sc">+</span></span>
<span id="cb28-7"><a href="#cb28-7" tabindex="-1"></a>  <span class="fu">geom_density_2d</span>(<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>)<span class="sc">+</span></span>
<span id="cb28-8"><a href="#cb28-8" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;CD3 BUV563&quot;</span>)<span class="sc">+</span></span>
<span id="cb28-9"><a href="#cb28-9" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;CD4 APC-Cy7&quot;</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-16-3.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="subsetting-data-and-summary-statistics"
class="section level1 tabset tabset-fade">
<h1 class="tabset tabset-fade">Subsetting data and summary
statistics</h1>
<p>While not as efficient as FlowJo, we can subset data after exporting.
We wouldn’t normally do this in R, but we’ll show you how you could
subset in R and for the purposes of visualisation.</p>
<div id="subsetting" class="section level2">
<h2>Subsetting</h2>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="co"># Plot a 2-marker plot of CD3 and CD4 expression values. </span></span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD3, <span class="at">y =</span> CD4))<span class="sc">+</span>        </span>
<span id="cb29-4"><a href="#cb29-4" tabindex="-1"></a>  <span class="fu">geom_bin_2d</span>(<span class="at">bins =</span> <span class="dv">200</span>)<span class="sc">+</span></span>
<span id="cb29-5"><a href="#cb29-5" tabindex="-1"></a>  <span class="fu">scale_fill_gradientn</span>(<span class="at">colors =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>,  <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;red&quot;</span>))<span class="sc">+</span></span>
<span id="cb29-6"><a href="#cb29-6" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;CD3 BUV563&quot;</span>)<span class="sc">+</span></span>
<span id="cb29-7"><a href="#cb29-7" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;CD4 APC-Cy7&quot;</span>)<span class="sc">+</span></span>
<span id="cb29-8"><a href="#cb29-8" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">1.3</span>)<span class="sc">+</span> <span class="co">#choose the x intercept</span></span>
<span id="cb29-9"><a href="#cb29-9" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">0.6</span>)  <span class="co">#choose the y intercept</span></span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Subset the dataset based on these thresholds.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="co"># This code creates a new column named &#39;subset&#39; in the &#39;ILC_transf&#39; dataset.</span></span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a><span class="co"># The &#39;subset&#39; column categorises rows based on the values of &#39;CD3&#39; and &#39;CD4&#39; columns:</span></span>
<span id="cb30-3"><a href="#cb30-3" tabindex="-1"></a><span class="co"># - &quot;CD3+CD4+&quot;: if CD3 &gt; 1.3 and CD4 &gt; 0.6</span></span>
<span id="cb30-4"><a href="#cb30-4" tabindex="-1"></a><span class="co"># - &quot;CD3+CD4-&quot;: if CD3 &gt; 1.3 and CD4 &lt;= 0.6</span></span>
<span id="cb30-5"><a href="#cb30-5" tabindex="-1"></a><span class="co"># - &quot;CD3-CD4+&quot;: if CD3 &lt;= 1.3 and CD4 &gt; 0.6</span></span>
<span id="cb30-6"><a href="#cb30-6" tabindex="-1"></a><span class="co"># - &quot;CD3-CD4-&quot;: if CD3 &lt;= 1.3 and CD4 &lt;= 0.6</span></span>
<span id="cb30-7"><a href="#cb30-7" tabindex="-1"></a><span class="co"># The &#39;mutate&#39; function from the dplyr package is used to add this new column.</span></span>
<span id="cb30-8"><a href="#cb30-8" tabindex="-1"></a></span>
<span id="cb30-9"><a href="#cb30-9" tabindex="-1"></a>ILC_transf <span class="ot">&lt;-</span> ILC_transf <span class="sc">%&gt;%</span> </span>
<span id="cb30-10"><a href="#cb30-10" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">subset =</span> <span class="fu">case_when</span>(</span>
<span id="cb30-11"><a href="#cb30-11" tabindex="-1"></a>    CD3 <span class="sc">&gt;</span> <span class="fl">1.3</span> <span class="sc">&amp;</span> CD4 <span class="sc">&gt;</span> <span class="fl">0.6</span> <span class="sc">~</span> <span class="st">&quot;CD3+CD4+&quot;</span>,</span>
<span id="cb30-12"><a href="#cb30-12" tabindex="-1"></a>    CD3 <span class="sc">&gt;</span> <span class="fl">1.3</span> <span class="sc">&amp;</span> CD4 <span class="sc">&lt;=</span> <span class="fl">0.6</span> <span class="sc">~</span> <span class="st">&quot;CD3+CD4-&quot;</span>,</span>
<span id="cb30-13"><a href="#cb30-13" tabindex="-1"></a>    CD3 <span class="sc">&lt;=</span> <span class="fl">1.3</span> <span class="sc">&amp;</span> CD4 <span class="sc">&gt;</span> <span class="fl">0.6</span> <span class="sc">~</span> <span class="st">&quot;CD3-CD4+&quot;</span>,</span>
<span id="cb30-14"><a href="#cb30-14" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">&quot;CD3-CD4-&quot;</span></span>
<span id="cb30-15"><a href="#cb30-15" tabindex="-1"></a>  ))</span></code></pre></div>
<div id="visualise-the-subsets-in-dot-plots" class="section level3">
<h3>Visualise the subsets in dot plots</h3>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD3, <span class="at">y =</span> CD4, <span class="at">color=</span>subset))<span class="sc">+</span>        </span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a>  <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb31-4"><a href="#cb31-4" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;CD3 BUV563&quot;</span>)<span class="sc">+</span></span>
<span id="cb31-5"><a href="#cb31-5" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;CD4 APC-Cy7&quot;</span>)<span class="sc">+</span><span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">1.3</span>)<span class="sc">+</span></span>
<span id="cb31-6"><a href="#cb31-6" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">0.6</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="co"># Look at different markers coloured by subset</span></span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb32-3"><a href="#cb32-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD56, <span class="at">y =</span> CD16, <span class="at">color=</span>subset))<span class="sc">+</span>        </span>
<span id="cb32-4"><a href="#cb32-4" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-19-2.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="co"># Split by subset</span></span>
<span id="cb33-2"><a href="#cb33-2" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb33-3"><a href="#cb33-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span>CD56, <span class="at">y =</span> CD16))<span class="sc">+</span>        </span>
<span id="cb33-4"><a href="#cb33-4" tabindex="-1"></a>  <span class="fu">geom_bin_2d</span>(<span class="at">bins =</span> <span class="dv">200</span>)<span class="sc">+</span></span>
<span id="cb33-5"><a href="#cb33-5" tabindex="-1"></a>  <span class="fu">scale_fill_gradientn</span>(<span class="at">colors =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>,  <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;red&quot;</span>))<span class="sc">+</span></span>
<span id="cb33-6"><a href="#cb33-6" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>subset)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-19-3.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span>FSC.A, <span class="at">y =</span> SSC.A))<span class="sc">+</span>        </span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a>  <span class="fu">geom_bin_2d</span>(<span class="at">bins =</span> <span class="dv">200</span>)<span class="sc">+</span></span>
<span id="cb34-4"><a href="#cb34-4" tabindex="-1"></a>  <span class="fu">scale_fill_gradientn</span>(<span class="at">colors =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>,  <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;red&quot;</span>))<span class="sc">+</span></span>
<span id="cb34-5"><a href="#cb34-5" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>subset)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-19-4.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="histograms-1" class="section level3">
<h3>Histograms</h3>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb35-2"><a href="#cb35-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD56, <span class="at">color =</span> subset, <span class="at">fill =</span> subset)) <span class="sc">+</span></span>
<span id="cb35-3"><a href="#cb35-3" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>, <span class="at">linewidth =</span> <span class="dv">1</span>)<span class="sc">+</span></span>
<span id="cb35-4"><a href="#cb35-4" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fu">quantile</span>(ILC_transf<span class="sc">$</span>CD56, <span class="fl">0.001</span>),<span class="fu">quantile</span>(ILC_transf<span class="sc">$</span>CD56, <span class="fl">0.999</span>))</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="co">#using ggridges</span></span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb36-3"><a href="#cb36-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD56, <span class="at">y =</span> subset, <span class="at">fill =</span> subset)) <span class="sc">+</span></span>
<span id="cb36-4"><a href="#cb36-4" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>(<span class="at">scale =</span> <span class="fl">1.2</span>, <span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb36-5"><a href="#cb36-5" tabindex="-1"></a>  <span class="fu">theme_ridges</span>() <span class="sc">+</span></span>
<span id="cb36-6"><a href="#cb36-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)<span class="sc">+</span></span>
<span id="cb36-7"><a href="#cb36-7" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">&quot;&quot;</span>, <span class="at">x=</span><span class="st">&quot;&quot;</span>, <span class="at">title =</span> <span class="st">&quot;CD56&quot;</span>)<span class="sc">+</span></span>
<span id="cb36-8"><a href="#cb36-8" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fu">quantile</span>(ILC_transf<span class="sc">$</span>CD56, <span class="fl">0.001</span>),<span class="fu">quantile</span>(ILC_transf<span class="sc">$</span>CD56, <span class="fl">0.999</span>))</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-20-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>You could mass print these using a loop, as we did for the density
plots above. You can also adjust the marker you output, as not all
markers or combinations of markers are useful.</p>
</div>
</div>
<div id="summary-statistics" class="section level2">
<h2>Summary statistics</h2>
<p>While subsetting is not as efficient as FlowJo, summarising data
is!</p>
<p>We can first collect subset percentages and output them in a table,
in text, and even in a plot. Then we can calculate summary statistics
like MFI &amp; gMFI.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="co"># total summary</span></span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">table</span>(ILC_transf<span class="sc">$</span>subset)<span class="sc">*</span><span class="dv">100</span><span class="sc">/</span><span class="fu">nrow</span>(ILC_transf),<span class="dv">2</span>) <span class="co"># % of each subset</span></span></code></pre></div>
<pre><code>
CD3-CD4- CD3-CD4+ CD3+CD4- CD3+CD4+ 
    6.56     0.22    17.79    75.44 </code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="co"># individual marker</span></span>
<span id="cb39-2"><a href="#cb39-2" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span> </span>
<span id="cb39-3"><a href="#cb39-3" tabindex="-1"></a>  <span class="fu">group_by</span>(subset) <span class="sc">%&gt;%</span> </span>
<span id="cb39-4"><a href="#cb39-4" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">Percent =</span> <span class="fu">round</span>(<span class="fu">n</span>()<span class="sc">/</span><span class="fu">nrow</span>(ILC_transf)<span class="sc">*</span><span class="dv">100</span>,<span class="dv">2</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb39-5"><a href="#cb39-5" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">subset =</span> <span class="fu">factor</span>(subset, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;CD3+CD4+&quot;</span>, <span class="st">&quot;CD3+CD4-&quot;</span>, <span class="st">&quot;CD3-CD4+&quot;</span>, <span class="st">&quot;CD3-CD4-&quot;</span>)))<span class="sc">%&gt;%</span></span>
<span id="cb39-6"><a href="#cb39-6" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Subsets =</span> <span class="st">&quot;Subsets&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-7"><a href="#cb39-7" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>Subsets,<span class="at">y =</span> Percent, <span class="at">fill =</span> subset)) <span class="sc">+</span></span>
<span id="cb39-8"><a href="#cb39-8" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&quot;identity&quot;</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb39-9"><a href="#cb39-9" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">&quot;Percentage&quot;</span>) <span class="sc">+</span></span>
<span id="cb39-10"><a href="#cb39-10" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">&quot;&quot;</span>) <span class="sc">+</span></span>
<span id="cb39-11"><a href="#cb39-11" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;CD3+CD4+&quot;</span> <span class="ot">=</span> <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;CD3+CD4-&quot;</span> <span class="ot">=</span> <span class="st">&quot;red&quot;</span>, <span class="st">&quot;CD3-CD4+&quot;</span> <span class="ot">=</span> <span class="st">&quot;green&quot;</span>, <span class="st">&quot;CD3-CD4-&quot;</span> <span class="ot">=</span> <span class="st">&quot;yellow&quot;</span>)) <span class="sc">+</span></span>
<span id="cb39-12"><a href="#cb39-12" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb39-13"><a href="#cb39-13" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb39-14"><a href="#cb39-14" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>) <span class="sc">+</span></span>
<span id="cb39-15"><a href="#cb39-15" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">size=</span><span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb39-16"><a href="#cb39-16" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">&quot;Subsets of CD3+/-CD4+/- cells&quot;</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We can also calculate the percentage of subsets in line, meaning we
don’t have to change these lines of text based on different results or
different data: - CD3+CD4+: 75.44%</p>
<ul>
<li><p>CD3+CD4-: 17.79%</p></li>
<li><p>CD3-CD4+: 0.22%</p></li>
<li><p>CD3-CD4-: 6.56%</p></li>
</ul>
<details>
<summary>
<strong>You can also quickly calculate summary statistics like MFI,
gMFI, medianFI</strong>:
</summary>
<p><strong>Arithmatic MFI</strong></p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a>ILC<span class="sc">$</span>subset  <span class="ot">&lt;-</span> ILC_transf<span class="sc">$</span>subset <span class="co"># Add the subset column to the original data.</span></span>
<span id="cb40-2"><a href="#cb40-2" tabindex="-1"></a></span>
<span id="cb40-3"><a href="#cb40-3" tabindex="-1"></a>ILC <span class="sc">%&gt;%</span> </span>
<span id="cb40-4"><a href="#cb40-4" tabindex="-1"></a>  <span class="fu">group_by</span>(subset) <span class="sc">%&gt;%</span> </span>
<span id="cb40-5"><a href="#cb40-5" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="dv">7</span><span class="sc">:</span><span class="dv">31</span>, mean, <span class="at">.names =</span> <span class="st">&quot;MFI_{col}&quot;</span>))</span></code></pre></div>
<pre><code># A tibble: 4 × 26
  subset  MFI_NKp44 MFI_Lineage MFI_CD45 MFI_CD103 MFI_FVS700 MFI_CD4 MFI_HLA.DR
  &lt;chr&gt;       &lt;dbl&gt;       &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;
1 CD3+CD…      49.9        119.    9901.      591.     -40.0   1419.        176.
2 CD3+CD…      55.5        131.   10543.      865.     -42.8   -115.        193.
3 CD3-CD…     113.         491.    6915.      431.      -5.70   918.      15030.
4 CD3-CD…     250.         242.    4114.     1158.     -64.6    -69.4      1247.
# ℹ 18 more variables: MFI_CD16 &lt;dbl&gt;, MFI_CD3 &lt;dbl&gt;, MFI_CD94 &lt;dbl&gt;,
#   MFI_CD39 &lt;dbl&gt;, MFI_CD56 &lt;dbl&gt;, MFI_TCRVA72 &lt;dbl&gt;, MFI_CKIT &lt;dbl&gt;,
#   MFI_CD14.CD19 &lt;dbl&gt;, MFI_CD49D &lt;dbl&gt;, MFI_CD244 &lt;dbl&gt;, MFI_CD161 &lt;dbl&gt;,
#   MFI_TCRgt &lt;dbl&gt;, MFI_CD69 &lt;dbl&gt;, MFI_CD25 &lt;dbl&gt;, MFI_ITAGB7 &lt;dbl&gt;,
#   MFI_CRTH2 &lt;dbl&gt;, MFI_NKp46 &lt;dbl&gt;, MFI_CD127 &lt;dbl&gt;</code></pre>
<p><strong>Median FI</strong></p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a>ILC <span class="sc">%&gt;%</span> </span>
<span id="cb42-2"><a href="#cb42-2" tabindex="-1"></a>  <span class="fu">group_by</span>(subset) <span class="sc">%&gt;%</span> </span>
<span id="cb42-3"><a href="#cb42-3" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="dv">7</span><span class="sc">:</span><span class="dv">31</span>, mean, <span class="at">.names =</span> <span class="st">&quot;MFI_{col}&quot;</span>))</span></code></pre></div>
<pre><code># A tibble: 4 × 26
  subset  MFI_NKp44 MFI_Lineage MFI_CD45 MFI_CD103 MFI_FVS700 MFI_CD4 MFI_HLA.DR
  &lt;chr&gt;       &lt;dbl&gt;       &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;
1 CD3+CD…      49.9        119.    9901.      591.     -40.0   1419.        176.
2 CD3+CD…      55.5        131.   10543.      865.     -42.8   -115.        193.
3 CD3-CD…     113.         491.    6915.      431.      -5.70   918.      15030.
4 CD3-CD…     250.         242.    4114.     1158.     -64.6    -69.4      1247.
# ℹ 18 more variables: MFI_CD16 &lt;dbl&gt;, MFI_CD3 &lt;dbl&gt;, MFI_CD94 &lt;dbl&gt;,
#   MFI_CD39 &lt;dbl&gt;, MFI_CD56 &lt;dbl&gt;, MFI_TCRVA72 &lt;dbl&gt;, MFI_CKIT &lt;dbl&gt;,
#   MFI_CD14.CD19 &lt;dbl&gt;, MFI_CD49D &lt;dbl&gt;, MFI_CD244 &lt;dbl&gt;, MFI_CD161 &lt;dbl&gt;,
#   MFI_TCRgt &lt;dbl&gt;, MFI_CD69 &lt;dbl&gt;, MFI_CD25 &lt;dbl&gt;, MFI_ITAGB7 &lt;dbl&gt;,
#   MFI_CRTH2 &lt;dbl&gt;, MFI_NKp46 &lt;dbl&gt;, MFI_CD127 &lt;dbl&gt;</code></pre>
<p><strong>geometric MFI</strong></p>
<p>The traditional formula written out would be like
<code>ILC_transf %&gt;% prod(CD56)^(1/nrow(CD56))</code>. However, as
our data is <em>compensated</em>, there are meaningful negative values
and 0s that we don’t want to get rid of, which if you run on this data
removes two subsets entirely. FlowJo calculates the <em>graph
spaces</em> as gMFI to enable it to deal with negatives and 0s. Here in
this case, our data is already asinh transformed to accomodate the -ve
and 0s and bizzare expression values. It may be better to stick with the
arithmatic mean instead of forcing the geometric mean.</p>
</details>
<p><br></p>
</div>
</div>
<div id="nxn-plots" class="section level1">
<h1>nxn plots</h1>
<p>You can also create a loop to view the nxn plots.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="co"># Select only numeric marker columns</span></span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a>marker_cols <span class="ot">&lt;-</span> <span class="fu">colnames</span>(ILC_transf)[<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>] <span class="co"># adjust if needed</span></span>
<span id="cb44-3"><a href="#cb44-3" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" tabindex="-1"></a><span class="co"># Make a customised function for plot streamline</span></span>
<span id="cb44-5"><a href="#cb44-5" tabindex="-1"></a>custom_bin_plot <span class="ot">&lt;-</span> <span class="cf">function</span>(data, mapping, ...) {</span>
<span id="cb44-6"><a href="#cb44-6" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">data =</span> data, <span class="at">mapping =</span> mapping) <span class="sc">+</span></span>
<span id="cb44-7"><a href="#cb44-7" tabindex="-1"></a>    <span class="fu">geom_bin_2d</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb44-8"><a href="#cb44-8" tabindex="-1"></a>    <span class="fu">scale_fill_gradientn</span>(<span class="at">colours =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;red&quot;</span>)) <span class="sc">+</span></span>
<span id="cb44-9"><a href="#cb44-9" tabindex="-1"></a>    <span class="fu">theme_minimal</span>()</span>
<span id="cb44-10"><a href="#cb44-10" tabindex="-1"></a>}</span>
<span id="cb44-11"><a href="#cb44-11" tabindex="-1"></a></span>
<span id="cb44-12"><a href="#cb44-12" tabindex="-1"></a><span class="co"># Generate plots</span></span>
<span id="cb44-13"><a href="#cb44-13" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggpairs</span>(</span>
<span id="cb44-14"><a href="#cb44-14" tabindex="-1"></a>  ILC_transf,</span>
<span id="cb44-15"><a href="#cb44-15" tabindex="-1"></a>  <span class="at">columns =</span> marker_cols,</span>
<span id="cb44-16"><a href="#cb44-16" tabindex="-1"></a>  <span class="at">lower =</span> <span class="fu">list</span>(<span class="at">continuous =</span> custom_bin_plot),</span>
<span id="cb44-17"><a href="#cb44-17" tabindex="-1"></a>  <span class="at">diag =</span> <span class="fu">list</span>(<span class="at">continuous =</span> <span class="fu">wrap</span>(<span class="st">&quot;densityDiag&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.4</span>)),</span>
<span id="cb44-18"><a href="#cb44-18" tabindex="-1"></a>  <span class="at">upper =</span> <span class="fu">list</span>(<span class="at">continuous =</span> <span class="fu">wrap</span>(<span class="st">&quot;cor&quot;</span>, <span class="at">size =</span> <span class="dv">3</span>))</span>
<span id="cb44-19"><a href="#cb44-19" tabindex="-1"></a>)</span>
<span id="cb44-20"><a href="#cb44-20" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">&quot;plots/nxn.png&quot;</span>,p, <span class="at">width=</span><span class="dv">25</span>, <span class="at">height=</span><span class="dv">25</span>,<span class="at">unit=</span><span class="st">&quot;in&quot;</span>)</span>
<span id="cb44-21"><a href="#cb44-21" tabindex="-1"></a></span>
<span id="cb44-22"><a href="#cb44-22" tabindex="-1"></a><span class="fu">rm</span>(p,custom_bin_plot, marker_cols)</span></code></pre></div>
</div>
<div id="dimension-reduction-and-clustering" class="section level1">
<h1>Dimension reduction and clustering</h1>
<p>There are several ways to perform clustering (e.g. FlowSOM, FuseSOM)
and dimension reduction (e.g. PCA, UMAP, tSNE).</p>
<p>In this section, we’ll cover <strong>clustering</strong> and
<strong>dimension reduction</strong> on our one sample.</p>
<div id="clustering" class="section level2 tabset tabset-fade">
<h2 class="tabset tabset-fade">Clustering</h2>
<p>Clustering groups our cells based on the input expressions we
choose.</p>
<div id="flowsom" class="section level3">
<h3>FlowSOM</h3>
<p><a
href="https://bioconductor.org/packages/release/bioc/html/FlowSOM.html">Read
more</a></p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="co"># Perform clustering</span></span>
<span id="cb45-2"><a href="#cb45-2" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> FlowSOM<span class="sc">::</span><span class="fu">FlowSOM</span>(</span>
<span id="cb45-3"><a href="#cb45-3" tabindex="-1"></a>  <span class="at">input=</span><span class="fu">as.matrix</span>(ILC_transf[,<span class="dv">1</span><span class="sc">:</span><span class="dv">32</span>]),</span>
<span id="cb45-4"><a href="#cb45-4" tabindex="-1"></a>  <span class="at">silent =</span> <span class="cn">FALSE</span>,</span>
<span id="cb45-5"><a href="#cb45-5" tabindex="-1"></a>  <span class="at">colsToUse =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">14</span>,<span class="dv">16</span><span class="sc">:</span><span class="dv">25</span>),</span>
<span id="cb45-6"><a href="#cb45-6" tabindex="-1"></a>  <span class="at">nClus =</span> <span class="dv">15</span>)</span>
<span id="cb45-7"><a href="#cb45-7" tabindex="-1"></a></span>
<span id="cb45-8"><a href="#cb45-8" tabindex="-1"></a><span class="co"># Annotate the cluster that the cell belongs to</span></span>
<span id="cb45-9"><a href="#cb45-9" tabindex="-1"></a>ILC_transf<span class="sc">$</span>flowSOM<span class="ot">&lt;-</span> <span class="fu">GetMetaclusters</span>(clusters)</span>
<span id="cb45-10"><a href="#cb45-10" tabindex="-1"></a>FlowSOM<span class="sc">::</span><span class="fu">GetMetaclusterMFIs</span>(clusters)</span></code></pre></div>
<pre><code>        NKp44   Lineage     CD45       CD103       FVS700          CD4
1  0.08767095 0.2075124 3.652728  0.66374838 -0.070054837  1.718446732
2  0.06092035 0.2608379 3.845011  0.48375735 -0.002694851  1.673946500
3  0.07769296 0.2096058 3.670606  0.44195807 -0.049197283  1.519431591
4  0.26814000 0.8422096 3.020440  0.25983192 -0.070561014  0.285355955
5  0.20384046 0.2494431 2.165039  0.04000831 -0.069922924 -0.159006745
6  0.07682344 0.4272733 2.800478 -0.15280791 -0.195778430 -0.178643271
7  0.15199903 0.5864034 2.634233  0.10985178 -0.073452704  0.001656005
8  0.10020243 0.2194812 3.694064  0.52694911 -0.055605981 -0.216673389
9  0.09238955 0.2335644 3.808754  1.61759275 -0.114163850 -0.282545939
10 0.09819077 0.2178801 3.599065  0.80909690 -0.121560473  1.810596168
11 0.12451813 0.1840846 3.692152  1.65920115 -0.104120027 -0.273976639
12 0.11406716 0.2838156 3.787179  1.99422741 -0.129626542  1.700919032
13 0.17637908 0.2869712 3.344381  1.70810831 -0.072664808 -0.280821741
14 0.70718268 0.4607808 3.019883  2.31473327 -0.219905265 -0.238584988
15 0.71015617 0.4519169 2.985281  2.12613797 -0.129041672 -0.132830180
      HLA.DR      CD16        CD3          CD94      CD39      CD56   TCRVA72
1  0.2585904 0.3014053 3.14893150 -0.0001283541 0.2069036 0.8791201 0.5654020
2  0.4412144 0.2544301 3.25660467 -0.0122394441 2.5714781 0.9501370 0.6968365
3  0.2215236 0.2948854 2.99801254 -0.0097803352 0.2119111 0.6668085 0.5375435
4  4.3851631 0.7932066 0.33294968  0.0760867558 2.9933397 0.8673116 0.4591199
5  0.4977376 0.3020615 0.08235943 -0.0044702403 0.2099833 1.9980837 0.1392810
6  0.4974132 0.4490478 0.11416380  0.6835254431 0.4359418 3.7976246 0.2433500
7  0.3032004 0.5342692 0.27236900  0.0273353718 0.5225141 0.3988376 0.1233970
8  0.2560977 0.3381878 3.09349370  0.0707866177 0.1877662 0.8110133 0.5478587
9  0.2631046 0.3468565 3.39602506  0.2483764440 0.2166930 1.3286474 0.5868405
10 0.2624349 0.3024804 3.27759159  0.0006247615 0.2403607 2.8603338 0.5975648
11 0.3739428 0.3097136 3.29970193  0.7232792079 0.4476766 1.0290836 3.3287402
12 0.3099208 0.3261117 3.23529434  0.1027751938 0.2045608 1.1476330 0.6706844
13 0.4888059 0.4864017 3.60711050  0.3930174112 2.1050779 1.8307076 0.4118729
14 0.4755351 0.5450906 0.13690468  0.3112615645 2.8756585 3.7010598 0.3551278
15 0.5272312 0.5735796 0.19247857  0.1627501398 3.3976748 1.1171057 0.2884356
        CKIT CD14.CD19     CD49D     CD244     CD161     TCRgt      CD69
1  0.1397609 0.1779661 0.2959321 0.6341562 2.3236103 0.2079571 3.0684111
2  0.2127095 0.2901192 0.3367369 0.4356396 1.7479348 0.1403525 2.2378151
3  0.1554675 0.2234804 0.3365653 0.5172071 0.8384712 0.1324767 1.0016602
4  0.6329466 0.8086958 1.1319287 1.3536867 0.5512469 0.2070493 0.2986748
5  1.5823526 0.1472093 0.5358167 0.6838114 1.8703939 0.1177222 2.7812300
6  0.1941248 0.0417533 0.5185655 2.2918398 2.1513495 0.2257744 1.5307155
7  0.4423564 0.2865438 0.5054426 1.1255010 0.1740590 0.1878523 1.0917209
8  0.1564134 0.2365468 0.3239223 1.9040470 1.5788118 0.1488845 2.6273956
9  0.1181918 0.2021875 0.4365611 2.2294233 1.6309892 0.2348087 2.9887999
10 0.2433276 0.1260104 0.3158472 0.7181301 2.3757333 0.2735456 3.3975062
11 0.1606230 0.2345162 0.3987706 2.2816653 1.7783434 0.2817596 3.1220107
12 0.1410920 0.1660120 0.4234817 1.0695182 1.7731206 0.2767200 3.4696200
13 0.2193442 0.3742922 0.5167282 2.5718895 0.5756401 2.0515333 3.6035721
14 0.3466539 0.2919101 0.7731310 2.8063775 1.0461413 0.2932870 3.4441962
15 0.4375035 0.5345708 0.6744828 2.7160012 1.0262325 0.2083201 3.2919337
        CD25       ITAGB7       CRTH2       NKp46     CD127    FSC.A    FSC.H
1  0.6439887  0.623462141 -0.07449993 -0.10346240 1.9387236 71242.25 43965.70
2  0.6330738  0.520151794 -0.04453720 -0.17500883 0.4855043 75832.98 47669.96
3  0.2266883  0.398961663 -0.05785664 -0.07085259 0.8520310 74653.98 46759.62
4  0.3406622  0.138838574 -0.06631734 -0.17470515 0.2055200 80909.42 50394.22
5  0.5917305  0.077761166 -0.07020644  0.10511041 1.9246757 65075.69 39761.16
6  0.3962337  0.009937556 -0.04532208  2.47066903 0.3197118 72070.27 43258.59
7  0.2819901 -0.014998766  0.23423104  0.03243181 0.1910981 73608.70 47642.46
8  0.3167695  0.829722762 -0.06298992 -0.03611803 1.4012212 75078.11 47320.37
9  0.6537131  1.722787678 -0.07341683 -0.02450578 1.8027968 79589.04 50587.21
10 0.7737251  0.695991188 -0.07791403 -0.10694155 2.0584894 70635.70 43593.60
11 0.9193848  1.687331736 -0.11279920 -0.01808203 1.6536188 78168.36 48935.48
12 1.0216262  1.758917928 -0.11142208 -0.04966688 2.3814716 70632.95 42753.42
13 0.9266705  2.040173292 -0.06207005  0.23646396 0.5336345 82701.09 53878.35
14 0.6812732  2.577180028 -0.09271222  2.87342191 0.2985345 84305.39 53874.07
15 0.6472251  2.477188349 -0.09255376  2.92880952 0.1796841 80065.25 50821.62
      FSC.W    SSC.A    SSC.H    SSC.W    Time
1  244059.2 18685.87 15139.36 171783.2 2079785
2  244993.4 16685.34 13484.70 168152.5 2000531
3  244719.7 13986.63 11490.51 162274.8 1935460
4  253807.2 44873.54 32974.96 199168.0 1988384
5  241489.4 20407.00 16644.30 175056.4 2221552
6  244649.3 26768.72 21487.72 181203.0 1910565
7  247586.6 61283.54 48365.29 201206.0 2042293
8  244641.4 17166.27 14022.40 169141.0 2092554
9  246403.3 20116.76 16281.69 173157.9 1984503
10 244163.6 21168.63 17050.64 175143.1 2017840
11 247032.2 20402.41 16691.55 173394.4 2139599
12 243986.8 21179.40 17106.62 175203.0 2109206
13 246364.4 21614.27 17728.73 176139.2 1938876
14 248790.6 29310.82 23717.70 183269.9 2189056
15 247282.0 29583.98 24032.16 183313.5 1794975</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="co"># Plot CD4 expression values on clusters</span></span>
<span id="cb47-2"><a href="#cb47-2" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb47-3"><a href="#cb47-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD56, <span class="at">y =</span> flowSOM, <span class="at">fill =</span> flowSOM)) <span class="sc">+</span></span>
<span id="cb47-4"><a href="#cb47-4" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>(<span class="at">scale =</span> <span class="fl">1.2</span>, <span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb47-5"><a href="#cb47-5" tabindex="-1"></a>  <span class="fu">theme_ridges</span>() <span class="sc">+</span></span>
<span id="cb47-6"><a href="#cb47-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)<span class="sc">+</span></span>
<span id="cb47-7"><a href="#cb47-7" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">&quot;&quot;</span>, <span class="at">x=</span><span class="st">&quot;&quot;</span>, <span class="at">title =</span> <span class="st">&quot;CD56&quot;</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-25-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="fusesom" class="section level3">
<h3>FuseSOM</h3>
<p><a
href="https://bioconductor.org/packages/release/bioc/html/FlowSOM.html">Read
more</a></p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a>risomMarkers <span class="ot">&lt;-</span> <span class="fu">colnames</span>(ILC_transf)[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">14</span>,<span class="dv">16</span><span class="sc">:</span><span class="dv">25</span>)]</span>
<span id="cb48-2"><a href="#cb48-2" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">runFuseSOM</span>(<span class="at">data =</span> <span class="fu">as.matrix</span>(ILC_transf[,<span class="dv">1</span><span class="sc">:</span><span class="dv">32</span>]), <span class="at">markers =</span> risomMarkers, </span>
<span id="cb48-3"><a href="#cb48-3" tabindex="-1"></a>                        <span class="at">numClusters =</span> <span class="dv">15</span>)</span>
<span id="cb48-4"><a href="#cb48-4" tabindex="-1"></a></span>
<span id="cb48-5"><a href="#cb48-5" tabindex="-1"></a>ILC_transf<span class="sc">$</span>fuseSOM<span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">str_split</span>(clusters<span class="sc">$</span>clusters, <span class="st">&quot;_&quot;</span>), <span class="cf">function</span>(x) x[[<span class="dv">2</span>]])))</span>
<span id="cb48-6"><a href="#cb48-6" tabindex="-1"></a></span>
<span id="cb48-7"><a href="#cb48-7" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb48-8"><a href="#cb48-8" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> CD39, <span class="at">y =</span> <span class="fu">as.factor</span>(fuseSOM), <span class="at">fill=</span><span class="fu">as.factor</span>(fuseSOM))) <span class="sc">+</span></span>
<span id="cb48-9"><a href="#cb48-9" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>(<span class="at">scale =</span> <span class="fl">1.2</span>, <span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb48-10"><a href="#cb48-10" tabindex="-1"></a>  <span class="fu">theme_ridges</span>() <span class="sc">+</span></span>
<span id="cb48-11"><a href="#cb48-11" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)<span class="sc">+</span></span>
<span id="cb48-12"><a href="#cb48-12" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">&quot;&quot;</span>, <span class="at">x=</span><span class="st">&quot;&quot;</span>, <span class="at">title =</span> <span class="st">&quot;CD39&quot;</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="dimension-reduction" class="section level2 tabset tabset-fade">
<h2 class="tabset tabset-fade">Dimension Reduction</h2>
<p>We’ll generate a PCA, a tSNE and a UMAP. I’ll then run through all of
this in the Spectre packge.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">as.matrix</span>(ILC_transf[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">14</span>,<span class="dv">16</span><span class="sc">:</span><span class="dv">25</span>)]), <span class="at">center =</span> <span class="cn">TRUE</span>)</span>
<span id="cb49-2"><a href="#cb49-2" tabindex="-1"></a></span>
<span id="cb49-3"><a href="#cb49-3" tabindex="-1"></a>ILC_transf <span class="ot">&lt;-</span> ILC_transf <span class="sc">%&gt;%</span> </span>
<span id="cb49-4"><a href="#cb49-4" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">PC1 =</span> pca<span class="sc">$</span>x[,<span class="dv">1</span>],</span>
<span id="cb49-5"><a href="#cb49-5" tabindex="-1"></a>         <span class="at">PC2 =</span> pca<span class="sc">$</span>x[,<span class="dv">2</span>])</span>
<span id="cb49-6"><a href="#cb49-6" tabindex="-1"></a></span>
<span id="cb49-7"><a href="#cb49-7" tabindex="-1"></a><span class="co"># can take a while</span></span>
<span id="cb49-8"><a href="#cb49-8" tabindex="-1"></a>tsne <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(<span class="fu">as.matrix</span>(ILC_transf[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">14</span>,<span class="dv">16</span><span class="sc">:</span><span class="dv">25</span>)]), <span class="at">dims =</span> <span class="dv">2</span>, <span class="at">perplexity =</span> <span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb49-9"><a href="#cb49-9" tabindex="-1"></a></span>
<span id="cb49-10"><a href="#cb49-10" tabindex="-1"></a>ILC_transf <span class="ot">&lt;-</span> ILC_transf <span class="sc">%&gt;%</span> </span>
<span id="cb49-11"><a href="#cb49-11" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">tSNE1 =</span> tsne<span class="sc">$</span>Y[,<span class="dv">1</span>],</span>
<span id="cb49-12"><a href="#cb49-12" tabindex="-1"></a>         <span class="at">tSNE2 =</span> tsne<span class="sc">$</span>Y[,<span class="dv">2</span>])</span>
<span id="cb49-13"><a href="#cb49-13" tabindex="-1"></a></span>
<span id="cb49-14"><a href="#cb49-14" tabindex="-1"></a>umap <span class="ot">&lt;-</span> <span class="fu">umap</span>(ILC_transf[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">14</span>,<span class="dv">16</span><span class="sc">:</span><span class="dv">25</span>)])</span>
<span id="cb49-15"><a href="#cb49-15" tabindex="-1"></a>ILC_transf <span class="ot">&lt;-</span> ILC_transf <span class="sc">%&gt;%</span> </span>
<span id="cb49-16"><a href="#cb49-16" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">UMAP1 =</span> umap<span class="sc">$</span>layout[,<span class="dv">1</span>],</span>
<span id="cb49-17"><a href="#cb49-17" tabindex="-1"></a>         <span class="at">UMAP2 =</span> umap<span class="sc">$</span>layout[,<span class="dv">2</span>])</span>
<span id="cb49-18"><a href="#cb49-18" tabindex="-1"></a></span>
<span id="cb49-19"><a href="#cb49-19" tabindex="-1"></a><span class="fu">rm</span>(pca,umap,tsne)</span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="fu">saveRDS</span>(ILC_transf,<span class="st">&quot;data/clustered_data.rds&quot;</span>)</span></code></pre></div>
<div id="visualisations-1" class="section level3">
<h3>Visualisations</h3>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a>ILC_transf <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;data/clustered_data.rds&quot;</span>)</span>
<span id="cb51-2"><a href="#cb51-2" tabindex="-1"></a><span class="fu">ggplot</span>(ILC_transf, <span class="fu">aes</span>(PC1,PC2, <span class="at">color=</span> <span class="fu">as.factor</span>(flowSOM))) <span class="sc">+</span></span>
<span id="cb51-3"><a href="#cb51-3" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a><span class="fu">ggplot</span>(ILC_transf, <span class="fu">aes</span>(tSNE1,tSNE2, <span class="at">color=</span> <span class="fu">as.factor</span>(flowSOM))) <span class="sc">+</span></span>
<span id="cb52-2"><a href="#cb52-2" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-29-2.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a><span class="fu">ggplot</span>(ILC_transf, <span class="fu">aes</span>(UMAP1,UMAP2, <span class="at">color=</span> <span class="fu">as.factor</span>(flowSOM))) <span class="sc">+</span></span>
<span id="cb53-2"><a href="#cb53-2" tabindex="-1"></a>  <span class="fu">geom_point</span>()<span class="sc">+</span><span class="fu">xlim</span>(<span class="sc">-</span><span class="dv">7</span>,<span class="dv">10</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-29-3.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a><span class="co"># Colour overlayed by the expression of one marker.</span></span>
<span id="cb54-2"><a href="#cb54-2" tabindex="-1"></a><span class="fu">ggplot</span>(ILC_transf, <span class="fu">aes</span>(UMAP1,UMAP2, <span class="at">color=</span> CD3)) <span class="sc">+</span></span>
<span id="cb54-3"><a href="#cb54-3" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb54-4"><a href="#cb54-4" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>()<span class="sc">+</span><span class="fu">xlim</span>(<span class="sc">-</span><span class="dv">7</span>,<span class="dv">10</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-29-4.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" tabindex="-1"></a><span class="fu">ggplot</span>(ILC_transf, <span class="fu">aes</span>(UMAP1,UMAP2, <span class="at">color=</span> CD16)) <span class="sc">+</span></span>
<span id="cb55-2"><a href="#cb55-2" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb55-3"><a href="#cb55-3" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>()<span class="sc">+</span><span class="fu">xlim</span>(<span class="sc">-</span><span class="dv">7</span>,<span class="dv">10</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-29-5.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="spectre" class="section level3">
<h3>Spectre</h3>
<p><a href="https://immunedynamics.io/spectre/">Read more</a></p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a><span class="co"># Spectre has easy plotting tools to visualize the results of clustering and dimension reduction.</span></span>
<span id="cb56-2"><a href="#cb56-2" tabindex="-1"></a><span class="fu">make.colour.plot</span>(ILC_transf, </span>
<span id="cb56-3"><a href="#cb56-3" tabindex="-1"></a>                 <span class="st">&quot;tSNE1&quot;</span>, <span class="st">&quot;tSNE2&quot;</span>, <span class="st">&quot;subset&quot;</span>,  </span>
<span id="cb56-4"><a href="#cb56-4" tabindex="-1"></a>                 <span class="at">add.label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb56-5"><a href="#cb56-5" tabindex="-1"></a>                 <span class="at">blank.axis =</span> T,</span>
<span id="cb56-6"><a href="#cb56-6" tabindex="-1"></a>                 <span class="at">path =</span> <span class="st">&quot;plots&quot;</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-30-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" tabindex="-1"></a><span class="fu">make.colour.plot</span>(ILC_transf, </span>
<span id="cb57-2"><a href="#cb57-2" tabindex="-1"></a>                 <span class="st">&quot;tSNE1&quot;</span>, <span class="st">&quot;tSNE2&quot;</span>, <span class="st">&quot;flowSOM&quot;</span>, </span>
<span id="cb57-3"><a href="#cb57-3" tabindex="-1"></a>                 <span class="at">col.type =</span> <span class="st">&#39;factor&#39;</span>, </span>
<span id="cb57-4"><a href="#cb57-4" tabindex="-1"></a>                 <span class="at">add.label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb57-5"><a href="#cb57-5" tabindex="-1"></a>                 <span class="at">blank.axis =</span> T,</span>
<span id="cb57-6"><a href="#cb57-6" tabindex="-1"></a>                 <span class="at">path =</span> <span class="st">&quot;plots&quot;</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-30-2.png" width="672" style="display: block; margin: auto;" /></p>
<details>
<summary>
<strong>Multi-plots:</strong>
</summary>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a><span class="fu">make.multi.plot</span>(ILC_transf, </span>
<span id="cb58-2"><a href="#cb58-2" tabindex="-1"></a>                <span class="st">&quot;tSNE1&quot;</span>, <span class="st">&quot;tSNE2&quot;</span>, </span>
<span id="cb58-3"><a href="#cb58-3" tabindex="-1"></a>                risomMarkers, </span>
<span id="cb58-4"><a href="#cb58-4" tabindex="-1"></a>                <span class="at">blank.axis =</span> T, </span>
<span id="cb58-5"><a href="#cb58-5" tabindex="-1"></a>                <span class="at">colours =</span> <span class="st">&quot;viridis&quot;</span>, </span>
<span id="cb58-6"><a href="#cb58-6" tabindex="-1"></a>                <span class="at">path =</span> <span class="st">&quot;plots&quot;</span>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-2.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-3.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-4.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-5.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-6.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-7.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-8.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-9.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-10.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-11.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-12.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-13.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-14.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-15.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-16.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-17.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-18.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-19.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-20.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-21.png" width="672" style="display: block; margin: auto;" /><img src="figure/202504_Workshop.Rmd/unnamed-chunk-31-22.png" width="672" style="display: block; margin: auto;" /></p>
</details>
<p>And we can make the heatmap you’re used to seeing.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" tabindex="-1"></a>exp <span class="ot">&lt;-</span> <span class="fu">do.aggregate</span>(<span class="fu">as.data.table</span>(ILC_transf), <span class="at">use.cols =</span> risomMarkers, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;flowSOM&quot;</span>))</span>
<span id="cb59-2"><a href="#cb59-2" tabindex="-1"></a><span class="fu">make.pheatmap</span>(exp, <span class="at">file.name =</span> <span class="st">&quot;plots/Pheatmap_FlowSOM.png&quot;</span>, <span class="at">sample.col=</span><span class="st">&quot;flowSOM&quot;</span>, <span class="at">plot.cols=</span>risomMarkers)</span></code></pre></div>
<p><em>Spectre has also embedded FlowSOM clustering and dimension
reduction tools such as umap in easy codes, such as:<br></em></p>
<ul>
<li><code>do.subsample()</code></li>
<li><code>run.flowsom()</code></li>
<li><code>run.umap()</code></li>
</ul>
</div>
</div>
</div>
<div id="additional-considerations"
class="section level1 tabset tabset-fade">
<h1 class="tabset tabset-fade">Additional Considerations</h1>
<div id="qc" class="section level2">
<h2>QC</h2>
<p>When looking at the raw data, or in the middle of data processing,
you may consider filtering some cell. In the 1-marker plots below, you
can see that there are a few extreme values that are displayed and
pushing the majority of values squeezed together. e.g. HLA-DR:</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a>ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb60-2"><a href="#cb60-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> HLA.DR))<span class="sc">+</span></span>
<span id="cb60-3"><a href="#cb60-3" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">linewidth =</span> <span class="fl">0.7</span>, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="fu">aes</span>(<span class="at">fill =</span> <span class="st">&quot;red&quot;</span>), <span class="at">alpha =</span> <span class="fl">0.5</span>,<span class="at">show.legend =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb60-4"><a href="#cb60-4" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>)<span class="sc">+</span></span>
<span id="cb60-5"><a href="#cb60-5" tabindex="-1"></a></span>
<span id="cb60-6"><a href="#cb60-6" tabindex="-1"></a><span class="co"># Visualise the values at the top 0.1% and bottom 0.1%.</span></span>
<span id="cb60-7"><a href="#cb60-7" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">quantile</span>(ILC_transf<span class="sc">$</span>HLA.DR, <span class="fl">0.999</span>), <span class="at">colour =</span> <span class="st">&quot;blue&quot;</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>)<span class="sc">+</span></span>
<span id="cb60-8"><a href="#cb60-8" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">quantile</span>(ILC_transf<span class="sc">$</span>HLA.DR, <span class="fl">0.001</span>), <span class="at">colour =</span> <span class="st">&quot;blue&quot;</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>)<span class="sc">+</span></span>
<span id="cb60-9"><a href="#cb60-9" tabindex="-1"></a></span>
<span id="cb60-10"><a href="#cb60-10" tabindex="-1"></a><span class="co"># We can label the threshold lines: annotate as a text, at the position of the top and bottom 0.01%, and slightly adjust the label position horizontally.</span></span>
<span id="cb60-11"><a href="#cb60-11" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="at">x =</span> <span class="fu">quantile</span>(ILC_transf<span class="sc">$</span>HLA.DR, <span class="fl">0.999</span>), <span class="at">y =</span> <span class="fl">0.003</span>, <span class="at">label =</span> <span class="st">&quot;Top and bottom 0.1%&quot;</span>, <span class="at">hjust =</span> <span class="sc">-</span><span class="fl">0.1</span>) </span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-33-1.png" width="672" style="display: block; margin: auto;" /></p>
<details>
<summary>
View them for all channels.
</summary>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" tabindex="-1"></a><span class="co"># Create the empty list</span></span>
<span id="cb61-2"><a href="#cb61-2" tabindex="-1"></a>plots_line <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb61-3"><a href="#cb61-3" tabindex="-1"></a></span>
<span id="cb61-4"><a href="#cb61-4" tabindex="-1"></a><span class="co"># Create the loop to generate plots.</span></span>
<span id="cb61-5"><a href="#cb61-5" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>){</span>
<span id="cb61-6"><a href="#cb61-6" tabindex="-1"></a>  </span>
<span id="cb61-7"><a href="#cb61-7" tabindex="-1"></a>  <span class="co"># Get the marker names</span></span>
<span id="cb61-8"><a href="#cb61-8" tabindex="-1"></a>  Marker <span class="ot">&lt;-</span> <span class="fu">colnames</span>(ILC_transf)[i]</span>
<span id="cb61-9"><a href="#cb61-9" tabindex="-1"></a>  </span>
<span id="cb61-10"><a href="#cb61-10" tabindex="-1"></a>  <span class="co"># Generate the plots and put it into the list</span></span>
<span id="cb61-11"><a href="#cb61-11" tabindex="-1"></a>  p <span class="ot">&lt;-</span> ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb61-12"><a href="#cb61-12" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> .data[[Marker]]))<span class="sc">+</span> </span>
<span id="cb61-13"><a href="#cb61-13" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">linewidth =</span> <span class="fl">0.7</span>, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;red&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>,<span class="at">show.legend =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb61-14"><a href="#cb61-14" tabindex="-1"></a>   <span class="fu">xlab</span>(Marker)<span class="sc">+</span></span>
<span id="cb61-15"><a href="#cb61-15" tabindex="-1"></a>   <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>)<span class="sc">+</span></span>
<span id="cb61-16"><a href="#cb61-16" tabindex="-1"></a>  </span>
<span id="cb61-17"><a href="#cb61-17" tabindex="-1"></a>    <span class="co"># Add the newly written lines to visualise the outliers to this loop.</span></span>
<span id="cb61-18"><a href="#cb61-18" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">quantile</span>(ILC_transf[[i]], <span class="fl">0.999</span>), <span class="at">colour =</span> <span class="st">&quot;blue&quot;</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>)<span class="sc">+</span></span>
<span id="cb61-19"><a href="#cb61-19" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">quantile</span>(ILC_transf[[i]], <span class="fl">0.001</span>), <span class="at">colour  =</span> <span class="st">&quot;blue&quot;</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>)</span>
<span id="cb61-20"><a href="#cb61-20" tabindex="-1"></a>    </span>
<span id="cb61-21"><a href="#cb61-21" tabindex="-1"></a>  <span class="co"># Put the plots into the list</span></span>
<span id="cb61-22"><a href="#cb61-22" tabindex="-1"></a>  plots_line[[<span class="fu">length</span>(plots_line)<span class="sc">+</span><span class="dv">1</span>]] <span class="ot">&lt;-</span> p</span>
<span id="cb61-23"><a href="#cb61-23" tabindex="-1"></a>}</span>
<span id="cb61-24"><a href="#cb61-24" tabindex="-1"></a></span>
<span id="cb61-25"><a href="#cb61-25" tabindex="-1"></a><span class="co"># Use a grid layout to view the plots</span></span>
<span id="cb61-26"><a href="#cb61-26" tabindex="-1"></a><span class="fu">library</span>(ggpubr)</span>
<span id="cb61-27"><a href="#cb61-27" tabindex="-1"></a>ggpubr<span class="sc">::</span><span class="fu">ggarrange</span>(<span class="at">plotlist =</span> plots_line, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>$`1`</code></pre>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-34-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>
$`2`</code></pre>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-34-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>
$`3`</code></pre>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-34-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>
$`4`</code></pre>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-34-4.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>
attr(,&quot;class&quot;)
[1] &quot;list&quot;      &quot;ggarrange&quot;</code></pre>
</details>
<p><br> To look at this in a more statistical way, you can create a
quick data summary of the minimum, maximum, mean, median and the
interquartile range.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" tabindex="-1"></a><span class="fu">summary</span>(ILC_transf)</span></code></pre></div>
<pre><code>     NKp44             Lineage              CD45            CD103        
 Min.   :-0.60944   Min.   :-0.74961   Min.   :0.8776   Min.   :-1.6764  
 1st Qu.:-0.02043   1st Qu.:-0.01364   1st Qu.:3.3551   1st Qu.: 0.2275  
 Median : 0.10043   Median : 0.22802   Median :3.6405   Median : 0.7676  
 Mean   : 0.12175   Mean   : 0.23908   Mean   :3.5623   Mean   : 0.8455  
 3rd Qu.: 0.23539   3rd Qu.: 0.47307   3rd Qu.:3.8650   3rd Qu.: 1.4379  
 Max.   : 2.56281   Max.   : 1.51645   Max.   :4.8584   Max.   : 3.8903  
                                                                         
     FVS700              CD4              HLA.DR             CD16        
 Min.   :-1.37739   Min.   :-1.0414   Min.   :-0.5972   Min.   :-0.2901  
 1st Qu.:-0.26103   1st Qu.: 0.7355   1st Qu.: 0.1318   1st Qu.: 0.2128  
 Median :-0.08059   Median : 1.5659   Median : 0.2741   Median : 0.3146  
 Mean   :-0.08100   Mean   : 1.2394   Mean   : 0.3409   Mean   : 0.3452  
 3rd Qu.: 0.09863   3rd Qu.: 1.8856   3rd Qu.: 0.4526   3rd Qu.: 0.4264  
 Max.   : 0.70678   Max.   : 3.0169   Max.   : 5.6978   Max.   : 4.4567  
                                                                         
      CD3              CD94               CD39               CD56        
 Min.   :-0.712   Min.   :-1.40325   Min.   :-0.98439   Min.   :-1.6483  
 1st Qu.: 2.812   1st Qu.:-0.21409   1st Qu.:-0.01198   1st Qu.: 0.5906  
 Median : 3.145   Median : 0.03539   Median : 0.24073   Median : 1.0224  
 Mean   : 2.968   Mean   : 0.06666   Mean   : 0.40428   Mean   : 1.2305  
 3rd Qu.: 3.447   3rd Qu.: 0.29662   3rd Qu.: 0.53662   3rd Qu.: 1.6042  
 Max.   : 4.601   Max.   : 3.19569   Max.   : 5.03303   Max.   : 5.8967  
                                                                         
    TCRVA72             CKIT            CD14.CD19            CD49D        
 Min.   :-0.9207   Min.   :-0.78687   Min.   :-0.62200   Min.   :-0.4446  
 1st Qu.: 0.3260   1st Qu.:-0.01042   1st Qu.: 0.02254   1st Qu.: 0.2099  
 Median : 0.5569   Median : 0.17073   Median : 0.18560   Median : 0.3337  
 Mean   : 0.5825   Mean   : 0.21415   Mean   : 0.21396   Mean   : 0.3559  
 3rd Qu.: 0.7734   3rd Qu.: 0.37904   3rd Qu.: 0.36621   3rd Qu.: 0.4694  
 Max.   : 4.7995   Max.   : 3.83387   Max.   : 1.25671   Max.   : 2.8257  
                                                                          
     CD244             CD161             TCRgt                CD69        
 Min.   :-0.6901   Min.   :-0.8723   Min.   :-0.936193   Min.   :-0.9441  
 1st Qu.: 0.4753   1st Qu.: 1.3699   1st Qu.: 0.006448   1st Qu.: 2.5221  
 Median : 0.8006   Median : 2.0976   Median : 0.210441   Median : 3.0271  
 Mean   : 1.0140   Mean   : 1.9402   Mean   : 0.226867   Mean   : 2.8921  
 3rd Qu.: 1.4233   3rd Qu.: 2.6183   3rd Qu.: 0.422976   3rd Qu.: 3.4394  
 Max.   : 3.6121   Max.   : 4.5151   Max.   : 3.696751   Max.   : 5.0315  
                                                                          
      CD25              ITAGB7            CRTH2              NKp46         
 Min.   :-2.04859   Min.   :-1.1604   Min.   :-0.81326   Min.   :-1.02354  
 1st Qu.: 0.08245   1st Qu.: 0.3493   1st Qu.:-0.17971   1st Qu.:-0.21610  
 Median : 0.62505   Median : 0.7189   Median :-0.07405   Median :-0.07461  
 Mean   : 0.55484   Mean   : 0.8315   Mean   :-0.06856   Mean   : 0.03578  
 3rd Qu.: 1.06689   3rd Qu.: 1.1891   3rd Qu.: 0.03066   3rd Qu.: 0.07985  
 Max.   : 3.55726   Max.   : 4.2459   Max.   : 2.94352   Max.   : 3.86214  
                                                                           
     CD127             FSC.A            FSC.H           FSC.W       
 Min.   :-0.9117   Min.   : 33329   Min.   :15411   Min.   :168801  
 1st Qu.: 1.2209   1st Qu.: 61433   1st Qu.:36705   1st Qu.:238038  
 Median : 1.8260   Median : 72555   Median :45029   Median :244372  
 Mean   : 1.6956   Mean   : 72136   Mean   :45157   Mean   :247282  
 3rd Qu.: 2.2689   3rd Qu.: 82906   3rd Qu.:53280   3rd Qu.:251189  
 Max.   : 3.7882   Max.   :141562   Max.   :95998   Max.   :395580  
                                                                    
     SSC.A           SSC.H           SSC.W             Time        
 Min.   :-2606   Min.   : 3338   Min.   : 90908   Min.   :   1182  
 1st Qu.:15561   1st Qu.:12672   1st Qu.:165542   1st Qu.:1042038  
 Median :19256   Median :15600   Median :172405   Median :2051315  
 Mean   :20342   Mean   :16431   Mean   :174628   Mean   :2077101  
 3rd Qu.:23568   3rd Qu.:18958   3rd Qu.:179231   3rd Qu.:3062559  
 Max.   :97782   Max.   :75316   Max.   :415995   Max.   :4599769  
                                                                   
    subset             flowSOM        fuseSOM            PC1         
 Length:19588       6      :9296   Min.   : 1.000   Min.   :-2.6306  
 Class :character   1      :2493   1st Qu.: 1.000   1st Qu.:-0.9985  
 Mode  :character   2      :1818   Median : 1.000   Median :-0.3773  
                    7      :1517   Mean   : 3.254   Mean   : 0.0000  
                    12     :1212   3rd Qu.: 5.000   3rd Qu.: 0.6658  
                    10     :1000   Max.   :15.000   Max.   : 6.8083  
                    (Other):2252                                     
      PC2               tSNE1              tSNE2              UMAP1         
 Min.   :-4.08455   Min.   :-35.4132   Min.   :-44.7404   Min.   :-35.2457  
 1st Qu.:-0.78765   1st Qu.:-15.6414   1st Qu.:-12.1809   1st Qu.: -2.5886  
 Median :-0.02214   Median : -0.6647   Median :  0.3175   Median : -0.7602  
 Mean   : 0.00000   Mean   :  0.0000   Mean   :  0.0000   Mean   :  0.0000  
 3rd Qu.: 0.79000   3rd Qu.: 12.6072   3rd Qu.: 12.2346   3rd Qu.:  2.1378  
 Max.   : 4.90431   Max.   : 43.0089   Max.   : 37.6567   Max.   :  9.1392  
                                                                            
     UMAP2       
 Min.   :-5.908  
 1st Qu.:-3.448  
 Median :-1.536  
 Mean   : 0.000  
 3rd Qu.: 2.129  
 Max.   :14.211  
                 </code></pre>
<p><br></p>
<p>Either being artificial or biological, these might interfere how the
computer interprets your data in the downstream clustering and dimension
reduction. Sometimes you may decide to investigate how these impact your
dataset and where they arise from. But if there are plenty of cells, and
the % outliers is super small, you might just want to filter them
out.<br></p>
<p><strong>Filtering by %</strong> <br> You might want to filter out the
top and bottom 0.1% of cells . Note that this approach removes the cells
that doesn’t fit within the top and bottom 0.1% of ANY channel, and
removes the cell from the entire dataset. So be cautious about the % so
you aren’t removing too many cells when they are added up ( we’ll use
0.1% all the way for demonstration purposes). <br></p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" tabindex="-1"></a><span class="co"># Create an empty list.</span></span>
<span id="cb69-2"><a href="#cb69-2" tabindex="-1"></a>outlier_flag <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb69-3"><a href="#cb69-3" tabindex="-1"></a></span>
<span id="cb69-4"><a href="#cb69-4" tabindex="-1"></a><span class="co"># Loop through the columns with markers</span></span>
<span id="cb69-5"><a href="#cb69-5" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>){</span>
<span id="cb69-6"><a href="#cb69-6" tabindex="-1"></a>  </span>
<span id="cb69-7"><a href="#cb69-7" tabindex="-1"></a>  <span class="co"># Find the cell values</span></span>
<span id="cb69-8"><a href="#cb69-8" tabindex="-1"></a>  value <span class="ot">&lt;-</span> ILC_transf[[i]]</span>
<span id="cb69-9"><a href="#cb69-9" tabindex="-1"></a>  </span>
<span id="cb69-10"><a href="#cb69-10" tabindex="-1"></a>  <span class="co"># Get the marker names</span></span>
<span id="cb69-11"><a href="#cb69-11" tabindex="-1"></a>  Marker <span class="ot">&lt;-</span> <span class="fu">colnames</span>(ILC_transf)[i]</span>
<span id="cb69-12"><a href="#cb69-12" tabindex="-1"></a>  </span>
<span id="cb69-13"><a href="#cb69-13" tabindex="-1"></a>  <span class="co"># Find the thresholds of outliers</span></span>
<span id="cb69-14"><a href="#cb69-14" tabindex="-1"></a>  upper <span class="ot">&lt;-</span> <span class="fu">quantile</span>(value, <span class="fl">0.999</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb69-15"><a href="#cb69-15" tabindex="-1"></a>  lower <span class="ot">&lt;-</span> <span class="fu">quantile</span>(value, <span class="fl">0.001</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb69-16"><a href="#cb69-16" tabindex="-1"></a></span>
<span id="cb69-17"><a href="#cb69-17" tabindex="-1"></a>  <span class="co"># Flag the outliers as 1, else as 0. And record them into the list.</span></span>
<span id="cb69-18"><a href="#cb69-18" tabindex="-1"></a>  flag <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(value <span class="sc">&lt;</span> lower <span class="sc">|</span> value <span class="sc">&gt;</span> upper, <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb69-19"><a href="#cb69-19" tabindex="-1"></a>  outlier_flag[[<span class="fu">length</span>(outlier_flag)<span class="sc">+</span><span class="dv">1</span>]] <span class="ot">&lt;-</span> flag</span>
<span id="cb69-20"><a href="#cb69-20" tabindex="-1"></a>}</span>
<span id="cb69-21"><a href="#cb69-21" tabindex="-1"></a><span class="co"># Turn the flags into a data table object.</span></span>
<span id="cb69-22"><a href="#cb69-22" tabindex="-1"></a>outlier_flag <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(outlier_flag)</span>
<span id="cb69-23"><a href="#cb69-23" tabindex="-1"></a><span class="fu">colnames</span>(outlier_flag)<span class="ot">&lt;-</span> <span class="fu">colnames</span>(ILC_transf)[<span class="fu">c</span>(<span class="dv">7</span><span class="sc">:</span><span class="dv">31</span>)]</span>
<span id="cb69-24"><a href="#cb69-24" tabindex="-1"></a>subset <span class="ot">&lt;-</span> ILC_transf[<span class="fu">rowSums</span>(outlier_flag)<span class="sc">==</span><span class="dv">0</span>,]</span>
<span id="cb69-25"><a href="#cb69-25" tabindex="-1"></a>  </span>
<span id="cb69-26"><a href="#cb69-26" tabindex="-1"></a><span class="co"># Calculate the % of cell loss after filtering, then format it as being displayed in less digits, then paste a sentence together.</span></span>
<span id="cb69-27"><a href="#cb69-27" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">&quot;Cell loss: &quot;</span>,<span class="fu">format</span>((<span class="dv">1</span><span class="sc">-</span><span class="fu">nrow</span>(subset)<span class="sc">/</span><span class="fu">nrow</span>(ILC_transf))<span class="sc">*</span><span class="dv">100</span>, <span class="at">digits =</span> <span class="dv">3</span>), <span class="st">&quot;%&quot;</span>)</span></code></pre></div>
<pre><code>[1] &quot;Cell loss: 4.44%&quot;</code></pre>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" tabindex="-1"></a><span class="co"># View the pre- and post-filtered data side by side, putting them in a grid.</span></span>
<span id="cb71-2"><a href="#cb71-2" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb71-3"><a href="#cb71-3" tabindex="-1"></a>  ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb71-4"><a href="#cb71-4" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> HLA.DR))<span class="sc">+</span></span>
<span id="cb71-5"><a href="#cb71-5" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">linewidth =</span> <span class="fl">0.7</span>, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;red&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb71-6"><a href="#cb71-6" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>),</span>
<span id="cb71-7"><a href="#cb71-7" tabindex="-1"></a></span>
<span id="cb71-8"><a href="#cb71-8" tabindex="-1"></a>subset <span class="sc">%&gt;%</span></span>
<span id="cb71-9"><a href="#cb71-9" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> HLA.DR))<span class="sc">+</span></span>
<span id="cb71-10"><a href="#cb71-10" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">linewidth =</span> <span class="fl">0.7</span>, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;red&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb71-11"><a href="#cb71-11" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>),</span>
<span id="cb71-12"><a href="#cb71-12" tabindex="-1"></a></span>
<span id="cb71-13"><a href="#cb71-13" tabindex="-1"></a> <span class="co"># number of column equals 1.</span></span>
<span id="cb71-14"><a href="#cb71-14" tabindex="-1"></a> <span class="at">ncol =</span> <span class="dv">1</span></span>
<span id="cb71-15"><a href="#cb71-15" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-37-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><strong>Filtering by fixed thresholds</strong> <br> You might want to
filter out the data that is outside of certain values per channel. A
commonly used standard in data cleaning is calculated based on
interquartile ranges (IQR). The thresholds are any data points that are
1.5 times of IQR above the 3rd quartile, or 1.5 times of IQR below the
1st quartile. This will keep the majority of data points, however, it
might also remove some true variance in the dataset at extreme
ends.<br></p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" tabindex="-1"></a><span class="co"># Take HLA-DR as an example</span></span>
<span id="cb72-2"><a href="#cb72-2" tabindex="-1"></a></span>
<span id="cb72-3"><a href="#cb72-3" tabindex="-1"></a><span class="co"># Calculate the interquartile range.</span></span>
<span id="cb72-4"><a href="#cb72-4" tabindex="-1"></a>   IQR <span class="ot">&lt;-</span> <span class="fu">IQR</span>(ILC_transf<span class="sc">$</span>HLA.DR)</span>
<span id="cb72-5"><a href="#cb72-5" tabindex="-1"></a>   Q1 <span class="ot">&lt;-</span> <span class="fu">quantile</span>(ILC_transf<span class="sc">$</span>HLA.DR, <span class="fl">0.25</span>)</span>
<span id="cb72-6"><a href="#cb72-6" tabindex="-1"></a>   Q3 <span class="ot">&lt;-</span> <span class="fu">quantile</span>(ILC_transf<span class="sc">$</span>HLA.DR, <span class="fl">0.75</span>)</span>
<span id="cb72-7"><a href="#cb72-7" tabindex="-1"></a>   </span>
<span id="cb72-8"><a href="#cb72-8" tabindex="-1"></a><span class="co"># Calculate the upper and lower thresholds</span></span>
<span id="cb72-9"><a href="#cb72-9" tabindex="-1"></a>   lowerfence <span class="ot">&lt;-</span> Q1 <span class="sc">-</span> <span class="fl">1.5</span><span class="sc">*</span>IQR</span>
<span id="cb72-10"><a href="#cb72-10" tabindex="-1"></a>   upperfence <span class="ot">&lt;-</span> Q3 <span class="sc">+</span> <span class="fl">1.5</span><span class="sc">*</span>IQR </span>
<span id="cb72-11"><a href="#cb72-11" tabindex="-1"></a> </span>
<span id="cb72-12"><a href="#cb72-12" tabindex="-1"></a><span class="co"># Only keep the cells which CD127 channel value within the thresholds.   </span></span>
<span id="cb72-13"><a href="#cb72-13" tabindex="-1"></a>filt <span class="ot">&lt;-</span> ILC_transf[ILC_transf<span class="sc">$</span>HLA.DR <span class="sc">&gt;</span> lowerfence <span class="sc">&amp;</span> ILC_transf<span class="sc">$</span>HLA.DR <span class="sc">&lt;</span> upperfence, ]</span>
<span id="cb72-14"><a href="#cb72-14" tabindex="-1"></a></span>
<span id="cb72-15"><a href="#cb72-15" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb72-16"><a href="#cb72-16" tabindex="-1"></a>  ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb72-17"><a href="#cb72-17" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> HLA.DR))<span class="sc">+</span></span>
<span id="cb72-18"><a href="#cb72-18" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">linewidth =</span> <span class="fl">0.7</span>, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;red&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>,<span class="at">show.legend =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb72-19"><a href="#cb72-19" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>),</span>
<span id="cb72-20"><a href="#cb72-20" tabindex="-1"></a>filt <span class="sc">%&gt;%</span></span>
<span id="cb72-21"><a href="#cb72-21" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> HLA.DR))<span class="sc">+</span></span>
<span id="cb72-22"><a href="#cb72-22" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">linewidth =</span> <span class="fl">0.7</span>, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;red&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>,<span class="at">show.legend =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb72-23"><a href="#cb72-23" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>),</span>
<span id="cb72-24"><a href="#cb72-24" tabindex="-1"></a></span>
<span id="cb72-25"><a href="#cb72-25" tabindex="-1"></a> <span class="co"># number of column equals 1.</span></span>
<span id="cb72-26"><a href="#cb72-26" tabindex="-1"></a> <span class="at">ncol =</span> <span class="dv">1</span></span>
<span id="cb72-27"><a href="#cb72-27" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-38-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" tabindex="-1"></a><span class="fu">rm</span>(Q1,Q3,IQR)</span></code></pre></div>
<p>Or, simply just choose an arbitrary threshold.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" tabindex="-1"></a><span class="co"># Only keep the cells which transformed HLA-DR expression value &lt; 2.</span></span>
<span id="cb74-2"><a href="#cb74-2" tabindex="-1"></a>manual <span class="ot">&lt;-</span> ILC_transf[ILC_transf<span class="sc">$</span>HLA.DR <span class="sc">&lt;</span> <span class="dv">2</span>, ]</span>
<span id="cb74-3"><a href="#cb74-3" tabindex="-1"></a></span>
<span id="cb74-4"><a href="#cb74-4" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb74-5"><a href="#cb74-5" tabindex="-1"></a>  ILC_transf <span class="sc">%&gt;%</span></span>
<span id="cb74-6"><a href="#cb74-6" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> HLA.DR))<span class="sc">+</span></span>
<span id="cb74-7"><a href="#cb74-7" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">linewidth =</span> <span class="fl">0.7</span>, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;red&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>,<span class="at">show.legend =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb74-8"><a href="#cb74-8" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>),</span>
<span id="cb74-9"><a href="#cb74-9" tabindex="-1"></a>manual <span class="sc">%&gt;%</span></span>
<span id="cb74-10"><a href="#cb74-10" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> HLA.DR))<span class="sc">+</span></span>
<span id="cb74-11"><a href="#cb74-11" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">linewidth =</span> <span class="fl">0.7</span>, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;red&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>,<span class="at">show.legend =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb74-12"><a href="#cb74-12" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Density&quot;</span>),</span>
<span id="cb74-13"><a href="#cb74-13" tabindex="-1"></a></span>
<span id="cb74-14"><a href="#cb74-14" tabindex="-1"></a> <span class="co"># number of column equals 1.</span></span>
<span id="cb74-15"><a href="#cb74-15" tabindex="-1"></a> <span class="at">ncol =</span> <span class="dv">1</span></span>
<span id="cb74-16"><a href="#cb74-16" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="figure/202504_Workshop.Rmd/unnamed-chunk-39-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>R version 4.4.0 (2024-04-24)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.3

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Sydney
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] ggpubr_0.6.0         pheatmap_1.0.12      ggpointdensity_0.1.0
 [4] RColorBrewer_1.1-3   ggthemes_5.1.0       colorRamps_2.3.4    
 [7] scales_1.3.0         data.table_1.17.0    Rtsne_0.17          
[10] umap_0.2.10.0        GGally_2.2.1         FlowSOM_2.12.0      
[13] igraph_2.1.4         FuseSOM_1.6.0        Spectre_1.2.0       
[16] cowplot_1.1.3        ggridges_0.5.6       lubridate_1.9.4     
[19] forcats_1.0.0        stringr_1.5.1        dplyr_1.1.4         
[22] purrr_1.0.4          readr_2.1.5          tidyr_1.3.1         
[25] tibble_3.2.1         ggplot2_3.5.1        tidyverse_2.0.0     
[28] workflowr_1.7.1     

loaded via a namespace (and not attached):
  [1] rstudioapi_0.17.1           jsonlite_1.9.1             
  [3] magrittr_2.0.3              DataVisualizations_1.3.3   
  [5] farver_2.1.2                profileModel_0.6.1         
  [7] rmarkdown_2.29              ragg_1.3.3                 
  [9] fs_1.6.5                    zlibbioc_1.50.0            
 [11] vctrs_0.6.5                 askpass_1.2.1              
 [13] rstatix_0.7.2               htmltools_0.5.8.1          
 [15] S4Arrays_1.4.1              broom_1.0.7                
 [17] SparseArray_1.4.8           Formula_1.2-5              
 [19] gridGraphics_0.5-1          sass_0.4.9                 
 [21] bslib_0.9.0                 princurve_2.1.6            
 [23] plyr_1.8.9                  cachem_1.1.0               
 [25] whisker_0.4.1               lifecycle_1.0.4            
 [27] pkgconfig_2.0.3             Matrix_1.7-3               
 [29] R6_2.6.1                    fastmap_1.2.0              
 [31] GenomeInfoDbData_1.2.12     MatrixGenerics_1.16.0      
 [33] digest_0.6.37               ggnewscale_0.5.1           
 [35] colorspace_2.1-1            S4Vectors_0.42.1           
 [37] ps_1.9.0                    rprojroot_2.0.4            
 [39] RSpectra_0.16-2             textshaping_1.0.0          
 [41] GenomicRanges_1.56.2        vegan_2.6-10               
 [43] labeling_0.4.3              cytolib_2.16.0             
 [45] timechange_0.3.0            httr_1.4.7                 
 [47] polyclip_1.10-7             abind_1.4-8                
 [49] coop_0.6-3                  mgcv_1.9-1                 
 [51] compiler_4.4.0              proxy_0.4-27               
 [53] withr_3.0.2                 ConsensusClusterPlus_1.68.0
 [55] backports_1.5.0             carData_3.0-5              
 [57] psych_2.4.12                ggstats_0.9.0              
 [59] ggforce_0.4.2               ggsignif_0.6.4             
 [61] MASS_7.3-65                 openssl_2.3.2              
 [63] DelayedArray_0.30.1         permute_0.9-7              
 [65] tools_4.4.0                 httpuv_1.6.15              
 [67] brglm_0.7.2                 glue_1.8.0                 
 [69] callr_3.7.6                 nlme_3.1-167               
 [71] promises_1.3.2              grid_4.4.0                 
 [73] getPass_0.2-4               cluster_2.1.8.1            
 [75] generics_0.1.3              isoband_0.2.7              
 [77] gtable_0.3.6                tzdb_0.5.0                 
 [79] hms_1.1.3                   utf8_1.2.4                 
 [81] sp_2.2-0                    car_3.1-3                  
 [83] XVector_0.44.0              BiocGenerics_0.50.0        
 [85] pillar_1.10.1               yulab.utils_0.2.0          
 [87] later_1.4.1                 splines_4.4.0              
 [89] flowCore_2.16.0             tweenr_2.0.3               
 [91] lattice_0.22-6              RProtoBufLib_2.16.0        
 [93] tidyselect_1.2.1            knitr_1.50                 
 [95] git2r_0.35.0                gridExtra_2.3              
 [97] IRanges_2.38.1              SummarizedExperiment_1.34.0
 [99] analogue_0.17-7             stats4_4.4.0               
[101] xfun_0.51                   Biobase_2.64.0             
[103] matrixStats_1.5.0           stringi_1.8.4              
[105] UCSC.utils_1.0.0            yaml_2.3.10                
[107] evaluate_1.0.3              ggplotify_0.1.2            
[109] cli_3.6.4                   systemfonts_1.2.1          
[111] reticulate_1.41.0.1         munsell_0.5.1              
[113] processx_3.8.6              jquerylib_0.1.4            
[115] Rcpp_1.0.14                 GenomeInfoDb_1.40.1        
[117] png_0.1-8                   FCPS_1.3.4                 
[119] fastcluster_1.2.6           XML_3.99-0.18              
[121] parallel_4.4.0              mclust_6.1.1               
[123] viridisLite_0.4.2           crayon_1.5.3               
[125] rlang_1.1.6                 mnormt_2.1.1               </code></pre>
</div>
</div>
</div>

<div id="rmd-source-code">---
output:
  workflowr::wflow_html:
    toc: no
    code_folding: "hide"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning=FALSE, error=F)
```

## Flow Cytometry in R: A Hands-On Introduction

This beginner-friendly, hands-on workshop will guide users through a complete flow cytometry analysis workflow in R, from loading data to visualizing results. Through practical exercises, we’ll cover essential steps like generating histograms and scatterplots, extracting summary statistics, subsetting data, and performing dimension reduction and clustering—all while demonstrating the advantages of reproducible reports with RMarkdown. 

Whether you're new to R or just curious about coding your analyses, this session will show you how accessible and powerful R can be for flow cytometry.

**Your Demonstrators:**

- Thomas O'Neil   
- Yuchen Li (Online)

**Author**: Yuchen Li, Thomas O'Neil

[Zoom link](https://uni-sydney.zoom.us/s/88109360455) (recorded)

# Introduction

In this workshop, we aim to give an overview of what is possible for flow analysis in R. We will touch on two workflows:

- data visualisation and data collection

- dimension reduction and clustering

We'll use a 25-colour flow cytometry dataset of innate lymphoid cells (ILC), natural killer cells (NK), and innate-like T cells in human intestinal tissues, after excluding dead cells, doublets, mast cells and basophils, hematopoietic stem cells, myeloid cells and granulocytes, and B cells. 

The original data was published as is available in [OMIP-082](https://pubmed.ncbi.nlm.nih.gov/35018731/) 

# Setup{.tabset .tabset-fade}

We'll take you through steps from the beginning, installing packages and loading data. 

## Preparation: packages and directories

Packages are collections of functions. Many are not in-built in the base R but they do pretty cool things, enabling generating pretty plots, performing data wrangling in an efficient way, etc. R also has a large developer community, with many community-developed packages available on platforms like `Bioconductor` or `GitHub`.

```{r, message=FALSE, warning=FALSE}
# Install packages used in this workshop.

#install.packages("tidyverse")
#install.packages("ggridges")
#install.packages("GGally")
#install.packages("cowplot")
#install.packages("data.table")
#install.packages("BiocManager")

#BiocManager::install("FuseSOM")
#BiocManager::install("FlowSOM")

#if(!require('remotes')) {install.packages('remotes')}
#remotes::install_github(repo = "immunedynamics/spectre")


# Load packages
library(tidyverse)
library(ggridges)
library(cowplot)
library(Spectre)
library(FuseSOM)
library(FlowSOM)
library(GGally)
library(umap)
library(Rtsne)
library(data.table)

theme_set(theme_classic())
```

Create some folders for plots and data.

```{r, message=FALSE, warning=FALSE, eval=F}
dir.create("plots")
dir.create("data")
download.file("https://raw.githubusercontent.com/DrThomasOneil/Digital-Research-Skills-Network/refs/heads/main/docs/assets/downloads/example_flowdata_omip082_20k.csv", destfile = "data/example_flowdata_omip082_20k.csv")
```

<details><summary>Setting seed is crucial for reproducibility:</summary>

It's a number of your choice, which pins down the "way" that how the random process will be performed when you re-run these codes. (Easier example: Say if you randomly take 10 numbers with a mean of 50, there will be tonnes of ways to generate the numbers. But a seed ensures that the report always returns the same set of numbers. It's the similar idea when you run other algorithms including randomness, e.g. clustering and generating a UMAP).

```{r}
set.seed(1337)
```
</details>

## Load flow cytometry data

The expression data is stored in a `.csv` file. We'll first load this in.

```{r}
# Read the .csv file in the working directory.
ILC <- read.csv("data/example_flowdata_omip082_20k.csv")
```

Have a look at the broad features of this dataset. In this dataset, the channel names are stored as column names. Each row represents a cell, storing expression values in each channel. 

```{r, eval=F}
# Dimension: number of cells x number of channels.
dim(ILC)

# Check the column names.
colnames(ILC)
```

The column names are wordy, with markers, fluorophores, and channels. We just need the marker names to proceed. The first column 'X' is just indexing the number of rows. It's not necessary either.<br>

Remove the first column.

```{r}
# This keeps the 2-33 columns in the dataset, and store the change.
ILC <- ILC[, -1]
colnames(ILC)

# Same with it, you can just enter the column indexes from 2-33.
# ILC <-ILC[,c(2:33)]
```

Change the column names.

```{r}
# Substitute everything after the first '_' with an empty "" in the column names that have the pattern "_" and something.
colnames(ILC) <- sub("_.*","",colnames(ILC)) #'_' is the exact character in the names, the '.' means 'any character', '*' means 'any of the preceding things'. 

colnames(ILC)

# Fix up the broken names FVS700 and CD14&CD19
# This tells R to: find which of the column equals "FVS" among all columns in the data, then change it to FVS700.
colnames(ILC)[colnames(ILC) == "FVS"] <- "FVS700"
colnames(ILC)[colnames(ILC) == "CD14"] <- "CD14.CD19"

# Have a look
colnames(ILC)
```

## Time gate QC. 

The data is already preprocessed in FlowJo. But if you are interested, you can view the Time gate, and filter the data based on this in R.

```{r}
ggplot(ILC, aes(x = Time, y = FSC.A))+
  geom_bin_2d(bins = 200, show.legend = FALSE)+
    scale_fill_gradientn(colours = c("blue", "green", "yellow", "red"))+
  geom_vline(xintercept = 4600000)

# Filter out the cells above the threshold.
ILC <- ILC[ILC$Time < 4600000,]
ggplot(ILC, aes(x = Time, y = FSC.A))+
  geom_bin_2d(bins = 200, show.legend = FALSE)+
    scale_fill_gradientn(colours = c("blue", "green", "yellow", "red"))
```

# Visualisations

## Transformations

Raw Flow Cytometry data looks bad!

```{r}
ILC %>% ggplot(aes(x=CD4, y=CD3))+        
  geom_point()
```

So we transform the data to make it look better. In FlowJo, when you adjust the axes, you can see several transformations available. One is arcsinh (the inverse of hyperbolic sin, if you want to get mathy). Smaller values centre around zero, and larger values separate, giving the appearance that we're familiar with. 

We are also able to adjust the level of transformation by dividing the values by some cofactor. 

```{r, fig.height=10, fig.width=10}
# decide on a cofactor - try changing it and seeing how the data changes
cofactor = 500
ILC_transf <- asinh(ILC[,7:31]/cofactor) #7:31 are the columns with markers.

ILC_transf <- cbind(ILC_transf, ILC[,c(1:6,32)]) # Add the first 6 columns & time back in.

ILC_transf %>% ggplot(aes(x=CD4, y=CD3))+
  geom_bin_2d(bins = 200, show.legend=T)
```


```{r, fig.height=10, fig.width=10, warning=FALSE}
# Initialize an empty list to store the plots
plots <- list()
# Define the x-axis marker for the plots
xaxis = "CD39" # Choose an x-axis marker for comparison

# Loop through each column in the dataset
for(i in 1:ncol(ILC_transf)) {
  # Get the marker name for the current column
  Marker <- colnames(ILC_transf)[i]
  
  # Generate a 2D bin plot for the current marker against the chosen x-axis marker
  p <- ILC_transf %>% 
        ggplot(aes_string(x = xaxis, y = Marker)) +        
        geom_bin_2d(bins = 200, show.legend = FALSE) + # Create a heatmap-style scatter plot
        scale_fill_gradientn(colors = c("blue", "yellow", "red")) # Define the color gradient
  
  # Add the generated plot to the list
  plots[[i]] <- p
}

# Combine all the plots into a grid layout and display them
cowplot::plot_grid(plotlist = plots)
rm(plots)
```

This code demonstrates how you would manually transform individual columns in the dataset using different cofactors for each marker. It creates a new dataset (`ILC_transf_indiv`) where specific columns are transformed using the `asinh` function with their respective cofactors. This allows for more granular control over transformations. 

```{r, eval=F}
ILC_transf_indiv <- ILC
ILC_transf_indiv[, "CD3"] <- asinh(ILC[,"CD3"]/1000)
ILC_transf_indiv[, "CD4"] <- asinh(ILC[,"CD4"]/500)
# and so on...

# Visualize it
ILC_transf_indiv %>% ggplot(aes(x=CD3, y=CD4))+        
        geom_bin_2d(bins = 200, show.legend=F) +
        scale_fill_gradientn(colors = c("blue",  "yellow", "red"))

rm(ILC_transf_indiv)
```

## Histograms

We can visualize the marker expressions via histograms and density plots.

```{r}
# For example, this generates a histogram of CD3 expression.
ILC_transf %>%
  ggplot(aes(x = CD3))+
  # Plot the histogram, and change the fill colour to lightblue, the line colour to black.
  geom_histogram(bins = 35, fill = "lightblue", colour ="black")+
  # Add labels
  ylab("Frequency")

# A density plot.
ILC_transf %>%
  ggplot(aes(x = CD3))+
  geom_density(linewidth = 0.7, colour = "red4", aes(fill = "red"), alpha = 0.5,show.legend = FALSE)+
  ylab("Density")

```

We can loop through the channels and create a density plot for all 25 markers.

```{r, message = FALSE}
# Initialise an empty list to store the plots
plots<- list()

# Loop through each column in the dataset
for (i in 1:ncol(ILC_transf)){

  # Get the marker name for the current column
  Marker <- colnames(ILC_transf)[i]

  # Generate the histogram for the current marker
  p <- ILC_transf %>%
  ggplot(aes(x = .data[[Marker]]))+
    geom_density(linewidth = 0.7, colour = "red4", aes(fill = "red"), alpha = 0.5,show.legend = FALSE)+
    labs(x=Marker,y="")+
    theme(axis.text = element_text(size=1))

  # Add the generated plot to the list
  plots[[i]] <- p
}

plot_grid(plotlist = plots[1:25])
rm(plots)
```
<br>

## Scatter plot

Like what we do in Flowjo, we can visualize 2 markers' expressions, We can generate scatterplots.

```{r}
ILC_transf %>%
  ggplot(aes(x = CD4, y = CD3))+
  geom_point(color="blue", alpha=0.1)

# Generate a scatterplot with a density overlay.
ILC_transf %>% ggplot(aes(x=CD4, y=CD3))+        
  geom_bin_2d(bins = 200, show.legend = F)+
  scale_fill_gradientn(colors = c("blue",  "yellow", "red"))

```

## Contour plot

And a 2-marker countour plot. You can easily make it more stylish by adding layers in ggplot!

```{r collapse=T}

# A contour plot of CD3 and CD4 expression values.
ILC_transf %>%
  ggplot(aes(x = CD3, y = CD4))+
  geom_density_2d(colour = "black", bins=50)+
  xlab("CD3 BUV563")+
  ylab("CD4 APC-Cy7")

# A contour plot of CD3 and CD4 expression over a scatter plot.
ILC_transf %>%
  ggplot(aes(x = CD3, y = CD4))+        
  geom_point(color = "grey", alpha=0.5)+
  geom_density_2d(colour = "black", bins=50, alpha=0.5)+
  xlab("CD3 BUV563")+
  ylab("CD4 APC-Cy7")

# A contour plot of CD3 and CD4 expression over a density plot.
ILC_transf %>%
  ggplot(aes(x = CD3, y = CD4))+        
  geom_bin_2d(bins = 200)+
  scale_fill_gradientn(colors = c("blue",  "yellow", "red"))+
  geom_density_2d(colour = "black")+
  xlab("CD3 BUV563")+
  ylab("CD4 APC-Cy7")

```

# Subsetting data and summary statistics{.tabset .tabset-fade}

While not as efficient as FlowJo, we can subset data after exporting. We wouldn't normally do this in R, but we'll show you how you could subset in R and for the purposes of visualisation. 

## Subsetting
```{r}
# Plot a 2-marker plot of CD3 and CD4 expression values. 
ILC_transf %>%
  ggplot(aes(x = CD3, y = CD4))+        
  geom_bin_2d(bins = 200)+
  scale_fill_gradientn(colors = c("blue",  "yellow", "red"))+
  xlab("CD3 BUV563")+
  ylab("CD4 APC-Cy7")+
  geom_vline(xintercept = 1.3)+ #choose the x intercept
  geom_hline(yintercept = 0.6)  #choose the y intercept
```

Subset the dataset based on these thresholds.

```{r}
# This code creates a new column named 'subset' in the 'ILC_transf' dataset.
# The 'subset' column categorises rows based on the values of 'CD3' and 'CD4' columns:
# - "CD3+CD4+": if CD3 > 1.3 and CD4 > 0.6
# - "CD3+CD4-": if CD3 > 1.3 and CD4 <= 0.6
# - "CD3-CD4+": if CD3 <= 1.3 and CD4 > 0.6
# - "CD3-CD4-": if CD3 <= 1.3 and CD4 <= 0.6
# The 'mutate' function from the dplyr package is used to add this new column.

ILC_transf <- ILC_transf %>% 
  mutate(subset = case_when(
    CD3 > 1.3 & CD4 > 0.6 ~ "CD3+CD4+",
    CD3 > 1.3 & CD4 <= 0.6 ~ "CD3+CD4-",
    CD3 <= 1.3 & CD4 > 0.6 ~ "CD3-CD4+",
    TRUE ~ "CD3-CD4-"
  ))
```

### Visualise the subsets in dot plots
```{r}
ILC_transf %>%
  ggplot(aes(x = CD3, y = CD4, color=subset))+        
  geom_point()+
  xlab("CD3 BUV563")+
  ylab("CD4 APC-Cy7")+geom_vline(xintercept = 1.3)+
  geom_hline(yintercept = 0.6)

# Look at different markers coloured by subset
ILC_transf %>%
  ggplot(aes(x = CD56, y = CD16, color=subset))+        
  geom_point()

# Split by subset
ILC_transf %>%
  ggplot(aes(x =CD56, y = CD16))+        
  geom_bin_2d(bins = 200)+
  scale_fill_gradientn(colors = c("blue",  "yellow", "red"))+
  facet_wrap(~subset)

ILC_transf %>%
  ggplot(aes(x =FSC.A, y = SSC.A))+        
  geom_bin_2d(bins = 200)+
  scale_fill_gradientn(colors = c("blue",  "yellow", "red"))+
  facet_wrap(~subset)
```

### Histograms

```{r, warning=FALSE}
ILC_transf %>%
  ggplot(aes(x = CD56, color = subset, fill = subset)) +
  geom_density(alpha = 0.2, linewidth = 1)+
  xlim(quantile(ILC_transf$CD56, 0.001),quantile(ILC_transf$CD56, 0.999))

#using ggridges
ILC_transf %>%
  ggplot(aes(x = CD56, y = subset, fill = subset)) +
  geom_density_ridges(scale = 1.2, alpha = 0.6, colour = "black") +
  theme_ridges() +
  theme(legend.position = "none")+
  labs(y="", x="", title = "CD56")+
  xlim(quantile(ILC_transf$CD56, 0.001),quantile(ILC_transf$CD56, 0.999))
```

You could mass print these using a loop, as we did for the density plots above. You can also adjust the marker you output, as not all markers or combinations of markers are useful.

## Summary statistics

While subsetting is not as efficient as FlowJo, summarising data is! 

We can first collect subset percentages and output them in a table, in text, and even in a plot. Then we can calculate summary statistics like MFI & gMFI.

```{r}
# total summary
round(table(ILC_transf$subset)*100/nrow(ILC_transf),2) # % of each subset

# individual marker
ILC_transf %>% 
  group_by(subset) %>% 
  summarise(Percent = round(n()/nrow(ILC_transf)*100,2)) %>%
  mutate(subset = factor(subset, levels = c("CD3+CD4+", "CD3+CD4-", "CD3-CD4+", "CD3-CD4-")))%>%
  mutate(Subsets = "Subsets") %>%
  ggplot(aes(x=Subsets,y = Percent, fill = subset)) +
    geom_bar(stat = "identity", show.legend = FALSE) +
    ylab("Percentage") +
    xlab("") +
    scale_fill_manual(values = c("CD3+CD4+" = "blue", "CD3+CD4-" = "red", "CD3-CD4+" = "green", "CD3-CD4-" = "yellow")) +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    theme(axis.text.y = element_text(size=10)) +
    theme(legend.position = "none") +
    theme(plot.title = element_text(hjust = 0.5, size=10)) +
    ggtitle("Subsets of CD3+/-CD4+/- cells")
```

We can also calculate the percentage of subsets in line, meaning we don't have to change these lines of text based on different results or different data:
- CD3+CD4+: `r round(sum(ILC_transf$subset == "CD3+CD4+")/nrow(ILC_transf)*100,2)`%

- CD3+CD4-: `r round(sum(ILC_transf$subset == "CD3+CD4-")/nrow(ILC_transf)*100,2)`%

- CD3-CD4+: `r round(sum(ILC_transf$subset == "CD3-CD4+")/nrow(ILC_transf)*100,2)`%

- CD3-CD4-: `r round(sum(ILC_transf$subset == "CD3-CD4-")/nrow(ILC_transf)*100,2)`%

<details>
<summary>**You can also quickly calculate summary statistics like MFI, gMFI, medianFI**:</summary>

**Arithmatic MFI**  

```{r}
ILC$subset  <- ILC_transf$subset # Add the subset column to the original data.

ILC %>% 
  group_by(subset) %>% 
  summarise(across(7:31, mean, .names = "MFI_{col}"))

```

**Median FI**

```{r}
ILC %>% 
  group_by(subset) %>% 
  summarise(across(7:31, mean, .names = "MFI_{col}"))
```

**geometric MFI**

The traditional formula written out would be like `ILC_transf %>% prod(CD56)^(1/nrow(CD56))`. However, as our data is *compensated*, there are meaningful negative values and 0s that we don't want to get rid of, which if you run on this data removes two subsets entirely. FlowJo calculates the *graph spaces* as gMFI to enable it to deal with negatives and 0s. Here in this case, our data is already asinh transformed to accomodate the -ve and 0s and bizzare expression values. It may be better to stick with the arithmatic mean instead of forcing the geometric mean.

</details>

<br>

# nxn plots

You can also create a loop to view the nxn plots.

```{r, fig.width=20, fig.height=20, warning=FALSE}
# Select only numeric marker columns
marker_cols <- colnames(ILC_transf)[1:25] # adjust if needed

# Make a customised function for plot streamline
custom_bin_plot <- function(data, mapping, ...) {
  ggplot(data = data, mapping = mapping) +
    geom_bin_2d(bins = 50) +
    scale_fill_gradientn(colours = c("blue", "yellow", "red")) +
    theme_minimal()
}

# Generate plots
p <- ggpairs(
  ILC_transf,
  columns = marker_cols,
  lower = list(continuous = custom_bin_plot),
  diag = list(continuous = wrap("densityDiag", alpha = 0.4)),
  upper = list(continuous = wrap("cor", size = 3))
)
ggsave("plots/nxn.png",p, width=25, height=25,unit="in")

rm(p,custom_bin_plot, marker_cols)
```


# Dimension reduction and clustering

There are several ways to perform clustering (e.g. FlowSOM, FuseSOM) and dimension reduction (e.g. PCA, UMAP, tSNE). 

In this section, we'll cover **clustering** and **dimension reduction** on our one sample. 

## Clustering{.tabset .tabset-fade}

Clustering groups our cells based on the input expressions we choose. 

### FlowSOM
[Read more](https://bioconductor.org/packages/release/bioc/html/FlowSOM.html)

```{r, message = FALSE}

# Perform clustering
clusters <- FlowSOM::FlowSOM(
  input=as.matrix(ILC_transf[,1:32]),
  silent = FALSE,
  colsToUse = c(1,4,6:14,16:25),
  nClus = 15)

# Annotate the cluster that the cell belongs to
ILC_transf$flowSOM<- GetMetaclusters(clusters)
FlowSOM::GetMetaclusterMFIs(clusters)

# Plot CD4 expression values on clusters
ILC_transf %>%
  ggplot(aes(x = CD56, y = flowSOM, fill = flowSOM)) +
  geom_density_ridges(scale = 1.2, alpha = 0.6, colour = "black") +
  theme_ridges() +
  theme(legend.position = "none")+
  labs(y="", x="", title = "CD56")
```

### FuseSOM
[Read more](https://bioconductor.org/packages/release/bioc/html/FlowSOM.html)

```{r, message = FALSE}

risomMarkers <- colnames(ILC_transf)[c(1,4,6:14,16:25)]
clusters <- runFuseSOM(data = as.matrix(ILC_transf[,1:32]), markers = risomMarkers, 
                        numClusters = 15)

ILC_transf$fuseSOM<- as.numeric(unlist(lapply(str_split(clusters$clusters, "_"), function(x) x[[2]])))

ILC_transf %>%
  ggplot(aes(x = CD39, y = as.factor(fuseSOM), fill=as.factor(fuseSOM))) +
  geom_density_ridges(scale = 1.2, alpha = 0.6, colour = "black") +
  theme_ridges() +
  theme(legend.position = "none")+
  labs(y="", x="", title = "CD39")
```

## Dimension Reduction{.tabset .tabset-fade}

We'll generate a PCA, a tSNE and a UMAP. I'll then run through all of this in the Spectre packge. 

```{r message=F, warning=F, eval=F}

pca <- prcomp(as.matrix(ILC_transf[,c(1,4,6:14,16:25)]), center = TRUE)

ILC_transf <- ILC_transf %>% 
  mutate(PC1 = pca$x[,1],
         PC2 = pca$x[,2])

# can take a while
tsne <- Rtsne(as.matrix(ILC_transf[,c(1,4,6:14,16:25)]), dims = 2, perplexity = 30, verbose = TRUE)

ILC_transf <- ILC_transf %>% 
  mutate(tSNE1 = tsne$Y[,1],
         tSNE2 = tsne$Y[,2])

umap <- umap(ILC_transf[,c(1,4,6:14,16:25)])
ILC_transf <- ILC_transf %>% 
  mutate(UMAP1 = umap$layout[,1],
         UMAP2 = umap$layout[,2])

rm(pca,umap,tsne)
```
```{r, eval=F}
saveRDS(ILC_transf,"data/clustered_data.rds")
```

### Visualisations

```{r}
ILC_transf <- readRDS("data/clustered_data.rds")
ggplot(ILC_transf, aes(PC1,PC2, color= as.factor(flowSOM))) +
  geom_point()

ggplot(ILC_transf, aes(tSNE1,tSNE2, color= as.factor(flowSOM))) +
  geom_point()

ggplot(ILC_transf, aes(UMAP1,UMAP2, color= as.factor(flowSOM))) +
  geom_point()+xlim(-7,10)

# Colour overlayed by the expression of one marker.
ggplot(ILC_transf, aes(UMAP1,UMAP2, color= CD3)) +
  geom_point() + 
  scale_color_viridis_c()+xlim(-7,10)

ggplot(ILC_transf, aes(UMAP1,UMAP2, color= CD16)) +
  geom_point() + 
  scale_color_viridis_c()+xlim(-7,10)

```

### Spectre

[Read more](https://immunedynamics.io/spectre/)

```{r, message = F, warning=FALSE}
# Spectre has easy plotting tools to visualize the results of clustering and dimension reduction.
make.colour.plot(ILC_transf, 
                 "tSNE1", "tSNE2", "subset",  
                 add.label = TRUE,
                 blank.axis = T,
                 path = "plots")
make.colour.plot(ILC_transf, 
                 "tSNE1", "tSNE2", "flowSOM", 
                 col.type = 'factor', 
                 add.label = TRUE,
                 blank.axis = T,
                 path = "plots")
```

<details><summary>**Multi-plots:**</summary>

```{r, message = F, warning=FALSE}
make.multi.plot(ILC_transf, 
                "tSNE1", "tSNE2", 
                risomMarkers, 
                blank.axis = T, 
                colours = "viridis", 
                path = "plots")
```

</details>

And we can make the heatmap you're used to seeing. 

```{r, message = F, warning=FALSE, error=F}
exp <- do.aggregate(as.data.table(ILC_transf), use.cols = risomMarkers, by = c("flowSOM"))
make.pheatmap(exp, file.name = "plots/Pheatmap_FlowSOM.png", sample.col="flowSOM", plot.cols=risomMarkers)
```


*Spectre has also embedded FlowSOM clustering and dimension reduction tools such as umap in easy codes, such as:<br>*

-   `do.subsample()`
-   `run.flowsom()`
-   `run.umap()`

# Additional Considerations{.tabset .tabset-fade}

## QC

When looking at the raw data, or in the middle of data processing, you may consider filtering some cell. In the 1-marker plots below, you can see that there are a few extreme values that are displayed and pushing the majority of values squeezed together. e.g. HLA-DR:

```{r}
ILC_transf %>%
  ggplot(aes(x = HLA.DR))+
  geom_density(linewidth = 0.7, colour = "red", aes(fill = "red"), alpha = 0.5,show.legend = FALSE)+
  ylab("Density")+

# Visualise the values at the top 0.1% and bottom 0.1%.
  geom_vline(xintercept = quantile(ILC_transf$HLA.DR, 0.999), colour = "blue", linewidth = 0.5)+
  geom_vline(xintercept = quantile(ILC_transf$HLA.DR, 0.001), colour = "blue", linewidth = 0.5)+

# We can label the threshold lines: annotate as a text, at the position of the top and bottom 0.01%, and slightly adjust the label position horizontally.
  annotate("text", x = quantile(ILC_transf$HLA.DR, 0.999), y = 0.003, label = "Top and bottom 0.1%", hjust = -0.1) 
```

<details>
<summary>View them for all channels.</summary>

```{r, message=F}
# Create the empty list
plots_line <- list()

# Create the loop to generate plots.
for (i in 1:25){
  
  # Get the marker names
  Marker <- colnames(ILC_transf)[i]
  
  # Generate the plots and put it into the list
  p <- ILC_transf %>%
  ggplot(aes(x = .data[[Marker]]))+ 
  geom_density(linewidth = 0.7, colour = "red", fill = "red", alpha = 0.3,show.legend = FALSE)+
   xlab(Marker)+
   ylab("Density")+
  
    # Add the newly written lines to visualise the outliers to this loop.
    geom_vline(xintercept = quantile(ILC_transf[[i]], 0.999), colour = "blue", linewidth = 0.5)+
    geom_vline(xintercept = quantile(ILC_transf[[i]], 0.001), colour  = "blue", linewidth = 0.5)
    
  # Put the plots into the list
  plots_line[[length(plots_line)+1]] <- p
}

# Use a grid layout to view the plots
library(ggpubr)
ggpubr::ggarrange(plotlist = plots_line, ncol = 2, nrow = 4)
```
</details>

<br>
To look at this in a more statistical way, you can create a quick data summary of the minimum, maximum, mean, median and the interquartile range.

```{r}
summary(ILC_transf)
```

<br>

Either being artificial or biological, these might interfere how the computer interprets your data in the downstream clustering and dimension reduction. Sometimes you may decide to investigate how these impact your dataset and where they arise from. But if there are plenty of cells, and the % outliers is super small, you might just want to filter them out.<br>

**Filtering by %** <br>
You might want to filter out the top and bottom 0.1% of cells . Note that this approach removes the cells that doesn't fit within the top and bottom 0.1% of ANY channel, and removes the cell from the entire dataset. So be cautious about the % so you aren't removing too many cells when they are added up ( we'll use 0.1% all the way for demonstration purposes). <br>

```{r}
# Create an empty list.
outlier_flag <- list()

# Loop through the columns with markers
for (i in 1:25){
  
  # Find the cell values
  value <- ILC_transf[[i]]
  
  # Get the marker names
  Marker <- colnames(ILC_transf)[i]
  
  # Find the thresholds of outliers
  upper <- quantile(value, 0.999, na.rm = TRUE)
  lower <- quantile(value, 0.001, na.rm = TRUE)

  # Flag the outliers as 1, else as 0. And record them into the list.
  flag <- ifelse(value < lower | value > upper, 1, 0)
  outlier_flag[[length(outlier_flag)+1]] <- flag
}
# Turn the flags into a data table object.
outlier_flag <- as.data.table(outlier_flag)
colnames(outlier_flag)<- colnames(ILC_transf)[c(7:31)]
subset <- ILC_transf[rowSums(outlier_flag)==0,]
  
# Calculate the % of cell loss after filtering, then format it as being displayed in less digits, then paste a sentence together.
    paste0("Cell loss: ",format((1-nrow(subset)/nrow(ILC_transf))*100, digits = 3), "%")
```


```{r}

# View the pre- and post-filtered data side by side, putting them in a grid.
cowplot::plot_grid(
  ILC_transf %>%
  ggplot(aes(x = HLA.DR))+
  geom_density(linewidth = 0.7, colour = "red", fill = "red", alpha = 0.3, show.legend = FALSE)+
  ylab("Density"),

subset %>%
  ggplot(aes(x = HLA.DR))+
  geom_density(linewidth = 0.7, colour = "red", fill = "red", alpha = 0.3, show.legend = FALSE)+
  ylab("Density"),

 # number of column equals 1.
 ncol = 1
)

```

**Filtering by fixed thresholds** <br>
You might want to filter out the data that is outside of certain values per channel. A commonly used standard in data cleaning is calculated based on interquartile ranges (IQR). The thresholds are any data points that are 1.5 times of IQR above the 3rd quartile, or 1.5 times of IQR below the 1st quartile. This will keep the majority of data points, however, it might also remove some true variance in the dataset at extreme ends.<br>

```{r}

# Take HLA-DR as an example

# Calculate the interquartile range.
   IQR <- IQR(ILC_transf$HLA.DR)
   Q1 <- quantile(ILC_transf$HLA.DR, 0.25)
   Q3 <- quantile(ILC_transf$HLA.DR, 0.75)
   
# Calculate the upper and lower thresholds
   lowerfence <- Q1 - 1.5*IQR
   upperfence <- Q3 + 1.5*IQR 
 
# Only keep the cells which CD127 channel value within the thresholds.   
filt <- ILC_transf[ILC_transf$HLA.DR > lowerfence & ILC_transf$HLA.DR < upperfence, ]

cowplot::plot_grid(
  ILC_transf %>%
  ggplot(aes(x = HLA.DR))+
  geom_density(linewidth = 0.7, colour = "red", fill = "red", alpha = 0.3,show.legend = FALSE)+
  ylab("Density"),
filt %>%
  ggplot(aes(x = HLA.DR))+
  geom_density(linewidth = 0.7, colour = "red", fill = "red", alpha = 0.3,show.legend = FALSE)+
  ylab("Density"),

 # number of column equals 1.
 ncol = 1
)

rm(Q1,Q3,IQR)

```

Or, simply just choose an arbitrary threshold.

```{r}

# Only keep the cells which transformed HLA-DR expression value < 2.
manual <- ILC_transf[ILC_transf$HLA.DR < 2, ]

cowplot::plot_grid(
  ILC_transf %>%
  ggplot(aes(x = HLA.DR))+
  geom_density(linewidth = 0.7, colour = "red", fill = "red", alpha = 0.3,show.legend = FALSE)+
  ylab("Density"),
manual %>%
  ggplot(aes(x = HLA.DR))+
  geom_density(linewidth = 0.7, colour = "red", fill = "red", alpha = 0.3,show.legend = FALSE)+
  ylab("Density"),

 # number of column equals 1.
 ncol = 1
)

```
</div>
<div class="custom-footer">
  <div class="footer-left" style="width: 30%; display: flex; justify-content: center; align-items: center;">
    <a href="https://wimr.org.au/" target="_blank">
      <img src="assets/footer/footer1.png" alt="wimr-logo" style="max-height: 100px; width: auto;">
    </a>
   </div>
  <div class="footer-center">
    <p><strong>Digital Research Skills Network</strong><br>
      <a href="mailto:thomas.oneil@sydney.edu.au"><strong>Contact us</strong></a>
    </p>
  </div>
  <div class="footer-right" style="width: 30%; display: flex; justify-content: center; align-items: center;">
    <a href="https://www.precisiongo.com.au/" target="_blank">
      <img src="assets/footer/PGlogo.png" alt="wimr-logo" style="max-height: 100px; width: auto;">
    </a>  
  </div>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("202504_Workshop.Rmd");
  window.initializeCodeFolding("hide" === "show");
});
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
