<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Thomas O’Neil" />


<title>Tidyverse 2: Data Manipulation</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; } /* Alert */
code span.an { color: #008000; } /* Annotation */
code span.at { } /* Attribute */
code span.bu { } /* BuiltIn */
code span.cf { color: #0000ff; } /* ControlFlow */
code span.ch { color: #008080; } /* Char */
code span.cn { } /* Constant */
code span.co { color: #008000; } /* Comment */
code span.cv { color: #008000; } /* CommentVar */
code span.do { color: #008000; } /* Documentation */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.im { } /* Import */
code span.in { color: #008000; } /* Information */
code span.kw { color: #0000ff; } /* Keyword */
code span.op { } /* Operator */
code span.ot { color: #ff4000; } /* Other */
code span.pp { color: #ff4000; } /* Preprocessor */
code span.sc { color: #008080; } /* SpecialChar */
code span.ss { color: #008080; } /* SpecialString */
code span.st { color: #008080; } /* String */
code span.va { } /* Variable */
code span.vs { color: #008080; } /* VerbatimString */
code span.wa { color: #008000; font-weight: bold; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>


<link rel="stylesheet" href="adit/custom.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#section-TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#section-TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="section-TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src='assets/logo/logo.png' style='height:30px; vertical-align:middle; align-items:center; margin-right:10px;'></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="0_start.html">Getting Started in R</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Introduction Series
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="202502_tidyverse1.html">Tidyverse 1: Intro to Plotting</a>
    </li>
    <li>
      <a href="202502_tidyverse2.html">Tidyverse 2: Intro to Manipulation</a>
    </li>
    <li>
      <a href="202503_tidyverse3.html">Tidyverse 3: Intro to Manipulation</a>
    </li>
    <li>
      <a href="202503_python.html">Python vs R</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="0_resources.html">Extra Resources</a>
</li>
<li>
  <a href="0_contacts.html">Contacts</a>
</li>
<li>
  <a href="https://github.com/DrThomasOneil/analysis-user-group">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="section-header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Tidyverse 2: Data Manipulation</h1>
<h4 class="author">Thomas O’Neil</h4>
<h4 class="date">2025-02</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-02-24
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>analysis-user-group/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="section-workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="section-strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="section-strongEnvironmentstrongempty"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1337code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1337)</code> </a>
</p>
</div>
<div id="section-strongSeedstrongcodesetseed1337code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1337)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="section-strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="section-strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="section-strongFilepathsstrongrelative"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomDrThomasOneilanalysisusergrouptreec9db666a9e7f8c332c6f7eb8056d91c1464772eatargetblankc9db666a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/DrThomasOneil/analysis-user-group/tree/c9db666a9e7f8c332c6f7eb8056d91c1464772ea" target="_blank">c9db666</a>
</a>
</p>
</div>
<div
id="section-strongRepositoryversionstrongahrefhttpsgithubcomDrThomasOneilanalysisusergrouptreec9db666a9e7f8c332c6f7eb8056d91c1464772eatargetblankc9db666a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/DrThomasOneil/analysis-user-group/tree/c9db666a9e7f8c332c6f7eb8056d91c1464772ea" target="_blank">c9db666</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/data/
    Ignored:    analysis/figure/

Unstaged changes:
    Modified:   workflow.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/202502_tidyverse2.Rmd</code>) and
HTML (<code>docs/202502_tidyverse2.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/analysis-user-group/blob/c9db666a9e7f8c332c6f7eb8056d91c1464772ea/analysis/202502_tidyverse2.Rmd" target="_blank">c9db666</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-02-24
</td>
<td>
wflow_publish(c("analysis/*.Rmd"))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/analysis-user-group/af42fbabcad839511c5c14ed3847c04df3f6d9a4/docs/202502_tidyverse2.html" target="_blank">af42fba</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-02-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/analysis-user-group/blob/24a9fefc515af1ddd0034edcfc2ab9bbc08980b4/analysis/202502_tidyverse2.Rmd" target="_blank">24a9fef</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-02-24
</td>
<td>
wflow_publish(c("analysis/*.Rmd"))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/analysis-user-group/1aeefc7a88b9f306f0a61389e72811dcf369b9d7/docs/202502_tidyverse2.html" target="_blank">1aeefc7</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-02-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/analysis-user-group/blob/756ebac1ac832b5f0324d8c87dc329b641b2a7b0/analysis/202502_tidyverse2.Rmd" target="_blank">756ebac</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-02-20
</td>
<td>
wflow_publish(c("analysis/*.Rmd"))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/analysis-user-group/5fe30de6735f75be6fe53a89cb092266f13ed73e/docs/202502_tidyverse2.html" target="_blank">5fe30de</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-02-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/analysis-user-group/blob/477062c98f2270023f89bf44908c43cf6bd3d446/analysis/202502_tidyverse2.Rmd" target="_blank">477062c</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-02-20
</td>
<td>
wflow_publish(c("analysis/*.Rmd"))
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/analysis-user-group/blob/a48de9d79d048625c7f7b54ca4bd4e57605ca16a/analysis/202502_tidyverse2.Rmd" target="_blank">a48de9d</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-02-20
</td>
<td>
new data
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="section-introduction"
class="section level1 tabset tabset-fade">
<h1 class="tabset tabset-fade">Introduction</h1>
<p>In this tutorial, we explore the functionalities of the Tidyverse by
working with two datasets:</p>
<ul>
<li><code>data.csv</code> contains fabricated summary flow cytometry
data, such as cell numbers and MFI</li>
<li><code>meta.csv</code> contains information regarding each donor,
such as age and sex.</li>
</ul>
<p>The Tidyverse is a collection of R packages that share an underlying
design philosophy and grammar, making data analysis more intuitive and
coherent. Here are some key benefits:</p>
<p><strong>dplyr</strong>: Provides a suite of verbs
(e.g. <code>filter()</code>, <code>mutate()</code>,
<code>select(</code>), and <code>left_join()</code>) that simplify data
manipulation. For example, <code>left_join()</code> easily combines
experimental data with metadata based on a common key. See the <a
href="https://drthomasoneil.github.io/analysis-user-group/r-tutorial/_book/chapter-4-data-manipulation-with-tidyverse.html">Introduction
to R: Chapter 4</a> for more content on dplyr.</p>
<p><strong>Pipes (%&gt;%)</strong>: Enhance code readability by allowing
you to chain multiple operations in a sequential, natural-language
style. This means you can finalise complex data transformations in a
single, clear pipeline without the need for numerous intermediate
variables.</p>
<p><strong>tidyr</strong>: Focuses on tidying data, ensuring that each
observation occupies a single row and each variable a single column.
Functions like <code>pivot_longer()</code> and pivot_wider() help
reshape data into a standard format, which is essential for further
analysis.</p>
<p><strong>ggplot2</strong>: Offers a powerful and flexible grammar for
data visualisation. With ggplot2, you can create customised plots that
effectively communicate insights from your data, building on the clean,
tidy data produced by the other Tidyverse packages. See the <a
href="https://drthomasoneil.github.io/analysis-user-group/r-tutorial/_book/chapter-5-data-visualisation-with-ggplot2.html">Introduction
to R: Chapter 5-6</a> and the <a
href="https://drthomasoneil.github.io/analysis-user-group/202502_tidyverse1.html">previous
workshop</a> by Harry for extra information on ggplot.</p>
<div class="hint-goals">
<p><strong>Goals:</strong> Learn data manipulation with Tidyverse<br>
  • Tidy Data<br>   • Select and Mutate data<br>   • Group and summarise
data<br>   • Join two data frames<br>   • Pivot data with
<code>tidyr</code><br></p>
</div>
<p>Whether you are combining datasets, reshaping data, or creating
compelling graphics, the Tidyverse offers a consistent and powerful set
of tools to support your analysis.</p>
<p><read-time><i><strong>Estimated length</strong>: 1
Hour</i></read-time><wimr></p>
</div>
<div id="section-set-up" class="section level1">
<h1>Set up</h1>
<details>
<summary>
<strong>Follow these instructions to get started:</strong>
</summary>
<p><br><br> 1. Save this script in a folder called
<strong>“202502_Tidyverse2”</strong></p>
<ol start="2" style="list-style-type: decimal">
<li>Install and load the relevant packages and functions</li>
</ol>
<div class="sourceCode" id="section-cb1"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb1-1"><a href="#section-cb1-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;tidyverse&quot;</span>)</span>
<span id="section-cb1-2"><a href="#section-cb1-2" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;shiny&quot;</span>)</span>
<span id="section-cb1-3"><a href="#section-cb1-3" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;readr&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="section-cb2"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb2-1"><a href="#section-cb2-1" tabindex="-1"></a><span class="fu">suppressWarnings</span>(<span class="fu">library</span>(shiny))</span>
<span id="section-cb2-2"><a href="#section-cb2-2" tabindex="-1"></a><span class="fu">suppressWarnings</span>(<span class="fu">library</span>(tidyverse))</span></code></pre></div>
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.0.4     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
<div class="sourceCode" id="section-cb4"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb4-1"><a href="#section-cb4-1" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_classic</span>())</span>
<span id="section-cb4-2"><a href="#section-cb4-2" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">url</span>(<span class="st">&quot;https://raw.githubusercontent.com/DrThomasOneil/SkillTrees/refs/heads/main/assets/.functions.R&quot;</span>)) </span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Create new folders in your directory.</li>
</ol>
<div class="sourceCode" id="section-cb5"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb5-1"><a href="#section-cb5-1" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">&quot;plots&quot;</span>)</span>
<span id="section-cb5-2"><a href="#section-cb5-2" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">&quot;data&quot;</span>)</span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Download the data.</li>
</ol>
<div class="sourceCode" id="section-cb6"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb6-1"><a href="#section-cb6-1" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url=</span><span class="st">&quot;https://raw.githubusercontent.com/DrThomasOneil/analysis-user-group/refs/heads/main/docs/r-tutorial/assets/synthetic_data.csv&quot;</span>, </span>
<span id="section-cb6-2"><a href="#section-cb6-2" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">&quot;data/data.csv&quot;</span>, <span class="at">method=</span><span class="st">&#39;curl&#39;</span>, <span class="at">mode=</span><span class="st">&#39;wb&#39;</span>)</span>
<span id="section-cb6-3"><a href="#section-cb6-3" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url=</span><span class="st">&quot;https://raw.githubusercontent.com/DrThomasOneil/analysis-user-group/refs/heads/main/docs/r-tutorial/assets/meta.csv&quot;</span>, </span>
<span id="section-cb6-4"><a href="#section-cb6-4" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">&quot;data/meta.csv&quot;</span>, <span class="at">method=</span><span class="st">&#39;curl&#39;</span>, <span class="at">mode=</span><span class="st">&#39;wb&#39;</span>)</span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>Load data into R</li>
</ol>
<div class="sourceCode" id="section-cb7"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb7-1"><a href="#section-cb7-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;~/Desktop/analysis-user-group/20250210_Tidyverse/data/data.csv&quot;</span>)</span>
<span id="section-cb7-2"><a href="#section-cb7-2" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;~/Desktop/analysis-user-group/20250210_Tidyverse/data/meta.csv&quot;</span>)</span></code></pre></div>
</details>
<p><wimr></p>
<div id="section-data-wrangling-with-dplyr"
class="section level2 tabset tabset-fade">
<h2 class="tabset tabset-fade">Data Wrangling with dplyr</h2>
<p>First, we can quickly explore the data.</p>
<div class="sourceCode" id="section-cb8"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb8-1"><a href="#section-cb8-1" tabindex="-1"></a><span class="fu">summary</span>(data)</span></code></pre></div>
<p>We have 5 categorical columns and 8 numeric values.</p>
<ul>
<li>CD3, CD8, CD4, HLADR and CCR5 appear to be cell counts.<br />
</li>
<li>CD28 appears to be as a percentage already.</li>
</ul>
<div class="sourceCode" id="section-cb9"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb9-1"><a href="#section-cb9-1" tabindex="-1"></a><span class="fu">View</span>(data)</span></code></pre></div>
<details>
<summary>
Dropdown for interactive session
</summary>
<p>You can only use this in your personal sessions.</p>
<div class="sourceCode" id="section-cb10"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb10-1"><a href="#section-cb10-1" tabindex="-1"></a><span class="fu">userCode</span>(<span class="at">title=</span><span class="st">&quot;Explore the Data yourself&quot;</span>, </span>
<span id="section-cb10-2"><a href="#section-cb10-2" tabindex="-1"></a>        <span class="at">qvalue=</span><span class="st">&quot;table(data$tissue, data$layer)&quot;</span>, </span>
<span id="section-cb10-3"><a href="#section-cb10-3" tabindex="-1"></a>        <span class="at">nrows=</span><span class="dv">1</span>, <span class="at">out_height=</span><span class="dv">350</span>)</span></code></pre></div>
<iframe data-deferred-src="app85c91f51b9f6f54808e3d9fa12a5ebeb/?w=&amp;__subapp__=1" width="100%" height="350" class="shiny-frame shiny-frame-deferred"></iframe>
</details>
<br>
<hr>
<p>This immediately reveals some issues where our data needs tidying,
but we’ll cover these in the next workshop.</p>
<p><wimr></p>
<div id="section-filtering-selecting-and-mutating-data-dplyr."
class="section level3">
<h3>Filtering, Selecting, and Mutating data: <code>dplyr()</code>.</h3>
<p>The <code>dplyr</code> package is a powerful tool for data
manipulation in R. It provides a set of functions that make it easy to
filter, arrange, group, and summarize data. Some of the most commonly
used functions in <code>dplyr</code> are:</p>
<div class="sourceCode" id="section-cb11"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb11-1"><a href="#section-cb11-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span> <span class="co"># this is the pipe. I like to think of pipes as &quot;using this&quot;</span></span>
<span id="section-cb11-2"><a href="#section-cb11-2" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="section-cb11-3"><a href="#section-cb11-3" tabindex="-1"></a>  <span class="fu">print</span>()</span></code></pre></div>
<pre><code>  experiment  donor  tissue      layer group   CD3   CD8   CD4 HLADR  CCR5
1       Exp1 Donor1 Abdomen Epithelium   A_E  4460  2205  2184   882  1932
2       Exp1 Donor1 Abdomen Underlying  A_UM 15295  7960  5529   808  3916
3       Exp2 Donor2 Abdomen Epithelium   A_E 49774 25149 25251 10671 22763
4       Exp2 Donor2 Abdomen Underlying  A_UM 68879 47729 33616  5662 27725
  HLADR_MFI CCR5_MFI CD28
1  652.0000 4455.553 29.6
2  471.4750 7121.196 59.8
3  652.0000 3666.176 72.8
4  206.1183 2025.143 89.2</code></pre>
<ul>
<li><code>filter()</code>: to filter rows based on a condition</li>
</ul>
<div class="sourceCode" id="section-cb13"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb13-1"><a href="#section-cb13-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span></span>
<span id="section-cb13-2"><a href="#section-cb13-2" tabindex="-1"></a>  <span class="fu">filter</span>(layer <span class="sc">==</span> <span class="st">&quot;Epithelium&quot;</span>)<span class="sc">%&gt;%</span></span>
<span id="section-cb13-3"><a href="#section-cb13-3" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="section-cb13-4"><a href="#section-cb13-4" tabindex="-1"></a>  <span class="fu">print</span>()</span></code></pre></div>
<pre><code>  experiment  donor  tissue      layer group    CD3   CD8   CD4 HLADR  CCR5
1       Exp1 Donor1 Abdomen Epithelium   A_E   4460  2205  2184   882  1932
2       Exp2 Donor2 Abdomen Epithelium   A_E  49774 25149 25251 10671 22763
3       Exp2 Donor3 Abdomen Epithelium   A_E 208665 92410 84610 26531 68033
4       Exp3 Donor4 Abdomen Epithelium   A_E   4460  2272  2293   985  2082
  HLADR_MFI CCR5_MFI CD28
1       652 4455.553 29.6
2       652 3666.176 72.8
3       858 7065.873 62.0
4       858 9238.000 65.6</code></pre>
<ul>
<li><code>arrange()</code>: to reorder rows</li>
</ul>
<div class="sourceCode" id="section-cb15"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb15-1"><a href="#section-cb15-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span> </span>
<span id="section-cb15-2"><a href="#section-cb15-2" tabindex="-1"></a>  <span class="fu">arrange</span>(experiment) <span class="sc">%&gt;%</span></span>
<span id="section-cb15-3"><a href="#section-cb15-3" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="section-cb15-4"><a href="#section-cb15-4" tabindex="-1"></a>  <span class="fu">print</span>()</span></code></pre></div>
<pre><code>  experiment   donor  tissue      layer group   CD3   CD8   CD4 HLADR  CCR5
1       Exp1  Donor1 Abdomen Epithelium   A_E  4460  2205  2184   882  1932
2       Exp1  Donor1 Abdomen Underlying  A_UM 15295  7960  5529   808  3916
3       Exp1 Donor16  Vagina Epithelium   V_E 30109 17011  6596   263  6264
4       Exp1 Donor16  Vagina Underlying  V_UM 30774 16713 22180  3245 18958
  HLADR_MFI CCR5_MFI CD28
1  652.0000 4455.553 29.6
2  471.4750 7121.196 59.8
3   62.1126 3078.970 52.6
4  192.5019 3534.601  0.0</code></pre>
<ul>
<li><code>select()</code>: to select columns</li>
</ul>
<div class="sourceCode" id="section-cb17"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb17-1"><a href="#section-cb17-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span></span>
<span id="section-cb17-2"><a href="#section-cb17-2" tabindex="-1"></a>  <span class="fu">filter</span>(layer <span class="sc">==</span> <span class="st">&quot;Epithelium&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="section-cb17-3"><a href="#section-cb17-3" tabindex="-1"></a>  <span class="fu">arrange</span>(CD4) <span class="sc">%&gt;%</span></span>
<span id="section-cb17-4"><a href="#section-cb17-4" tabindex="-1"></a>  <span class="fu">select</span>(CD4) <span class="sc">%&gt;%</span></span>
<span id="section-cb17-5"><a href="#section-cb17-5" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="section-cb17-6"><a href="#section-cb17-6" tabindex="-1"></a>  <span class="fu">print</span>()</span></code></pre></div>
<pre><code>  CD4
1 262
2 472
3 596
4 596</code></pre>
<ul>
<li><code>mutate()</code>: to create new columns.</li>
</ul>
<div class="sourceCode" id="section-cb19"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb19-1"><a href="#section-cb19-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span></span>
<span id="section-cb19-2"><a href="#section-cb19-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">CD4_percent =</span> <span class="dv">100</span><span class="sc">*</span>CD4<span class="sc">/</span>CD3) <span class="sc">%&gt;%</span> </span>
<span id="section-cb19-3"><a href="#section-cb19-3" tabindex="-1"></a>  <span class="fu">select</span>(CD4_percent, CD3, CD4) <span class="sc">%&gt;%</span></span>
<span id="section-cb19-4"><a href="#section-cb19-4" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="section-cb19-5"><a href="#section-cb19-5" tabindex="-1"></a>  <span class="fu">print</span>()</span></code></pre></div>
<pre><code>  CD4_percent   CD3   CD4
1    48.96861  4460  2184
2    36.14907 15295  5529
3    50.73131 49774 25251
4    48.80443 68879 33616</code></pre>
<p><wimr></p>
</div>
<div id="section-grouping-and-summarising" class="section level3">
<h3>Grouping and Summarising</h3>
<p>Grouping and summarising in the Tidyverse allows you to split your
data into subsets using <code>group_by()</code> and then compute
aggregate statistics for each subgroup with <code>summarise()</code>.
This approach enables quick, clear insights into trends and differences
within your data by reducing complex datasets to meaningful
summaries.</p>
<div class="sourceCode" id="section-cb21"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb21-1"><a href="#section-cb21-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span></span>
<span id="section-cb21-2"><a href="#section-cb21-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">percent_CD4 =</span> <span class="dv">100</span><span class="sc">*</span>CD4<span class="sc">/</span>CD3) <span class="sc">%&gt;%</span></span>
<span id="section-cb21-3"><a href="#section-cb21-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(tissue, percent_CD4, <span class="at">fill=</span>layer))<span class="sc">+</span><span class="fu">geom_boxplot</span>()</span></code></pre></div>
<p><img src="figure/202502_tidyverse2.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-14-1">
Past versions of unnamed-chunk-14-1.png
</button>
</p>
<div id="section-fig-unnamed-chunk-14-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DrThomasOneil/analysis-user-group/blob/5fe30de6735f75be6fe53a89cb092266f13ed73e/docs/figure/202502_tidyverse2.Rmd/unnamed-chunk-14-1.png" target="_blank">5fe30de</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-02-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="sourceCode" id="section-cb22"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb22-1"><a href="#section-cb22-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span></span>
<span id="section-cb22-2"><a href="#section-cb22-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">percent_CD4 =</span> <span class="dv">100</span><span class="sc">*</span>CD4<span class="sc">/</span>CD3) <span class="sc">%&gt;%</span></span>
<span id="section-cb22-3"><a href="#section-cb22-3" tabindex="-1"></a>  <span class="fu">group_by</span>(tissue, layer) <span class="sc">%&gt;%</span></span>
<span id="section-cb22-4"><a href="#section-cb22-4" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_CD4 =</span> <span class="fu">mean</span>(percent_CD4), </span>
<span id="section-cb22-5"><a href="#section-cb22-5" tabindex="-1"></a>            <span class="at">median_CD4 =</span> <span class="fu">median</span>(percent_CD4)) <span class="sc">%&gt;%</span></span>
<span id="section-cb22-6"><a href="#section-cb22-6" tabindex="-1"></a>  <span class="fu">print</span>()</span></code></pre></div>
<pre><code>`summarise()` has grouped output by &#39;tissue&#39;. You can override using the
`.groups` argument.</code></pre>
<pre><code># A tibble: 9 × 4
# Groups:   tissue [5]
  tissue  layer      mean_CD4 median_CD4
  &lt;chr&gt;   &lt;chr&gt;         &lt;dbl&gt;      &lt;dbl&gt;
1 Abdo    Epithelium     37.7       37.7
2 Abdo    Underlying     35.5       35.5
3 Abdomen Epithelium     43.3       44.0
4 Abdomen Underlying     37.2       35.5
5 Labia   Epithelium     28.8       28.1
6 Labia   Underlying     65.9       66.3
7 Vagina  Epithelium     26.0       23.6
8 Vagina  Underlying     52.5       53.5
9 abdomen Underlying     21.3       21.3</code></pre>
<p><wimr></p>
</div>
<div id="section-joining-data-frames" class="section level3">
<h3>Joining Data Frames</h3>
<p>Combining data from multiple sources is often necessary, even when
the datasets don’t align perfectly. For instance, your long-format data
might include multiple entries per donor for tissues like epithelium and
mucosa, while the metadata contains donor-specific details that you
prefer not to duplicate.</p>
<p><read-time><i>This scenario is common in public single cell RNA
sequencing data, where you’ll integrate diverse layers of information
using shared keys.</read-time></p>
<div class="sourceCode" id="section-cb25"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb25-1"><a href="#section-cb25-1" tabindex="-1"></a><span class="co"># View data values of Donor1</span></span>
<span id="section-cb25-2"><a href="#section-cb25-2" tabindex="-1"></a>data[data<span class="sc">$</span>donor<span class="sc">==</span><span class="st">&quot;Donor1&quot;</span>,]</span></code></pre></div>
<pre><code>  experiment  donor  tissue      layer group   CD3  CD8  CD4 HLADR CCR5
1       Exp1 Donor1 Abdomen Epithelium   A_E  4460 2205 2184   882 1932
2       Exp1 Donor1 Abdomen Underlying  A_UM 15295 7960 5529   808 3916
  HLADR_MFI CCR5_MFI CD28
1   652.000 4455.553 29.6
2   471.475 7121.196 59.8</code></pre>
<div class="sourceCode" id="section-cb27"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb27-1"><a href="#section-cb27-1" tabindex="-1"></a><span class="co"># View data values of Donor1</span></span>
<span id="section-cb27-2"><a href="#section-cb27-2" tabindex="-1"></a>meta[meta<span class="sc">$</span>donor<span class="sc">==</span><span class="st">&quot;Donor1&quot;</span>,]</span></code></pre></div>
<pre><code>   donor experiment      date sex age clinical
1 Donor1       Exp1 19/2/2024   F  35  Healthy</code></pre>
<p>So we expect Donor 1 to be a <em>female</em>, aged <em>35</em> and
classified <em>Healthy</em></p>
<div class="sourceCode" id="section-cb29"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb29-1"><a href="#section-cb29-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span></span>
<span id="section-cb29-2"><a href="#section-cb29-2" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>experiment)<span class="sc">%&gt;%</span> <span class="co"># we can remove columns using select(-...)</span></span>
<span id="section-cb29-3"><a href="#section-cb29-3" tabindex="-1"></a>  <span class="fu">right_join</span>(meta, <span class="at">by=</span><span class="st">&quot;donor&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="section-cb29-4"><a href="#section-cb29-4" tabindex="-1"></a>  <span class="fu">filter</span>(donor <span class="sc">==</span> <span class="st">&quot;Donor1&quot;</span>)</span></code></pre></div>
<pre><code>   donor  tissue      layer group   CD3  CD8  CD4 HLADR CCR5 HLADR_MFI CCR5_MFI
1 Donor1 Abdomen Epithelium   A_E  4460 2205 2184   882 1932   652.000 4455.553
2 Donor1 Abdomen Underlying  A_UM 15295 7960 5529   808 3916   471.475 7121.196
  CD28 experiment      date sex age clinical
1 29.6       Exp1 19/2/2024   F  35  Healthy
2 59.8       Exp1 19/2/2024   F  35  Healthy</code></pre>
<p>With this, we wouldn’t need to save a combined dataframe to
analyse.</p>
<div class="sourceCode" id="section-cb31"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb31-1"><a href="#section-cb31-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span></span>
<span id="section-cb31-2"><a href="#section-cb31-2" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>experiment)<span class="sc">%&gt;%</span></span>
<span id="section-cb31-3"><a href="#section-cb31-3" tabindex="-1"></a>  <span class="fu">right_join</span>(meta, <span class="at">by=</span><span class="st">&quot;donor&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="section-cb31-4"><a href="#section-cb31-4" tabindex="-1"></a>  <span class="fu">filter</span>(tissue <span class="sc">!=</span> <span class="st">&quot;Abdo&quot;</span> <span class="sc">&amp;</span> tissue <span class="sc">!=</span> <span class="st">&#39;abdomen&#39;</span>) <span class="sc">%&gt;%</span> <span class="co"># remove errors</span></span>
<span id="section-cb31-5"><a href="#section-cb31-5" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">percent =</span> <span class="dv">100</span><span class="sc">*</span>CD4<span class="sc">/</span>CD3) <span class="sc">%&gt;%</span></span>
<span id="section-cb31-6"><a href="#section-cb31-6" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(clinical, percent, <span class="at">fill=</span>tissue))<span class="sc">+</span></span>
<span id="section-cb31-7"><a href="#section-cb31-7" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outliers=</span>F)</span></code></pre></div>
<p><img src="figure/202502_tidyverse2.Rmd/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-18-1">
Past versions of unnamed-chunk-18-1.png
</button>
</p>
<div id="section-fig-unnamed-chunk-18-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DrThomasOneil/analysis-user-group/blob/5fe30de6735f75be6fe53a89cb092266f13ed73e/docs/figure/202502_tidyverse2.Rmd/unnamed-chunk-18-1.png" target="_blank">5fe30de</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-02-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><wimr></p>
</div>
</div>
<div id="section-reshaping-data-with-tidyr" class="section level2">
<h2>Reshaping Data with tidyr</h2>
<p><code>tidyr</code> provides essential tools like
<code>pivot_longer()</code> and <code>pivot_wider()</code> to transform
your dataset between long and wide formats. This reshaping makes it
easier to align variables and observations for analysis, ensuring that
each variable forms a column and each observation a row. Such
flexibility is crucial when preparing complex experimental or single
cell RNA sequencing data for further analysis and visualisation.</p>
<p><strong>long format</strong><br />
In long format, each observation is represented by a single row, with
one column holding the categorical variable (e.g. the type of
measurement) and another column holding the corresponding values. This
format is particularly useful for generating boxplots or other
visualisations that compare distributions across groups.</p>
<p>For example, if you want to compare the percentages of CD4 and CD8
cells on the same plot, you can pivot the data longer:</p>
<div class="sourceCode" id="section-cb32"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb32-1"><a href="#section-cb32-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span> </span>
<span id="section-cb32-2"><a href="#section-cb32-2" tabindex="-1"></a>  <span class="fu">filter</span>(tissue <span class="sc">!=</span> <span class="st">&quot;Abdo&quot;</span> <span class="sc">&amp;</span> tissue <span class="sc">!=</span> <span class="st">&#39;abdomen&#39;</span>) <span class="sc">%&gt;%</span> <span class="co"># remove errors</span></span>
<span id="section-cb32-3"><a href="#section-cb32-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">CD4percent =</span> <span class="dv">100</span><span class="sc">*</span>CD4<span class="sc">/</span>CD3, </span>
<span id="section-cb32-4"><a href="#section-cb32-4" tabindex="-1"></a>         <span class="at">CD8percent =</span> <span class="dv">100</span><span class="sc">*</span>CD8<span class="sc">/</span>CD3) <span class="sc">%&gt;%</span></span>
<span id="section-cb32-5"><a href="#section-cb32-5" tabindex="-1"></a>  <span class="co"># pivot longer - select the columns, and the name of the columns for the names and values</span></span>
<span id="section-cb32-6"><a href="#section-cb32-6" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(CD4percent, CD8percent), </span>
<span id="section-cb32-7"><a href="#section-cb32-7" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">&quot;subset&quot;</span>, </span>
<span id="section-cb32-8"><a href="#section-cb32-8" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">&quot;percent&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="section-cb32-9"><a href="#section-cb32-9" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(experiment,group,CCR5_MFI,HLADR_MFI,CD3,CD8,CD4,HLADR,CCR5,CD28))<span class="sc">%&gt;%</span></span>
<span id="section-cb32-10"><a href="#section-cb32-10" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">8</span>) <span class="sc">%&gt;%</span></span>
<span id="section-cb32-11"><a href="#section-cb32-11" tabindex="-1"></a>  <span class="fu">print</span>()</span></code></pre></div>
<pre><code># A tibble: 8 × 5
  donor  tissue  layer      subset     percent
  &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;        &lt;dbl&gt;
1 Donor1 Abdomen Epithelium CD4percent    49.0
2 Donor1 Abdomen Epithelium CD8percent    49.4
3 Donor1 Abdomen Underlying CD4percent    36.1
4 Donor1 Abdomen Underlying CD8percent    52.0
5 Donor2 Abdomen Epithelium CD4percent    50.7
6 Donor2 Abdomen Epithelium CD8percent    50.5
7 Donor2 Abdomen Underlying CD4percent    48.8
8 Donor2 Abdomen Underlying CD8percent    69.3</code></pre>
<p>This code creates a new column <code>subset</code> that indicates
whether the value corresponds to <code>%CD4</code> or <code>%CD8</code>,
and a column <code>percent</code> for the computed values. The data is
now structured with one column for the measurement type and another for
the percentage, making it straightforward to plot:</p>
<div class="sourceCode" id="section-cb34"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb34-1"><a href="#section-cb34-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span> </span>
<span id="section-cb34-2"><a href="#section-cb34-2" tabindex="-1"></a>  <span class="fu">filter</span>(tissue <span class="sc">!=</span> <span class="st">&quot;Abdo&quot;</span> <span class="sc">&amp;</span> tissue <span class="sc">!=</span> <span class="st">&#39;abdomen&#39;</span>) <span class="sc">%&gt;%</span> <span class="co"># remove errors</span></span>
<span id="section-cb34-3"><a href="#section-cb34-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">CD4percent =</span> <span class="dv">100</span><span class="sc">*</span>CD4<span class="sc">/</span>CD3, </span>
<span id="section-cb34-4"><a href="#section-cb34-4" tabindex="-1"></a>         <span class="at">CD8percent =</span> <span class="dv">100</span><span class="sc">*</span>CD8<span class="sc">/</span>CD3) <span class="sc">%&gt;%</span></span>
<span id="section-cb34-5"><a href="#section-cb34-5" tabindex="-1"></a>  <span class="co"># pivot longer - select the columns, and the name of the columns for the names and values</span></span>
<span id="section-cb34-6"><a href="#section-cb34-6" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(CD4percent, CD8percent), </span>
<span id="section-cb34-7"><a href="#section-cb34-7" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">&quot;subset&quot;</span>, </span>
<span id="section-cb34-8"><a href="#section-cb34-8" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">&quot;percent&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="section-cb34-9"><a href="#section-cb34-9" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(layer,percent, <span class="at">fill=</span>subset))<span class="sc">+</span></span>
<span id="section-cb34-10"><a href="#section-cb34-10" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code></pre></div>
<p><img src="figure/202502_tidyverse2.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-20-1">
Past versions of unnamed-chunk-20-1.png
</button>
</p>
<div id="section-fig-unnamed-chunk-20-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DrThomasOneil/analysis-user-group/blob/5fe30de6735f75be6fe53a89cb092266f13ed73e/docs/figure/202502_tidyverse2.Rmd/unnamed-chunk-20-1.png" target="_blank">5fe30de</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-02-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><strong>wide format</strong><br />
Sometimes you want to examine relationships between measurements
directly—for example, to see if there’s a relationship between values
from different layers (e.g. Epithelium and Underlying mucosa). In wide
format, each type of measurement occupies its own column. This structure
is ideal for scatterplots or correlation analyses.</p>
<p>Consider this example where we pivot the data to have separate
columns for each layer:</p>
<div class="sourceCode" id="section-cb35"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb35-1"><a href="#section-cb35-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span> </span>
<span id="section-cb35-2"><a href="#section-cb35-2" tabindex="-1"></a>  <span class="fu">filter</span>(tissue <span class="sc">!=</span> <span class="st">&quot;Abdo&quot;</span> <span class="sc">&amp;</span> tissue <span class="sc">!=</span> <span class="st">&#39;abdomen&#39;</span>) <span class="sc">%&gt;%</span> <span class="co"># remove errors</span></span>
<span id="section-cb35-3"><a href="#section-cb35-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">percent =</span> <span class="dv">100</span><span class="sc">*</span>HLADR<span class="sc">/</span>CD4) <span class="sc">%&gt;%</span></span>
<span id="section-cb35-4"><a href="#section-cb35-4" tabindex="-1"></a>  <span class="fu">select</span>(donor,tissue,layer,percent) <span class="sc">%&gt;%</span></span>
<span id="section-cb35-5"><a href="#section-cb35-5" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> layer, <span class="at">values_from =</span> percent) <span class="sc">%&gt;%</span></span>
<span id="section-cb35-6"><a href="#section-cb35-6" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>Epithelium, <span class="at">y=</span>Underlying, <span class="at">color=</span>tissue))<span class="sc">+</span></span>
<span id="section-cb35-7"><a href="#section-cb35-7" tabindex="-1"></a>    <span class="fu">geom_point</span>()<span class="sc">+</span> </span>
<span id="section-cb35-8"><a href="#section-cb35-8" tabindex="-1"></a>    <span class="fu">stat_ellipse</span>()</span></code></pre></div>
<p><img src="figure/202502_tidyverse2.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-21-1">
Past versions of unnamed-chunk-21-1.png
</button>
</p>
<div id="section-fig-unnamed-chunk-21-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DrThomasOneil/analysis-user-group/blob/5fe30de6735f75be6fe53a89cb092266f13ed73e/docs/figure/202502_tidyverse2.Rmd/unnamed-chunk-21-1.png" target="_blank">5fe30de</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-02-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Here, the pivot_wider() function converts the long data into a wide
format where separate columns for Epithelium and Underlying values are
created. This allows you to directly plot and explore relationships
between these layers using scatterplots, with
<code>stat_ellipse()</code> adding confidence ellipses to highlight
group trends.</p>
<p>Both long and wide formats serve specific purposes in data analysis
and visualisation. Long format is flexible for creating grouped
comparisons, while wide format facilitates direct relationship analysis
between variables.</p>
<p><wimr></p>
</div>
<div id="section-conclusions" class="section level2">
<h2>Conclusions</h2>
<p>In conclusion, this tutorial has demonstrated how the
<strong>Tidyverse</strong> streamlines data analysis by providing a
coherent set of tools that simplify data manipulation, reshaping, and
visualisation. Using <code>dplyr</code>, we efficiently filtered,
mutated, grouped, summarised, and joined our experimental and metadata,
while the use of pipes (<code>%&gt;%</code>) allowed us to finalise
complex workflows in a readable and intuitive manner.</p>
<p>Additionally, <code>tidyr</code> enabled us to transform our data
between long and wide formats, ensuring that it is in the optimal
structure for analysis. With ggplot2, these tidy datasets were then
translated into compelling graphics that highlight key patterns and
relationships. Together, these tools empower analysts to create
reproducible and insightful workflows, ultimately enhancing the rigour
and clarity of data-driven research.</p>
<p><e></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="section-workflowr-sessioninfo" class="collapse">
<div class="sourceCode" id="section-cb36"><pre
class="sourceCode r"><code class="sourceCode r"><span id="section-cb36-1"><a href="#section-cb36-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>R version 4.4.0 (2024-04-24)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.3

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Sydney
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.4 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [5] purrr_1.0.4     readr_2.1.5     tidyr_1.3.1     tibble_3.2.1   
 [9] ggplot2_3.5.1   tidyverse_2.0.0 shiny_1.10.0    workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] utf8_1.2.4        sass_0.4.9        generics_0.1.3    stringi_1.8.4    
 [5] hms_1.1.3         digest_0.6.37     magrittr_2.0.3    timechange_0.3.0 
 [9] evaluate_1.0.3    grid_4.4.0        fastmap_1.2.0     rprojroot_2.0.4  
[13] jsonlite_1.8.9    processx_3.8.4    whisker_0.4.1     ps_1.8.1         
[17] promises_1.3.2    httr_1.4.7        scales_1.3.0      jquerylib_0.1.4  
[21] cli_3.6.3         rlang_1.1.5       munsell_0.5.1     withr_3.0.2      
[25] cachem_1.1.0      yaml_2.3.10       tools_4.4.0       tzdb_0.4.0       
[29] colorspace_2.1-1  httpuv_1.6.15     vctrs_0.6.5       R6_2.5.1         
[33] mime_0.12         lifecycle_1.0.4   git2r_0.35.0      fs_1.6.5         
[37] MASS_7.3-61       pkgconfig_2.0.3   callr_3.7.6       pillar_1.10.1    
[41] bslib_0.9.0       later_1.4.1       gtable_0.3.6      glue_1.8.0       
[45] Rcpp_1.0.14       xfun_0.50         tidyselect_1.2.1  rstudioapi_0.17.1
[49] knitr_1.49        farver_2.1.2      xtable_1.8-4      htmltools_0.5.8.1
[53] labeling_0.4.3    rmarkdown_2.29    compiler_4.4.0    getPass_0.2-4    </code></pre>
</div>
</div>
</div>

<div id="rmd-source-code">---
title: "Tidyverse 2: Data Manipulation"
author: "Thomas O'Neil"
date: "2025-02" 
output:
  html_document:
    code_folding: show
editor_options: 
  markdown: 
    wrap: 72
runtime: shiny
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, collapse=F, warning = F, error=F, root.dir =  getwd())
```

# Introduction{.tabset .tabset-fade}

In this tutorial, we explore the functionalities of the Tidyverse by working with two datasets:

- `data.csv` contains fabricated summary flow cytometry data, such as cell numbers and MFI
- `meta.csv` contains information regarding each donor, such as age and sex. 

The Tidyverse is a collection of R packages that share an underlying design philosophy and grammar, making data analysis more intuitive and coherent. Here are some key benefits:

**dplyr**:
Provides a suite of verbs (e.g. `filter()`, `mutate()`, `select(`), and `left_join()`) that simplify data manipulation. For example, `left_join()` easily combines experimental data with metadata based on a common key. See the [Introduction to R: Chapter 4](https://drthomasoneil.github.io/analysis-user-group/r-tutorial/_book/chapter-4-data-manipulation-with-tidyverse.html) for more content on dplyr. 

**Pipes (%>%)**:
Enhance code readability by allowing you to chain multiple operations in a sequential, natural-language style. This means you can finalise complex data transformations in a single, clear pipeline without the need for numerous intermediate variables.

**tidyr**:
Focuses on tidying data, ensuring that each observation occupies a single row and each variable a single column. Functions like `pivot_longer()` and pivot_wider() help reshape data into a standard format, which is essential for further analysis.

**ggplot2**:
Offers a powerful and flexible grammar for data visualisation. With ggplot2, you can create customised plots that effectively communicate insights from your data, building on the clean, tidy data produced by the other Tidyverse packages. See the [Introduction to R: Chapter 5-6](https://drthomasoneil.github.io/analysis-user-group/r-tutorial/_book/chapter-5-data-visualisation-with-ggplot2.html) and the [previous workshop](https://drthomasoneil.github.io/analysis-user-group/202502_tidyverse1.html) by Harry for extra information on ggplot. 

<div class="hint-goals">
**Goals:** Learn data manipulation with Tidyverse<br>
&nbsp;&nbsp;&#8226; Tidy Data<br> 
&nbsp;&nbsp;&#8226; Select and Mutate data<br> 
&nbsp;&nbsp;&#8226; Group and summarise data<br> 
&nbsp;&nbsp;&#8226; Join two data frames<br> 
&nbsp;&nbsp;&#8226; Pivot data with `tidyr`<br> 

</div>

Whether you are combining datasets, reshaping data, or creating compelling graphics, the Tidyverse offers a consistent and powerful set of tools to support your analysis.

<read-time><i>**Estimated length**: 1 Hour</i></read-time><wimr>

# Set up

<details><summary>**Follow these instructions to get started:**</summary>
<br><br>
1. Save this script in a folder called **"202502_Tidyverse2"**

2. Install and load the relevant packages and functions

```{r, eval=F}
install.packages("tidyverse")
install.packages("shiny")
install.packages("readr")
```

```{r}
suppressWarnings(library(shiny))
suppressWarnings(library(tidyverse))
theme_set(theme_classic())
source(url("https://raw.githubusercontent.com/DrThomasOneil/SkillTrees/refs/heads/main/assets/.functions.R")) 
```

2. Create new folders in your directory.

```{r, eval=F}
dir.create("plots")
dir.create("data")
```

3. Download the data.

```{r, eval=F}
download.file(url="https://raw.githubusercontent.com/DrThomasOneil/analysis-user-group/refs/heads/main/docs/r-tutorial/assets/synthetic_data.csv", 
              destfile = "data/data.csv", method='curl', mode='wb')
download.file(url="https://raw.githubusercontent.com/DrThomasOneil/analysis-user-group/refs/heads/main/docs/r-tutorial/assets/meta.csv", 
              destfile = "data/meta.csv", method='curl', mode='wb')
```

4. Load data into R

```{r}
data <- read.csv("~/Desktop/analysis-user-group/20250210_Tidyverse/data/data.csv")
meta <- read.csv("~/Desktop/analysis-user-group/20250210_Tidyverse/data/meta.csv")
```

</details>

<wimr>

## Data Wrangling with dplyr{.tabset .tabset-fade}

First, we can quickly explore the data. 

```{r, eval=F}
summary(data)
```

We have 5 categorical columns and 8 numeric values. 

- CD3, CD8, CD4, HLADR and CCR5 appear to be cell counts.   
- CD28 appears to be as a percentage already.

```{r, eval=F}
View(data)
```

<details><summary>Dropdown for interactive session</summary>

You can only use this in your personal sessions.
```{r, eval=T}
userCode(title="Explore the Data yourself", 
        qvalue="table(data$tissue, data$layer)", 
        nrows=1, out_height=350)
```

</details><br><hr>

This immediately reveals some issues where our data needs tidying, but we'll cover these in the next workshop. 

<wimr>

### Filtering, Selecting, and Mutating data: `dplyr()`.

The `dplyr` package is a powerful tool for data manipulation in R. It
provides a set of functions that make it easy to filter, arrange, group,
and summarize data. Some of the most commonly used functions in `dplyr`
are:    

```{r}
data %>% # this is the pipe. I like to think of pipes as "using this"
  head(4) %>%
  print()
```

- `filter()`: to filter rows based on a condition   

```{r}
data %>%
  filter(layer == "Epithelium")%>%
  head(4) %>%
  print()
```

- `arrange()`: to reorder rows    

```{r}
data %>% 
  arrange(experiment) %>%
  head(4) %>%
  print()
```

- `select()`: to select columns   

```{r}
data %>%
  filter(layer == "Epithelium") %>%
  arrange(CD4) %>%
  select(CD4) %>%
  head(4) %>%
  print()
```

- `mutate()`: to create new columns.

```{r}
data %>%
  mutate(CD4_percent = 100*CD4/CD3) %>% 
  select(CD4_percent, CD3, CD4) %>%
  head(4) %>%
  print()
```

<wimr>

### Grouping and Summarising

Grouping and summarising in the Tidyverse allows you to split your data into subsets using `group_by()` and then compute aggregate statistics for each subgroup with `summarise()`. This approach enables quick, clear insights into trends and differences within your data by reducing complex datasets to meaningful summaries.

```{r}
data %>%
  mutate(percent_CD4 = 100*CD4/CD3) %>%
  ggplot(aes(tissue, percent_CD4, fill=layer))+geom_boxplot()
```

```{r}
data %>%
  mutate(percent_CD4 = 100*CD4/CD3) %>%
  group_by(tissue, layer) %>%
  summarise(mean_CD4 = mean(percent_CD4), 
            median_CD4 = median(percent_CD4)) %>%
  print()
```

<wimr>

### Joining Data Frames

Combining data from multiple sources is often necessary, even when the datasets don’t align perfectly. For instance, your long-format data might include multiple entries per donor for tissues like epithelium and mucosa, while the metadata contains donor-specific details that you prefer not to duplicate. 

<read-time><i>This scenario is common in public single cell RNA sequencing data, where you'll integrate diverse layers of information using shared keys.</read-time>

```{r}
# View data values of Donor1
data[data$donor=="Donor1",]

# View data values of Donor1
meta[meta$donor=="Donor1",]
```

So we expect Donor 1 to be a *female*, aged *35* and classified *Healthy*

```{r}
data %>%
  select(-experiment)%>% # we can remove columns using select(-...)
  right_join(meta, by="donor") %>%
  filter(donor == "Donor1")
```

With this, we wouldn't need to save a combined dataframe to analyse. 

```{r}
data %>%
  select(-experiment)%>%
  right_join(meta, by="donor") %>%
  filter(tissue != "Abdo" & tissue != 'abdomen') %>% # remove errors
  mutate(percent = 100*CD4/CD3) %>%
  ggplot(aes(clinical, percent, fill=tissue))+
  geom_boxplot(outliers=F)
```

<wimr>

## Reshaping Data with tidyr

`tidyr` provides essential tools like `pivot_longer()` and `pivot_wider()` to transform your dataset between long and wide formats. This reshaping makes it easier to align variables and observations for analysis, ensuring that each variable forms a column and each observation a row. Such flexibility is crucial when preparing complex experimental or single cell RNA sequencing data for further analysis and visualisation. 

**long format**   
In long format, each observation is represented by a single row, with one column holding the categorical variable (e.g. the type of measurement) and another column holding the corresponding values. This format is particularly useful for generating boxplots or other visualisations that compare distributions across groups.

For example, if you want to compare the percentages of CD4 and CD8 cells on the same plot, you can pivot the data longer:

```{r}
data %>% 
  filter(tissue != "Abdo" & tissue != 'abdomen') %>% # remove errors
  mutate(CD4percent = 100*CD4/CD3, 
         CD8percent = 100*CD8/CD3) %>%
  # pivot longer - select the columns, and the name of the columns for the names and values
  pivot_longer(cols = c(CD4percent, CD8percent), 
               names_to = "subset", 
               values_to = "percent") %>%
  select(-c(experiment,group,CCR5_MFI,HLADR_MFI,CD3,CD8,CD4,HLADR,CCR5,CD28))%>%
  head(8) %>%
  print()
```

This code creates a new column `subset` that indicates whether the value corresponds to `%CD4` or `%CD8`, and a column `percent` for the computed values. The data is now structured with one column for the measurement type and another for the percentage, making it straightforward to plot:

```{r}
data %>% 
  filter(tissue != "Abdo" & tissue != 'abdomen') %>% # remove errors
  mutate(CD4percent = 100*CD4/CD3, 
         CD8percent = 100*CD8/CD3) %>%
  # pivot longer - select the columns, and the name of the columns for the names and values
  pivot_longer(cols = c(CD4percent, CD8percent), 
               names_to = "subset", 
               values_to = "percent") %>%
  ggplot(aes(layer,percent, fill=subset))+
  geom_boxplot()
```

**wide format**   
Sometimes you want to examine relationships between measurements directly—for example, to see if there's a relationship between values from different layers (e.g. Epithelium and Underlying mucosa). In wide format, each type of measurement occupies its own column. This structure is ideal for scatterplots or correlation analyses.

Consider this example where we pivot the data to have separate columns for each layer:

```{r}
data %>% 
  filter(tissue != "Abdo" & tissue != 'abdomen') %>% # remove errors
  mutate(percent = 100*HLADR/CD4) %>%
  select(donor,tissue,layer,percent) %>%
  pivot_wider(names_from = layer, values_from = percent) %>%
  ggplot(aes(x=Epithelium, y=Underlying, color=tissue))+
    geom_point()+ 
    stat_ellipse()

```

Here, the pivot_wider() function converts the long data into a wide format where separate columns for Epithelium and Underlying values are created. This allows you to directly plot and explore relationships between these layers using scatterplots, with `stat_ellipse()` adding confidence ellipses to highlight group trends.

Both long and wide formats serve specific purposes in data analysis and visualisation. Long format is flexible for creating grouped comparisons, while wide format facilitates direct relationship analysis between variables.

<wimr>

## Conclusions

In conclusion, this tutorial has demonstrated how the **Tidyverse** streamlines data analysis by providing a coherent set of tools that simplify data manipulation, reshaping, and visualisation. Using `dplyr`, we efficiently filtered, mutated, grouped, summarised, and joined our experimental and metadata, while the use of pipes (`%>%`) allowed us to finalise complex workflows in a readable and intuitive manner.

Additionally, `tidyr` enabled us to transform our data between long and wide formats, ensuring that it is in the optimal structure for analysis. With ggplot2, these tidy datasets were then translated into compelling graphics that highlight key patterns and relationships. Together, these tools empower analysts to create reproducible and insightful workflows, ultimately enhancing the rigour and clarity of data-driven research.

<e>




</div>
<div class="custom-footer">
    <div class="footer-left" style="width: 30%; float: centre">
      <img src="assets/footer/footer1.png" alt="wimr-logo">
    </div>
    <div class="footer-center">
      <p><strong>Creators</strong><br>
        Thomas O'Neil<br>
        Harry Robertson<br>
        <a href="mailto:thomas.oneil@sydney.edu.au"><strong>Contact us</strong></a>
      </p>
    </div>
    <div class="footer-right" style="width: 30%; float: centre">
      <img src="assets/footer/footer2.png" alt="usyd-logo">
    </div>
  </div>
  <div class="footer-updated">
    <p>This site was last updated</p>
    <p><script>document.write(new Date().toISOString().slice(0, 10));</script></p>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("section-TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("202502_tidyverse2.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#section-TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
