<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Author" />


<title>Introduction to Visium Analysis</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="icon" type="image/png" href="assets/logo/DRSN_logo_smaller_filt.png">

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; } /* Alert */
code span.an { color: #008000; } /* Annotation */
code span.at { } /* Attribute */
code span.bu { } /* BuiltIn */
code span.cf { color: #0000ff; } /* ControlFlow */
code span.ch { color: #008080; } /* Char */
code span.cn { } /* Constant */
code span.co { color: #008000; } /* Comment */
code span.cv { color: #008000; } /* CommentVar */
code span.do { color: #008000; } /* Documentation */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.im { } /* Import */
code span.in { color: #008000; } /* Information */
code span.kw { color: #0000ff; } /* Keyword */
code span.op { } /* Operator */
code span.ot { color: #ff4000; } /* Other */
code span.pp { color: #ff4000; } /* Preprocessor */
code span.sc { color: #008080; } /* SpecialChar */
code span.ss { color: #008080; } /* SpecialString */
code span.st { color: #008080; } /* String */
code span.va { } /* Variable */
code span.vs { color: #008080; } /* VerbatimString */
code span.wa { color: #008000; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>


<link rel="stylesheet" href="adit/custom.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src='assets/logo/DRSN_logo_smaller.png' style='height:30px; vertical-align:middle; align-items:center; margin-right:10px;'></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="0_schedule.html">Schedule</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Introduction Series
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="202502_tidyverse1.html">Tidyverse 1: Intro to Plotting</a>
    </li>
    <li>
      <a href="202502_tidyverse2.html">Tidyverse 2: Intro to Manipulation</a>
    </li>
    <li>
      <a href="202503_tidyverse3.html">Tidyverse 3: Creating a Reproducible Report</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Events
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="202504.html">April: Flow Cytometry</a>
    </li>
    <li>
      <a href="202505.html">May: Spatial Transcriptomics</a>
    </li>
    <li>
      <a href="202506.html">June: Bring Your Own Data</a>
    </li>
    <li>
      <a href="202507.html">July: TBA</a>
    </li>
    <li>
      <a href="202508.html">August: TBA</a>
    </li>
    <li>
      <a href="202509.html">September: TBA</a>
    </li>
    <li>
      <a href="2025010.html">October: TBA</a>
    </li>
    <li>
      <a href="2025011.html">November: TBA</a>
    </li>
    <li>
      <a href="2025012.html">December: TBA</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Contact Us!
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="0_contacts.html">Contact Us!</a>
    </li>
    <li>
      <a href="0_contacts_committee.html">The Committee</a>
    </li>
    <li>
      <a href="0_contacts_feedback.html">Feedback Form</a>
    </li>
  </ul>
</li>
<li>
  <a href="0_archive.html">Archive</a>
</li>
<li>
  <a href="0_resources.html">Extra Resources</a>
</li>
<li>
  <a href="https://github.com/DrThomasOneil/analysis-user-group">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
<li role="separator" class="divider"></li>
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Introduction to Visium Analysis</h1>
<h4 class="author">Author</h4>
<h4 class="date">2025-02</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-05-12
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong>
<code>DigitalResearchSkillsNetwork/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1337code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1337)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1337code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1337)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomDrThomasOneilDigitalResearchSkillsNetworktreef349cafbbc6959115d7f63fea1f23c977a2907e4targetblankf349cafa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/tree/f349cafbbc6959115d7f63fea1f23c977a2907e4" target="_blank">f349caf</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomDrThomasOneilDigitalResearchSkillsNetworktreef349cafbbc6959115d7f63fea1f23c977a2907e4targetblankf349cafa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/tree/f349cafbbc6959115d7f63fea1f23c977a2907e4" target="_blank">f349caf</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/.RData
    Ignored:    analysis/.Rhistory
    Ignored:    analysis/adit/.DS_Store
    Ignored:    plots/
    Ignored:    raw/

Unstaged changes:
    Modified:   analysis/202505.Rmd
    Modified:   analysis/202505_InFocus.Rmd
    Modified:   analysis/202505_Seminar.Rmd
    Modified:   workflow.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/202505_Workshop.Rmd</code>) and
HTML (<code>docs/202505_Workshop.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/blob/f349cafbbc6959115d7f63fea1f23c977a2907e4/analysis/202505_Workshop.Rmd" target="_blank">f349caf</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-05-12
</td>
<td>
wflow_publish("analysis/202505_Workshop.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/cff79238bf16b8671c9c52ee5e87aa7cf2504018/docs/202505_Workshop.html" target="_blank">cff7923</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-05-07
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/blob/ff95f8fe1f24a1144fb4d0c7581e92ee527f4643/analysis/202505_Workshop.Rmd" target="_blank">ff95f8f</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-05-07
</td>
<td>
wflow_publish(files)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/16a96138fdcae45a20a992a3dfe2e0ba9bbbc20e/docs/202505_Workshop.html" target="_blank">16a9613</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-04-30
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DrThomasOneil/Digital-Research-Skills-Network/862d6b6d567bdbd4bbceec6c50cd780dbdb57b29/docs/202505_Workshop.html" target="_blank">862d6b6</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-04-29
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DrThomasOneil/Digital-Research-Skills-Network/blob/b68287f08cfd472eb7e732d5a1bf6e0391b9ab7d/analysis/202505_Workshop.Rmd" target="_blank">b68287f</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2025-04-29
</td>
<td>
wflow_publish(files)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div style="text-align: center">
<p><read-time><span style="font-size: 1.3em">Please try to follow the
instructions below <strong>prior</strong> to attending the
workshop.<br><br><i><strong>Be aware!</strong></i></span><br> It is
difficult to debug over zoom. <br>If you want the best feedback and
experience, make sure you attend the workshop in person!</read-time></p>
<p>You can download the script on the top right or from <a
href="https://github.com/DrThomasOneil/digital-research-skills-network/blob/master/analysis/202504.Rmd">this
link</a>.</p>
</div>
<p><wimr></p>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
</div>
<div id="setup" class="section level1 tabset tabset-fade">
<h1 class="tabset tabset-fade">Setup</h1>
<ul>
<li><p>Create folders</p></li>
<li><p>Install Packages</p></li>
<li><p>Download Data</p></li>
</ul>
<div id="create-folders" class="section level2">
<h2>Create Folders</h2>
<pre class="text"><code>├── YourFolder
│     └── raw (where we can deposit raw downloaded data)
│     └── data (where we can process data and store it)
│     └── plots (where we can store plots)</code></pre>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="st">&quot;raw&quot;</span>)){<span class="fu">dir.create</span>(<span class="st">&quot;raw&quot;</span>)}</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="st">&quot;data&quot;</span>)){<span class="fu">dir.create</span>(<span class="st">&quot;data&quot;</span>)}</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="st">&quot;plots&quot;</span>)){<span class="fu">dir.create</span>(<span class="st">&quot;plots&quot;</span>)}</span></code></pre></div>
<p><wimr></p>
</div>
<div id="install-packages" class="section level2">
<h2>Install Packages</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;Seurat&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;tidyverse&quot;</span>)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;cowplot&quot;</span>)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;GEOquery&quot;</span>)</span></code></pre></div>
<p>We have a function that will let you check the setup</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;https://github.com/DrThomasOneil/Digital-Research-Skills-Network/raw/refs/heads/main/docs/adit/checksetup.R&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">checkSetup</span>(</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="at">cran_packages =</span> <span class="fu">c</span>(<span class="st">&quot;Seurat&quot;</span>, <span class="st">&quot;tidyverse&quot;</span>, <span class="st">&quot;cowplot&quot;</span>), </span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="at">bioc_packages=</span><span class="fu">c</span>(<span class="st">&quot;GEOquery&quot;</span>) </span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>)</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="sc">--------------------------------------</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="er">***</span>Installing General Packages<span class="sc">**</span><span class="er">*</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>Seurat is loaded<span class="sc">!</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>tidyverse is loaded<span class="sc">!</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>cowplot is loaded<span class="sc">!</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a>GEOquery is loaded<span class="sc">!</span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a><span class="sc">--------------------------------------</span></span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a> </span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a> All packages are loaded<span class="sc">!</span></span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a></span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a> Happy Coding<span class="sc">!</span> <span class="er">:)</span></span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1337</span>)</span></code></pre></div>
<p><wimr></p>
</div>
<div id="download-data" class="section level2">
<h2>Download data</h2>
<p>We will download the data directly from the <a
href="https://www.ncbi.nlm.nih.gov/geo/">GEO</a>. We’ll use <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE290350">this</a>
dataset, which was <a
href="https://pmc.ncbi.nlm.nih.gov/articles/PMC11646855/">published in
2024</a>. Additionally, this group have published <a
href="https://github.com/vildeka/Spatial_DMPA?tab=readme-ov-file">all of
their methods and scripts for analysis</a>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Go to the GEO page and right click on the http link under download. It should look like this</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>data_file <span class="ot">&lt;-</span> <span class="st">&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE290350&amp;format=file&quot;</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co"># download the data</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(<span class="st">&quot;raw/GSE290350_RAW.tar&quot;</span>)){<span class="fu">download.file</span>(<span class="at">url=</span>data_file, <span class="at">destfile =</span> <span class="st">&quot;raw/GSE290350_RAW.tar&quot;</span>, <span class="at">method=</span><span class="st">&#39;curl&#39;</span>)}</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co"># repeat for metadata</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(<span class="st">&quot;raw/GSE290350_metadata.csv.gz&quot;</span>)){<span class="fu">download.file</span>(<span class="at">url=</span><span class="st">&#39;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE290350&amp;format=file&amp;file=GSE290350%5Fmetadata%2Ecsv%2Egz&#39;</span>, <span class="at">destfile =</span> <span class="st">&quot;raw/GSE290350_metadata.csv.gz&quot;</span>, <span class="at">method=</span><span class="st">&#39;curl&#39;</span>)}</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co"># and the supposed processed Seurat data</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(<span class="st">&quot;raw/GSE290350_seuratObj_spatial_dist.RDS&quot;</span>)){<span class="fu">download.file</span>(<span class="at">url=</span><span class="st">&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE290350&amp;format=file&amp;file=GSE290350%5FseuratObj%5Fspatial%5Fdist%2ERDS&quot;</span>, <span class="at">destfile =</span> <span class="st">&quot;raw/GSE290350_seuratObj_spatial_dist.RDS&quot;</span>, <span class="at">method=</span><span class="st">&#39;curl&#39;</span>)}</span></code></pre></div>
<p><wimr></p>
</div>
<div id="unzip-and-untar-the-raw-data" class="section level2">
<h2>Unzip and untar the raw data</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">untar</span>(<span class="st">&quot;raw/GSE290350_RAW.tar&quot;</span>, <span class="at">exdir =</span> <span class="st">&quot;../raw/GSE290350_RAW&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">&quot;raw/GSE290350_RAW&quot;</span>, <span class="at">pattern =</span> <span class="st">&quot;</span><span class="sc">\\</span><span class="st">.tar</span><span class="sc">\\</span><span class="st">.gz$&quot;</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(samples)) {</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>  tar_path <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">&quot;</span><span class="sc">\\</span><span class="st">.gz$&quot;</span>, <span class="st">&quot;&quot;</span>, samples[i])</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>  <span class="co"># Unzip the .tar.gz file</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>  GEOquery<span class="sc">::</span><span class="fu">gunzip</span>(samples[i], <span class="at">destname =</span> tar_path, <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>  output_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">&quot;raw/GSE290350_RAW&quot;</span>, <span class="fu">sub</span>(<span class="st">&quot;</span><span class="sc">\\</span><span class="st">.tar$&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="fu">basename</span>(tar_path)))</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>  <span class="fu">dir.create</span>(output_dir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>  <span class="co"># Extract tar to output directory</span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a>  <span class="fu">untar</span>(tar_path, <span class="at">exdir =</span> output_dir)</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a>}</span></code></pre></div>
<p><wimr></p>
</div>
</div>
<div id="pre-processing" class="section level1 tabset tabset-fade">
<h1 class="tabset tabset-fade">Pre-processing</h1>
<p>Here we’ll load in the data. Fortunately, this data is quite neatly
organised. When present, we can load the <code>.h5</code> object. Please
see the <code>Archive</code> page of the <a
href="https://digitalresearchskills.network/0_archive.html">website</a>
to find vignettes on downloading and installing GEO-sourced objects that
are not as well-organised!</p>
<p>We will:</p>
<ul>
<li><p>Load the data</p></li>
<li><p>Add the meta data</p></li>
<li><p>Check the QC, filter and process the data</p></li>
</ul>
<div>
<div id="load-in-the-data" class="section level2">
<h2>Load in the data</h2>
<p>We’ll load in the supplied metadata, and then we can load 10X data
quite easily using the Load10X_Spatial function, which requires a
certain format. We can load in the filtered matrix, which is filtered
based on whether the initial processing detected that that spot aligned
with tissue. Alternatively, you can load the raw object, and filter as
you wish.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;raw/GSE290350_metadata.csv.gz&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>  <span class="at">data.dir=</span><span class="st">&quot;raw/GSE290350_RAW/GSM8811056_P004/P004&quot;</span>,</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="at">filename =</span> <span class="st">&quot;filtered_feature_bc_matrix.h5&quot;</span>,</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="at">filter.matrix =</span> F</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>)</span></code></pre></div>
<p><wimr></p>
</div>
<div id="add-meta-data" class="section level2">
<h2>Add meta data</h2>
<p>The metadata provided is not found in the object itself. So we will
run a small piece of code that lets you add metadata to a Seurat
object.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>data<span class="sc">$</span>ID <span class="ot">&lt;-</span> <span class="st">&quot;P004&quot;</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(data,</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>                    data<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> </span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>                      <span class="fu">right_join</span>(meta[meta<span class="sc">$</span>ID<span class="sc">==</span><span class="st">&quot;P004&quot;</span>, ], <span class="at">by=</span><span class="st">&quot;ID&quot;</span>))</span></code></pre></div>
</div>
<div id="qc" class="section level2 tabset tabset-fade">
<h2 class="tabset tabset-fade">QC</h2>
<p>There are several ways to QC transcriptomic data. First, you should
check and scrutinise the <a
href="https://github.com/vildeka/Spatial_DMPA?tab=readme-ov-file">original
manuscript</a>.</p>
<p>We will assess:</p>
<ul>
<li><p>number of genes per spot</p></li>
<li><p>number of unique genes per spot</p></li>
<li><p>mitochondrial, haemoglobin, and ribosomal percentage.</p></li>
</ul>
<p>We’ll also filter the genes themselves, to remove noisy/pointless
genes.</p>
<ul>
<li><p>genes that are not present in more than x number of
spots</p></li>
<li><p>high expressing spots.</p></li>
</ul>
<div id="genes-per-spot-unique-genes-per-spot" class="section level3">
<h3>Genes per spot &amp; Unique genes per spot</h3>
<p>This is a typical <code>single-cell RNA</code> analysis metric. It
should be carefully considered how this data is interpretted and
filtered, as this is <strong>not</strong> single cell. Spots can have
double the average or half the average based on the density of cells.
Indeed, this is a consideration for normalisation, but this
consideration is for another workshop.</p>
<p><i><strong>As there is <span style="color:red">no</span> gold
standard</strong></i>, we will demonstrate how you visualise the QC
metrics and then filter.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>p1<span class="ot">=</span><span class="fu">FeatureScatter</span>(data, <span class="st">&quot;nCount_Spatial&quot;</span>, <span class="st">&quot;nFeature_Spatial&quot;</span>)<span class="sc">+</span><span class="fu">NoLegend</span>()</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co"># number of genes</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>p2<span class="ot">=</span><span class="fu">SpatialFeaturePlot</span>(data, <span class="st">&quot;nCount_Spatial&quot;</span>, </span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>                   <span class="at">crop=</span>F,</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>                   <span class="at">pt.size=</span><span class="dv">2</span>,</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>                   <span class="at">alpha=</span><span class="fu">c</span>(.<span class="dv">2</span>,<span class="dv">1</span>))<span class="sc">+</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="co"># number of unique features</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>p3<span class="ot">=</span><span class="fu">SpatialFeaturePlot</span>(data, <span class="st">&quot;nFeature_Spatial&quot;</span>, </span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>                   <span class="at">crop=</span>F,</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>                   <span class="at">pt.size=</span><span class="dv">2</span>,</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>                   <span class="at">alpha=</span><span class="fu">c</span>(.<span class="dv">2</span>,<span class="dv">1</span>))<span class="sc">+</span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="fu">plot_grid</span>(p1,<span class="fu">plot_grid</span>(p2,p3, <span class="at">nrow=</span><span class="dv">2</span>))</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The minimum number of Features per spot
(<code>r</code>min(data$nFeature_Spatial)`) and minimum number of Counts
per spot (151) are greater than the paper’s cut off (presuming they did
not use the filtered matrix). However, for the purposes of demonstrating
how this is done:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">subset</span>(data, </span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>               <span class="at">subset =</span> nCount_Spatial <span class="sc">&gt;</span> <span class="dv">100</span> <span class="sc">&amp;</span> nFeature_Spatial <span class="sc">&gt;</span><span class="dv">100</span>)</span></code></pre></div>
</div>
<div id="module-scoring" class="section level3">
<h3>Module Scoring</h3>
<p>In this paper, they remove cells that are &gt;15% Mt genes &amp; &gt;
10% Hb reads.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(data, <span class="st">&quot;^MT-&quot;</span>, <span class="at">col.name =</span> <span class="st">&quot;percent_mito&quot;</span>)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(data, <span class="st">&quot;^HB[^(P)]&quot;</span>, <span class="at">col.name =</span> <span class="st">&quot;percent_hb&quot;</span>)</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(data, <span class="st">&quot;^RP[SL]&quot;</span>, <span class="at">col.name =</span> <span class="st">&quot;percent_ribo&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>feature <span class="ot">&lt;-</span>  <span class="fu">c</span>(<span class="st">&quot;nCount_Spatial&quot;</span>, <span class="st">&quot;nFeature_Spatial&quot;</span>,<span class="st">&quot;percent_mito&quot;</span>,<span class="st">&quot;percent_hb&quot;</span>, <span class="st">&quot;percent_ribo&quot;</span>)</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="fu">sapply</span>(data<span class="sc">@</span>meta.data[feature], summary) <span class="sc">%&gt;%</span> </span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>  <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">&quot;stat&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="at">digits =</span> <span class="dv">1</span>)</span></code></pre></div>
<table style="width:100%;">
<colgroup>
<col width="10%" />
<col width="19%" />
<col width="22%" />
<col width="16%" />
<col width="14%" />
<col width="16%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">stat</th>
<th align="right">nCount_Spatial</th>
<th align="right">nFeature_Spatial</th>
<th align="right">percent_mito</th>
<th align="right">percent_hb</th>
<th align="right">percent_ribo</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Min.</td>
<td align="right">151.0</td>
<td align="right">107.0</td>
<td align="right">1.9</td>
<td align="right">0.0</td>
<td align="right">2.3</td>
</tr>
<tr class="even">
<td align="left">1st Qu.</td>
<td align="right">1068.2</td>
<td align="right">680.0</td>
<td align="right">5.4</td>
<td align="right">0.0</td>
<td align="right">8.0</td>
</tr>
<tr class="odd">
<td align="left">Median</td>
<td align="right">2644.5</td>
<td align="right">1363.0</td>
<td align="right">6.5</td>
<td align="right">0.0</td>
<td align="right">9.6</td>
</tr>
<tr class="even">
<td align="left">Mean</td>
<td align="right">6114.8</td>
<td align="right">1934.1</td>
<td align="right">6.7</td>
<td align="right">0.1</td>
<td align="right">9.9</td>
</tr>
<tr class="odd">
<td align="left">3rd Qu.</td>
<td align="right">6933.2</td>
<td align="right">2584.2</td>
<td align="right">7.8</td>
<td align="right">0.1</td>
<td align="right">11.8</td>
</tr>
<tr class="even">
<td align="left">Max.</td>
<td align="right">70736.0</td>
<td align="right">9935.0</td>
<td align="right">14.6</td>
<td align="right">12.1</td>
<td align="right">20.9</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>p1<span class="ot">=</span><span class="fu">SpatialFeaturePlot</span>(data, <span class="st">&quot;percent_mito&quot;</span>, </span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>                   <span class="at">crop=</span>F,</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>                   <span class="at">pt.size=</span><span class="dv">2</span>,</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>                   <span class="at">alpha=</span><span class="fu">c</span>(.<span class="dv">2</span>,<span class="dv">1</span>))<span class="sc">+</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>  <span class="fu">NoLegend</span>()<span class="sc">+</span><span class="fu">ggtitle</span>(<span class="st">&quot;Mito&quot;</span>)</span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>p2<span class="ot">=</span><span class="fu">SpatialFeaturePlot</span>(data, <span class="st">&quot;percent_hb&quot;</span>, </span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>                   <span class="at">crop=</span>F,</span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a>                   <span class="at">pt.size=</span><span class="dv">2</span>,</span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a>                   <span class="at">alpha=</span><span class="fu">c</span>(.<span class="dv">2</span>,<span class="dv">1</span>))<span class="sc">+</span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a>  <span class="fu">NoLegend</span>()<span class="sc">+</span><span class="fu">ggtitle</span>(<span class="st">&quot;Hb-genes&quot;</span>)</span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a>p3<span class="ot">=</span><span class="fu">SpatialFeaturePlot</span>(data, <span class="st">&quot;percent_ribo&quot;</span>, </span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a>                   <span class="at">crop=</span>F,</span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a>                   <span class="at">pt.size=</span><span class="dv">2</span>,</span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a>                   <span class="at">alpha=</span><span class="fu">c</span>(.<span class="dv">2</span>,<span class="dv">1</span>))<span class="sc">+</span></span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a>  <span class="fu">NoLegend</span>()<span class="sc">+</span><span class="fu">ggtitle</span>(<span class="st">&quot;Ribo&quot;</span>)</span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a><span class="fu">plot_grid</span>(p1,p2,p3)</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Filter:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">subset</span>(data, <span class="at">subset =</span> percent_mito<span class="sc">&lt;</span><span class="dv">15</span> <span class="sc">&amp;</span> percent_hb <span class="sc">&lt;</span><span class="dv">10</span>)</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="fu">sapply</span>(data<span class="sc">@</span>meta.data[feature], summary) <span class="sc">%&gt;%</span> </span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>  <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">&quot;stat&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="at">digits =</span> <span class="dv">1</span>)</span></code></pre></div>
<table style="width:100%;">
<colgroup>
<col width="10%" />
<col width="19%" />
<col width="22%" />
<col width="16%" />
<col width="14%" />
<col width="16%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">stat</th>
<th align="right">nCount_Spatial</th>
<th align="right">nFeature_Spatial</th>
<th align="right">percent_mito</th>
<th align="right">percent_hb</th>
<th align="right">percent_ribo</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Min.</td>
<td align="right">151.0</td>
<td align="right">107.0</td>
<td align="right">1.9</td>
<td align="right">0.0</td>
<td align="right">2.3</td>
</tr>
<tr class="even">
<td align="left">1st Qu.</td>
<td align="right">1067.5</td>
<td align="right">681.0</td>
<td align="right">5.4</td>
<td align="right">0.0</td>
<td align="right">8.0</td>
</tr>
<tr class="odd">
<td align="left">Median</td>
<td align="right">2654.0</td>
<td align="right">1363.0</td>
<td align="right">6.5</td>
<td align="right">0.0</td>
<td align="right">9.6</td>
</tr>
<tr class="even">
<td align="left">Mean</td>
<td align="right">6118.8</td>
<td align="right">1935.1</td>
<td align="right">6.7</td>
<td align="right">0.1</td>
<td align="right">9.9</td>
</tr>
<tr class="odd">
<td align="left">3rd Qu.</td>
<td align="right">6952.5</td>
<td align="right">2585.5</td>
<td align="right">7.8</td>
<td align="right">0.1</td>
<td align="right">11.8</td>
</tr>
<tr class="even">
<td align="left">Max.</td>
<td align="right">70736.0</td>
<td align="right">9935.0</td>
<td align="right">14.6</td>
<td align="right">4.7</td>
<td align="right">20.9</td>
</tr>
</tbody>
</table>
</div>
<div id="filter-genes" class="section level3">
<h3>Filter genes</h3>
<p>We’ll first remove certain genes. MALAT1 is highly expressed in long
non-coding RNA and is pretty ubiquitous. So we dont want to filter and
subsequently cluster according to the expression of this gene.
Similarly, Hb genes are highly expressed in rbc, which may represent
contamination. We can remove these too.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>keep_genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(data)[<span class="sc">!</span><span class="fu">rownames</span>(data) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="fu">grep</span>(<span class="st">&quot;MALAT1&quot;</span>, <span class="fu">rownames</span>(data), <span class="at">value=</span>T), <span class="fu">grep</span>(<span class="st">&quot;^HB[^(P)]&quot;</span>, <span class="fu">rownames</span>(data), <span class="at">value=</span>T))]</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">subset</span>(data, <span class="at">features =</span> keep_genes)</span></code></pre></div>
<p>Low genes can create noise and take up unnecessary space in the
object. Lets remove genes according to the original manuscript, whereby
a gene not found in &gt; 2 spots is removed. You can adjust this as you
wish.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">rowSums</span>(data<span class="sc">@</span>assays<span class="sc">$</span>Spatial<span class="sc">@</span>layers<span class="sc">$</span>counts<span class="sc">&gt;=</span><span class="dv">1</span>)<span class="sc">&gt;=</span><span class="dv">2</span>)</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="cn">FALSE</span>  <span class="cn">TRUE</span> </span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="dv">17971</span> <span class="dv">18617</span> </span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a>keep_genes <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(data<span class="sc">@</span>assays<span class="sc">$</span>Spatial<span class="sc">@</span>layers<span class="sc">$</span>counts <span class="sc">&gt;=</span> <span class="dv">1</span>) <span class="sc">&gt;=</span> <span class="dv">2</span></span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a><span class="co">#before filtering</span></span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a><span class="fu">dim</span>(data)</span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">36588</span>  <span class="dv">1255</span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">subset</span>(data, <span class="at">features =</span> <span class="fu">rownames</span>(data)[keep_genes])</span>
<span id="cb18-10"><a href="#cb18-10" tabindex="-1"></a><span class="co"># after filtering</span></span>
<span id="cb18-11"><a href="#cb18-11" tabindex="-1"></a><span class="fu">dim</span>(data)</span>
<span id="cb18-12"><a href="#cb18-12" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">18617</span>  <span class="dv">1255</span></span></code></pre></div>
</div>
</div>
</div>
<hr>
<p>We recommend taking some time to carefully consider your QC strategy.
<i><strong>There is no universal gold standard</i></strong>. One common
approach is to start with lenient QC, proceed with processing, and
during clustering and annotation, remain aware of the relaxed filtering.
If unexpected results arise, you can revisit and refine your QC.
Personally, I prefer to be permissive at first, while tagging potential
outliers for future consideration. For example, we initially filtered
out genes not present in at least 2 spots, but I may also label genes
present in fewer than 5 spots as “toQC” for later review—particularly
when assessing differential expression results. Similarly, we applied a
15% mitochondrial threshold but did not filter based on ribosomal
content. At a later stage, we could tighten the mitochondrial threshold
to 10% and include a filter for high ribosomal spots. If unusual
clusters appear, we can then assess how many of those cells would have
been excluded under stricter QC.</p>
</div>
<div id="data-processing" class="section level1 tabset tabset-fade">
<h1 class="tabset tabset-fade">Data Processing</h1>
<p>Seurat is an efficient package with simple functions for standard
processing.</p>
<ul>
<li><p><code>NormalizeData</code>: The default is to log-normalize
<em>(Feature counts for each cell are divided by the total counts for
that cell and multiplied by the scale.factor. This is then natural-log
transformed using log1p)</em>. Spatial data may be better normalised
with Area/cell count considered. As we don’t have cell counts per spot
here, we’ll stick to the standard approach and that taken by the
authors</p></li>
<li><p><code>FindVariableFeatures</code>: The standard approach is 2000
top variable genes + ‘vst’ selection method. <em>(First, fits a line to
the relationship of log(variance) and log(mean) using local polynomial
regression (loess). Then standardizes the feature values using the
observed mean and expected variance (given by the fitted line). Feature
variance is then calculated on the standardized values after clipping to
a maximum)</em>. Standard is 2000 genes</p></li>
<li><p><code>ScaleData</code>: here, we scale just the variable genes to
save space. If you find yourself unable to visualise your gene of
interest in subsequent visualisations, it was not in this top variable
list. In that case, you can scale all data.</p></li>
<li><p><code>RunPCA</code> &amp; <code>RunUMAP</code>: Visualisation
tools. There are a few things to consider in the UMAP function, but most
important is the <code>dims</code> argument. We’ll cover this
below.</p></li>
</ul>
<div>
<div id="normalisation" class="section level2">
<h2>Normalisation</h2>
<p>Again, normalisation is quite simple.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(data, </span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>                      <span class="at">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>,</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>                      <span class="at">verbose=</span>F)</span></code></pre></div>
</div>
<div id="variable-features" class="section level2">
<h2>Variable Features</h2>
<p>We’ll find the top variable features.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(data, </span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>                             <span class="at">method =</span> <span class="st">&#39;vst&#39;</span>, </span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>                             <span class="at">nfeatures =</span> <span class="dv">2000</span>, </span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a>                             <span class="at">verbose=</span>F)</span></code></pre></div>
<p>We can output/visualise them in several ways.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">VariableFeatures</span>(data) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">50</span>)</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a> [<span class="dv">1</span>] <span class="st">&quot;PI3&quot;</span>        <span class="st">&quot;SPRR1A&quot;</span>     <span class="st">&quot;IGLC3&quot;</span>      <span class="st">&quot;FLG&quot;</span>        <span class="st">&quot;IGHA2&quot;</span>     </span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a> [<span class="dv">6</span>] <span class="st">&quot;IGLC2&quot;</span>      <span class="st">&quot;IGHA1&quot;</span>      <span class="st">&quot;JCHAIN&quot;</span>     <span class="st">&quot;IGHG1&quot;</span>      <span class="st">&quot;AC019349.1&quot;</span></span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a>[<span class="dv">11</span>] <span class="st">&quot;IGKC&quot;</span>       <span class="st">&quot;IGLC1&quot;</span>      <span class="st">&quot;IGHG4&quot;</span>      <span class="st">&quot;KRT10&quot;</span>      <span class="st">&quot;IGHM&quot;</span>      </span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a>[<span class="dv">16</span>] <span class="st">&quot;MUSTN1&quot;</span>     <span class="st">&quot;CRISP3&quot;</span>     <span class="st">&quot;CEACAM5&quot;</span>    <span class="st">&quot;SPINK7&quot;</span>     <span class="st">&quot;SERPINB4&quot;</span>  </span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a>[<span class="dv">21</span>] <span class="st">&quot;KRTDAP&quot;</span>     <span class="st">&quot;SERPINB3&quot;</span>   <span class="st">&quot;PI15&quot;</span>       <span class="st">&quot;RPTN&quot;</span>       <span class="st">&quot;SPRR2B&quot;</span>    </span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a>[<span class="dv">26</span>] <span class="st">&quot;KLK6&quot;</span>       <span class="st">&quot;MGST1&quot;</span>      <span class="st">&quot;KRT6B&quot;</span>      <span class="st">&quot;SLURP1&quot;</span>     <span class="st">&quot;ERO1A&quot;</span>     </span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a>[<span class="dv">31</span>] <span class="st">&quot;CCL14&quot;</span>      <span class="st">&quot;KRT1&quot;</span>       <span class="st">&quot;SPRR2E&quot;</span>     <span class="st">&quot;RNASE7&quot;</span>     <span class="st">&quot;TOP2A&quot;</span>     </span>
<span id="cb21-9"><a href="#cb21-9" tabindex="-1"></a>[<span class="dv">36</span>] <span class="st">&quot;CRCT1&quot;</span>      <span class="st">&quot;MT1X&quot;</span>       <span class="st">&quot;RRM2&quot;</span>       <span class="st">&quot;S100A7&quot;</span>     <span class="st">&quot;DUOXA2&quot;</span>    </span>
<span id="cb21-10"><a href="#cb21-10" tabindex="-1"></a>[<span class="dv">41</span>] <span class="st">&quot;STMN1&quot;</span>      <span class="st">&quot;LCE3D&quot;</span>      <span class="st">&quot;PCNA&quot;</span>       <span class="st">&quot;FAM25A&quot;</span>     <span class="st">&quot;IGHG3&quot;</span>     </span>
<span id="cb21-11"><a href="#cb21-11" tabindex="-1"></a>[<span class="dv">46</span>] <span class="st">&quot;C15orf48&quot;</span>   <span class="st">&quot;OLFM4&quot;</span>      <span class="st">&quot;PTTG1&quot;</span>      <span class="st">&quot;YOD1&quot;</span>       <span class="st">&quot;CEACAM7&quot;</span>   </span>
<span id="cb21-12"><a href="#cb21-12" tabindex="-1"></a></span>
<span id="cb21-13"><a href="#cb21-13" tabindex="-1"></a><span class="fu">LabelPoints</span>(<span class="at">plot =</span> <span class="fu">VariableFeaturePlot</span>(data), </span>
<span id="cb21-14"><a href="#cb21-14" tabindex="-1"></a>            <span class="at">points =</span> <span class="fu">head</span>(<span class="fu">VariableFeatures</span>(data), <span class="dv">50</span>), <span class="at">repel =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="scale-data" class="section level2">
<h2>Scale Data</h2>
<p>Scaling the data is also easy!</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(data, </span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>                  <span class="at">features =</span> <span class="fu">VariableFeatures</span>(data), <span class="co">#if you want to scale all features, change this to rownames(data)</span></span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>        <span class="at">verbose=</span>F)</span></code></pre></div>
</div>
<div id="pca-umap" class="section level2">
<h2>PCA &amp; UMAP</h2>
<p>Also very easy with Seurat. We’ll determine the number of PCs to use
for the UMAP using ElbowPlot</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(data,</span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>               <span class="at">npcs =</span> <span class="dv">50</span>,</span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>               <span class="at">verbose=</span>F)</span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a><span class="fu">ElbowPlot</span>(data)</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The elbow plot drops off drastically and plateaus at ~6, meaning the
variation just becomes noisy around PC6-onwards. So we’ll just use the
first 6.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(data, </span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>                <span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, </span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a>                <span class="at">verbose=</span>F)</span></code></pre></div>
</div>
</div>
<hr>
</div>
<div id="visualisation" class="section level1 tabset tabset-fade">
<h1 class="tabset tabset-fade">Visualisation</h1>
<p>There are several visualisations we can do:</p>
<p><strong>Graph</strong>:</p>
<ul>
<li><p>UMAP visualisation</p></li>
<li><p>Cluster based expressions</p></li>
</ul>
<p><strong>Spatial</strong>:</p>
<ul>
<li><p>Genes overlaid on the tissue</p></li>
<li><p>Module scores overlaid on the tissue</p></li>
</ul>
<div id="umap" class="section level2 tabset tabset-fade">
<h2 class="tabset tabset-fade">UMAP</h2>
<div id="umap-visualisation" class="section level3">
<h3>UMAP Visualisation</h3>
<p>We’ll first cluster the data.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>data <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span> </span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a>  <span class="fu">FindNeighbors</span>(<span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="at">verbose=</span>F) <span class="sc">%&gt;%</span></span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>  <span class="fu">FindClusters</span>(<span class="at">resolution=</span><span class="dv">1</span>, <span class="at">verbose=</span>F)</span></code></pre></div>
<p>And now we can create a UMAP.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a>p1<span class="ot">=</span><span class="fu">UMAPPlot</span>(data,</span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a>            <span class="at">label=</span>T, <span class="at">label.box=</span>T)<span class="sc">+</span></span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a>  <span class="fu">NoLegend</span>()<span class="sc">+</span></span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a>  <span class="fu">NoAxes</span>()<span class="sc">+</span></span>
<span id="cb26-5"><a href="#cb26-5" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;11 Clusters&quot;</span>)</span></code></pre></div>
<details>
<summary>
<strong>Open to view more Plotting options</strong>
</summary>
<p>Here are some additional arguments that change the default
visualisations. Chop and change to see how they work:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="fu">UMAPPlot</span>(data, </span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>         <span class="at">pt.size=</span><span class="dv">2</span>, </span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a>         <span class="at">label=</span>T, </span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a>         <span class="at">label.size=</span><span class="dv">10</span>,</span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a>         <span class="at">label.box=</span>T,</span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a>         <span class="at">label.color=</span><span class="st">&quot;white&quot;</span>,</span>
<span id="cb27-7"><a href="#cb27-7" tabindex="-1"></a>         <span class="at">repel=</span>T)<span class="sc">+</span></span>
<span id="cb27-8"><a href="#cb27-8" tabindex="-1"></a>  <span class="fu">NoLegend</span>()<span class="sc">+</span></span>
<span id="cb27-9"><a href="#cb27-9" tabindex="-1"></a>  <span class="fu">NoAxes</span>()<span class="sc">+</span></span>
<span id="cb27-10"><a href="#cb27-10" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span> <span class="fu">c</span>(<span class="st">&quot;blue4&quot;</span>, <span class="st">&#39;blue2&#39;</span>, <span class="st">&#39;lightblue&#39;</span>, <span class="st">&#39;green2&#39;</span>, <span class="st">&#39;green4&#39;</span>,<span class="st">&#39;brown&#39;</span>, <span class="st">&#39;orange&#39;</span>, <span class="st">&#39;red&#39;</span>, <span class="st">&#39;red3&#39;</span>, <span class="st">&#39;red4&#39;</span>, <span class="st">&#39;black&#39;</span>))<span class="sc">+</span></span>
<span id="cb27-11"><a href="#cb27-11" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values=</span> <span class="fu">c</span>(<span class="st">&quot;blue4&quot;</span>, <span class="st">&#39;blue2&#39;</span>, <span class="st">&#39;lightblue&#39;</span>, <span class="st">&#39;green2&#39;</span>, <span class="st">&#39;green4&#39;</span>,<span class="st">&#39;brown&#39;</span>, <span class="st">&#39;orange&#39;</span>, <span class="st">&#39;red&#39;</span>, <span class="st">&#39;red3&#39;</span>, <span class="st">&#39;red4&#39;</span>, <span class="st">&#39;black&#39;</span>))</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;" /></p>
</details>
</div>
<div id="featureplots" class="section level3">
<h3>Featureplots</h3>
<p>We can view the gene expression on the UMAP itself.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a>p2<span class="ot">=</span><span class="fu">FeaturePlot</span>(data, </span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a>            <span class="st">&quot;CD3E&quot;</span>,</span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a>            <span class="at">order=</span>T,</span>
<span id="cb28-4"><a href="#cb28-4" tabindex="-1"></a>            <span class="at">pt.size=</span><span class="dv">2</span>, </span>
<span id="cb28-5"><a href="#cb28-5" tabindex="-1"></a>            <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">&quot;yellow2&quot;</span>, <span class="st">&#39;blue3&#39;</span>))<span class="sc">+</span><span class="fu">NoAxes</span>()<span class="sc">+</span><span class="fu">NoLegend</span>()</span>
<span id="cb28-6"><a href="#cb28-6" tabindex="-1"></a>p3<span class="ot">=</span><span class="fu">FeaturePlot</span>(data, </span>
<span id="cb28-7"><a href="#cb28-7" tabindex="-1"></a>            <span class="st">&quot;F13A1&quot;</span>,</span>
<span id="cb28-8"><a href="#cb28-8" tabindex="-1"></a>            <span class="at">order=</span>T,</span>
<span id="cb28-9"><a href="#cb28-9" tabindex="-1"></a>            <span class="at">pt.size=</span><span class="dv">2</span>, </span>
<span id="cb28-10"><a href="#cb28-10" tabindex="-1"></a>            <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">&quot;yellow2&quot;</span>, <span class="st">&#39;blue3&#39;</span>))<span class="sc">+</span><span class="fu">NoAxes</span>()<span class="sc">+</span><span class="fu">NoLegend</span>()</span>
<span id="cb28-11"><a href="#cb28-11" tabindex="-1"></a>p4<span class="ot">=</span><span class="fu">FeaturePlot</span>(data, </span>
<span id="cb28-12"><a href="#cb28-12" tabindex="-1"></a>            <span class="st">&quot;COL1A1&quot;</span>,</span>
<span id="cb28-13"><a href="#cb28-13" tabindex="-1"></a>            <span class="at">order=</span>T,</span>
<span id="cb28-14"><a href="#cb28-14" tabindex="-1"></a>            <span class="at">pt.size=</span><span class="dv">2</span>, </span>
<span id="cb28-15"><a href="#cb28-15" tabindex="-1"></a>            <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">&quot;yellow2&quot;</span>, <span class="st">&#39;blue3&#39;</span>))<span class="sc">+</span><span class="fu">NoAxes</span>()<span class="sc">+</span><span class="fu">NoLegend</span>()</span>
<span id="cb28-16"><a href="#cb28-16" tabindex="-1"></a>p5<span class="ot">=</span><span class="fu">FeaturePlot</span>(data, </span>
<span id="cb28-17"><a href="#cb28-17" tabindex="-1"></a>            <span class="st">&quot;KRT5&quot;</span>,</span>
<span id="cb28-18"><a href="#cb28-18" tabindex="-1"></a>            <span class="at">order=</span>T,</span>
<span id="cb28-19"><a href="#cb28-19" tabindex="-1"></a>            <span class="at">pt.size=</span><span class="dv">2</span>,</span>
<span id="cb28-20"><a href="#cb28-20" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="dv">3</span>,</span>
<span id="cb28-21"><a href="#cb28-21" tabindex="-1"></a>            <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">&quot;yellow2&quot;</span>, <span class="st">&#39;blue3&#39;</span>))<span class="sc">+</span><span class="fu">NoAxes</span>()<span class="sc">+</span><span class="fu">NoLegend</span>()</span>
<span id="cb28-22"><a href="#cb28-22" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, <span class="fu">plot_grid</span>(p2,p3,p4,p5, <span class="at">ncol=</span><span class="dv">2</span>))</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-27-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="de-and-dotplots" class="section level3">
<h3>DE and DotPlots</h3>
<p>And we can simply assess the clusters themselves using differential
analysis and expression graphs.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a></span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(data, </span>
<span id="cb29-4"><a href="#cb29-4" tabindex="-1"></a>                          <span class="at">logfc.threshold =</span> <span class="fl">0.1</span>,<span class="co">#default - increasing speeds it up, but may miss weaker signals.</span></span>
<span id="cb29-5"><a href="#cb29-5" tabindex="-1"></a>                          <span class="at">min.pct =</span> <span class="fl">0.01</span>, <span class="co">#default - can be increased to ensure that the minimum expression of genes are considered. </span></span>
<span id="cb29-6"><a href="#cb29-6" tabindex="-1"></a>                          <span class="at">min.diff.pct =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="co"># you can adjust this too. IF you want there to be at least a 10% difference in percentage of spots expressing the gene, you&#39;d change this to 0.1. Can ensure that its not only the level of expression, but that you&#39;re capturing highly expressed genes</span></span>
<span id="cb29-7"><a href="#cb29-7" tabindex="-1"></a>                          <span class="at">verbose=</span>F <span class="co">#change this to true if you want to track the speed at which DE is being calculated. It is off here to save it outputting into the document </span></span>
<span id="cb29-8"><a href="#cb29-8" tabindex="-1"></a>                          ) </span>
<span id="cb29-9"><a href="#cb29-9" tabindex="-1"></a>top10 <span class="ot">&lt;-</span> markers <span class="sc">%&gt;%</span> </span>
<span id="cb29-10"><a href="#cb29-10" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span> </span>
<span id="cb29-11"><a href="#cb29-11" tabindex="-1"></a>  <span class="fu">top_n</span>(<span class="at">wt=</span>avg_log2FC, <span class="at">n=</span><span class="dv">10</span>)</span></code></pre></div>
<details>
<summary>
Heatmap
</summary>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="fu">DoHeatmap</span>(<span class="fu">AggregateExpression</span>(<span class="fu">ScaleData</span>(data, <span class="fu">rownames</span>(data), <span class="at">verbose=</span>F), <span class="at">return.seurat =</span> T, <span class="at">verbose=</span>F),</span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a>          <span class="at">features =</span> top10<span class="sc">$</span>gene, </span>
<span id="cb30-3"><a href="#cb30-3" tabindex="-1"></a>          <span class="at">draw.lines =</span> F)</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;" /></p>
<hr>
</details>
<details>
<summary>
DotPlot
</summary>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="fu">DotPlot</span>(data, </span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;CD3E&quot;</span>, <span class="st">&quot;CLEC10A&quot;</span>, <span class="st">&quot;CD207&quot;</span>, <span class="st">&quot;KRT5&quot;</span>, <span class="st">&quot;COL1A1&quot;</span>), </span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a>        <span class="at">cols=</span><span class="fu">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;red&quot;</span>))</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-30-1.png" width="672" style="display: block; margin: auto;" /></p>
</details>
<details>
<summary>
Violin
</summary>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="fu">VlnPlot</span>(data, </span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;CD3E&quot;</span>, <span class="st">&quot;CLEC10A&quot;</span>, <span class="st">&quot;CD207&quot;</span>, <span class="st">&quot;KRT5&quot;</span>, <span class="st">&quot;COL1A1&quot;</span>))</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-31-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="fu">VlnPlot</span>(data, </span>
<span id="cb33-2"><a href="#cb33-2" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;CD3E&quot;</span>, <span class="st">&quot;CLEC10A&quot;</span>, <span class="st">&quot;CD207&quot;</span>, <span class="st">&quot;KRT5&quot;</span>, <span class="st">&quot;COL1A1&quot;</span>), <span class="at">stack=</span>T)</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-31-2.png" width="672" style="display: block; margin: auto;" /></p>
</details>
</div>
</div>
<div id="spatial" class="section level2 tabset tabset-fade">
<h2 class="tabset tabset-fade">Spatial</h2>
<p>We can do a few things here.</p>
<div id="clusters-overlaid-spatially" class="section level3">
<h3>Clusters overlaid spatially</h3>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="fu">SpatialDimPlot</span>(data, <span class="at">pt.size=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-32-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="fu">ImageDimPlot</span>(data, <span class="at">size=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-32-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="expressions-overlaid-spatially" class="section level3">
<h3>Expressions overlaid spatially</h3>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(data, </span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a>                   <span class="at">features=</span><span class="st">&quot;CD3E&quot;</span>,</span>
<span id="cb36-3"><a href="#cb36-3" tabindex="-1"></a>                   <span class="at">pt.size=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-33-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a></span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(data, </span>
<span id="cb37-3"><a href="#cb37-3" tabindex="-1"></a>                   <span class="at">features=</span><span class="st">&quot;CD3E&quot;</span>,</span>
<span id="cb37-4"><a href="#cb37-4" tabindex="-1"></a>                   <span class="at">pt.size=</span><span class="dv">3</span>)<span class="sc">+</span></span>
<span id="cb37-5"><a href="#cb37-5" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_b</span>()</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-33-2.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a></span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a><span class="fu">ImageFeaturePlot</span>(data, </span>
<span id="cb38-3"><a href="#cb38-3" tabindex="-1"></a>                   <span class="at">features=</span><span class="st">&quot;CD3E&quot;</span>,</span>
<span id="cb38-4"><a href="#cb38-4" tabindex="-1"></a>                   <span class="at">size=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="figure/202505_Workshop.Rmd/unnamed-chunk-33-3.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb39-2"><a href="#cb39-2" tabindex="-1"></a>R version <span class="dv">4</span>.<span class="fl">4.0</span> (<span class="dv">2024-04-24</span>)</span>
<span id="cb39-3"><a href="#cb39-3" tabindex="-1"></a>Platform<span class="sc">:</span> aarch64<span class="sc">-</span>apple<span class="sc">-</span>darwin20</span>
<span id="cb39-4"><a href="#cb39-4" tabindex="-1"></a>Running under<span class="sc">:</span> macOS Sonoma <span class="fl">14.3</span></span>
<span id="cb39-5"><a href="#cb39-5" tabindex="-1"></a></span>
<span id="cb39-6"><a href="#cb39-6" tabindex="-1"></a>Matrix products<span class="sc">:</span> default</span>
<span id="cb39-7"><a href="#cb39-7" tabindex="-1"></a>BLAS<span class="sc">:</span>   <span class="er">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.4</span><span class="sc">-</span>arm64<span class="sc">/</span>Resources<span class="sc">/</span>lib<span class="sc">/</span>libRblas.<span class="fl">0.</span>dylib </span>
<span id="cb39-8"><a href="#cb39-8" tabindex="-1"></a>LAPACK<span class="sc">:</span> <span class="er">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.4</span><span class="sc">-</span>arm64<span class="sc">/</span>Resources<span class="sc">/</span>lib<span class="sc">/</span>libRlapack.dylib;  LAPACK version <span class="dv">3</span>.<span class="fl">12.0</span></span>
<span id="cb39-9"><a href="#cb39-9" tabindex="-1"></a></span>
<span id="cb39-10"><a href="#cb39-10" tabindex="-1"></a>locale<span class="sc">:</span></span>
<span id="cb39-11"><a href="#cb39-11" tabindex="-1"></a>[<span class="dv">1</span>] en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>C<span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span></span>
<span id="cb39-12"><a href="#cb39-12" tabindex="-1"></a></span>
<span id="cb39-13"><a href="#cb39-13" tabindex="-1"></a>time zone<span class="sc">:</span> Australia<span class="sc">/</span>Sydney</span>
<span id="cb39-14"><a href="#cb39-14" tabindex="-1"></a>tzcode source<span class="sc">:</span> internal</span>
<span id="cb39-15"><a href="#cb39-15" tabindex="-1"></a></span>
<span id="cb39-16"><a href="#cb39-16" tabindex="-1"></a>attached base packages<span class="sc">:</span></span>
<span id="cb39-17"><a href="#cb39-17" tabindex="-1"></a>[<span class="dv">1</span>] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span id="cb39-18"><a href="#cb39-18" tabindex="-1"></a></span>
<span id="cb39-19"><a href="#cb39-19" tabindex="-1"></a>other attached packages<span class="sc">:</span></span>
<span id="cb39-20"><a href="#cb39-20" tabindex="-1"></a> [<span class="dv">1</span>] GEOquery_2.<span class="fl">72.0</span>     Biobase_2.<span class="fl">64.0</span>      BiocGenerics_0.<span class="fl">50.0</span></span>
<span id="cb39-21"><a href="#cb39-21" tabindex="-1"></a> [<span class="dv">4</span>] cowplot_1.<span class="fl">1.3</span>       lubridate_1.<span class="fl">9.4</span>     forcats_1.<span class="fl">0.0</span>      </span>
<span id="cb39-22"><a href="#cb39-22" tabindex="-1"></a> [<span class="dv">7</span>] stringr_1.<span class="fl">5.1</span>       dplyr_1.<span class="fl">1.4</span>         purrr_1.<span class="fl">0.4</span>        </span>
<span id="cb39-23"><a href="#cb39-23" tabindex="-1"></a>[<span class="dv">10</span>] readr_2.<span class="fl">1.5</span>         tidyr_1.<span class="fl">3.1</span>         tibble_3.<span class="fl">2.1</span>       </span>
<span id="cb39-24"><a href="#cb39-24" tabindex="-1"></a>[<span class="dv">13</span>] ggplot2_3.<span class="fl">5.1</span>       tidyverse_2.<span class="fl">0.0</span>     Seurat_5.<span class="fl">2.1</span>       </span>
<span id="cb39-25"><a href="#cb39-25" tabindex="-1"></a>[<span class="dv">16</span>] SeuratObject_5.<span class="fl">0.2</span>  sp_2<span class="fl">.2</span><span class="dv">-0</span>            cli_3.<span class="fl">6.4</span>          </span>
<span id="cb39-26"><a href="#cb39-26" tabindex="-1"></a>[<span class="dv">19</span>] workflowr_1.<span class="fl">7.1</span>    </span>
<span id="cb39-27"><a href="#cb39-27" tabindex="-1"></a></span>
<span id="cb39-28"><a href="#cb39-28" tabindex="-1"></a>loaded via a <span class="fu">namespace</span> (and not attached)<span class="sc">:</span></span>
<span id="cb39-29"><a href="#cb39-29" tabindex="-1"></a>  [<span class="dv">1</span>] RColorBrewer_1<span class="fl">.1</span><span class="dv">-3</span>     rstudioapi_0.<span class="fl">17.1</span>      jsonlite_1.<span class="fl">9.1</span>        </span>
<span id="cb39-30"><a href="#cb39-30" tabindex="-1"></a>  [<span class="dv">4</span>] magrittr_2.<span class="fl">0.3</span>         ggbeeswarm_0.<span class="fl">7.2</span>       spatstat.utils_3<span class="fl">.1</span><span class="dv">-3</span>  </span>
<span id="cb39-31"><a href="#cb39-31" tabindex="-1"></a>  [<span class="dv">7</span>] farver_2.<span class="fl">1.2</span>           rmarkdown_2<span class="fl">.29</span>         fs_1.<span class="fl">6.5</span>              </span>
<span id="cb39-32"><a href="#cb39-32" tabindex="-1"></a> [<span class="dv">10</span>] vctrs_0.<span class="fl">6.5</span>            ROCR_1<span class="fl">.0</span><span class="dv">-11</span>            spatstat.explore_3<span class="fl">.3</span><span class="dv">-4</span></span>
<span id="cb39-33"><a href="#cb39-33" tabindex="-1"></a> [<span class="dv">13</span>] htmltools_0.<span class="dv">5</span>.<span class="fl">8.1</span>      sass_0.<span class="fl">4.9</span>             sctransform_0.<span class="fl">4.1</span>     </span>
<span id="cb39-34"><a href="#cb39-34" tabindex="-1"></a> [<span class="dv">16</span>] parallelly_1.<span class="fl">42.0</span>      KernSmooth_2<span class="fl">.23</span><span class="dv">-26</span>     bslib_0.<span class="fl">9.0</span>           </span>
<span id="cb39-35"><a href="#cb39-35" tabindex="-1"></a> [<span class="dv">19</span>] htmlwidgets_1.<span class="fl">6.4</span>      ica_1<span class="fl">.0</span><span class="dv">-3</span>              plyr_1.<span class="fl">8.9</span>            </span>
<span id="cb39-36"><a href="#cb39-36" tabindex="-1"></a> [<span class="dv">22</span>] plotly_4.<span class="fl">10.4</span>          zoo_1<span class="fl">.8</span><span class="dv">-13</span>             cachem_1.<span class="fl">1.0</span>          </span>
<span id="cb39-37"><a href="#cb39-37" tabindex="-1"></a> [<span class="dv">25</span>] whisker_0.<span class="fl">4.1</span>          igraph_2.<span class="fl">1.4</span>           mime_0<span class="fl">.13</span>             </span>
<span id="cb39-38"><a href="#cb39-38" tabindex="-1"></a> [<span class="dv">28</span>] lifecycle_1.<span class="fl">0.4</span>        pkgconfig_2.<span class="fl">0.3</span>        Matrix_1<span class="fl">.7</span><span class="dv">-3</span>          </span>
<span id="cb39-39"><a href="#cb39-39" tabindex="-1"></a> [<span class="dv">31</span>] R6_2.<span class="fl">6.1</span>               fastmap_1.<span class="fl">2.0</span>          fitdistrplus_1<span class="fl">.2</span><span class="dv">-2</span>    </span>
<span id="cb39-40"><a href="#cb39-40" tabindex="-1"></a> [<span class="dv">34</span>] future_1.<span class="fl">34.0</span>          shiny_1.<span class="fl">10.0</span>           digest_0.<span class="fl">6.37</span>         </span>
<span id="cb39-41"><a href="#cb39-41" tabindex="-1"></a> [<span class="dv">37</span>] colorspace_2<span class="fl">.1</span><span class="dv">-1</span>       patchwork_1.<span class="fl">3.0</span>        ps_1.<span class="fl">9.0</span>              </span>
<span id="cb39-42"><a href="#cb39-42" tabindex="-1"></a> [<span class="dv">40</span>] rprojroot_2.<span class="fl">0.4</span>        tensor_1<span class="fl">.5</span>             RSpectra_0<span class="fl">.16</span><span class="dv">-2</span>       </span>
<span id="cb39-43"><a href="#cb39-43" tabindex="-1"></a> [<span class="dv">43</span>] irlba_2.<span class="dv">3</span>.<span class="fl">5.1</span>          labeling_0.<span class="fl">4.3</span>         progressr_0.<span class="fl">15.1</span>      </span>
<span id="cb39-44"><a href="#cb39-44" tabindex="-1"></a> [<span class="dv">46</span>] timechange_0.<span class="fl">3.0</span>       spatstat.sparse_3<span class="fl">.1</span><span class="dv">-0</span>  httr_1.<span class="fl">4.7</span>            </span>
<span id="cb39-45"><a href="#cb39-45" tabindex="-1"></a> [<span class="dv">49</span>] polyclip_1<span class="fl">.10</span><span class="dv">-7</span>        abind_1<span class="fl">.4</span><span class="dv">-8</span>            compiler_4.<span class="fl">4.0</span>        </span>
<span id="cb39-46"><a href="#cb39-46" tabindex="-1"></a> [<span class="dv">52</span>] bit64_4.<span class="fl">6.0</span><span class="dv">-1</span>          withr_3.<span class="fl">0.2</span>            fastDummies_1.<span class="fl">7.5</span>     </span>
<span id="cb39-47"><a href="#cb39-47" tabindex="-1"></a> [<span class="dv">55</span>] MASS_7<span class="fl">.3</span><span class="dv">-65</span>            tools_4.<span class="fl">4.0</span>            vipor_0.<span class="fl">4.7</span>           </span>
<span id="cb39-48"><a href="#cb39-48" tabindex="-1"></a> [<span class="dv">58</span>] lmtest_0<span class="fl">.9</span><span class="dv">-40</span>          beeswarm_0.<span class="fl">4.0</span>         httpuv_1.<span class="fl">6.15</span>         </span>
<span id="cb39-49"><a href="#cb39-49" tabindex="-1"></a> [<span class="dv">61</span>] future.apply_1.<span class="fl">11.3</span>    goftest_1<span class="fl">.2</span><span class="dv">-3</span>          glue_1.<span class="fl">8.0</span>            </span>
<span id="cb39-50"><a href="#cb39-50" tabindex="-1"></a> [<span class="dv">64</span>] callr_3.<span class="fl">7.6</span>            nlme_3<span class="fl">.1</span><span class="dv">-167</span>           promises_1.<span class="fl">3.2</span>        </span>
<span id="cb39-51"><a href="#cb39-51" tabindex="-1"></a> [<span class="dv">67</span>] grid_4.<span class="fl">4.0</span>             Rtsne_0<span class="fl">.17</span>             getPass_0<span class="fl">.2</span><span class="dv">-4</span>         </span>
<span id="cb39-52"><a href="#cb39-52" tabindex="-1"></a> [<span class="dv">70</span>] cluster_2.<span class="dv">1</span>.<span class="fl">8.1</span>        reshape2_1.<span class="fl">4.4</span>         generics_0.<span class="fl">1.3</span>        </span>
<span id="cb39-53"><a href="#cb39-53" tabindex="-1"></a> [<span class="dv">73</span>] hdf5r_1.<span class="fl">3.12</span>           gtable_0.<span class="fl">3.6</span>           spatstat.data_3<span class="fl">.1</span><span class="dv">-6</span>   </span>
<span id="cb39-54"><a href="#cb39-54" tabindex="-1"></a> [<span class="dv">76</span>] tzdb_0.<span class="fl">5.0</span>             hms_1.<span class="fl">1.3</span>              data.table_1.<span class="fl">17.0</span>     </span>
<span id="cb39-55"><a href="#cb39-55" tabindex="-1"></a> [<span class="dv">79</span>] xml2_1.<span class="fl">3.8</span>             spatstat.geom_3<span class="fl">.3</span><span class="dv">-6</span>    RcppAnnoy_0.<span class="fl">0.22</span>      </span>
<span id="cb39-56"><a href="#cb39-56" tabindex="-1"></a> [<span class="dv">82</span>] ggrepel_0.<span class="fl">9.6</span>          RANN_2.<span class="fl">6.2</span>             pillar_1.<span class="fl">10.1</span>         </span>
<span id="cb39-57"><a href="#cb39-57" tabindex="-1"></a> [<span class="dv">85</span>] limma_3.<span class="fl">60.6</span>           spam_2<span class="fl">.11</span><span class="dv">-1</span>            RcppHNSW_0.<span class="fl">6.0</span>        </span>
<span id="cb39-58"><a href="#cb39-58" tabindex="-1"></a> [<span class="dv">88</span>] later_1.<span class="fl">4.1</span>            splines_4.<span class="fl">4.0</span>          lattice_0<span class="fl">.22</span><span class="dv">-6</span>        </span>
<span id="cb39-59"><a href="#cb39-59" tabindex="-1"></a> [<span class="dv">91</span>] bit_4.<span class="fl">6.0</span>              survival_3<span class="fl">.8</span><span class="dv">-3</span>         deldir_2<span class="fl">.0</span><span class="dv">-4</span>          </span>
<span id="cb39-60"><a href="#cb39-60" tabindex="-1"></a> [<span class="dv">94</span>] tidyselect_1.<span class="fl">2.1</span>       miniUI_0.<span class="dv">1</span>.<span class="fl">1.1</span>         pbapply_1<span class="fl">.7</span><span class="dv">-2</span>         </span>
<span id="cb39-61"><a href="#cb39-61" tabindex="-1"></a> [<span class="dv">97</span>] knitr_1<span class="fl">.50</span>             git2r_0.<span class="fl">35.0</span>           gridExtra_2<span class="fl">.3</span>         </span>
<span id="cb39-62"><a href="#cb39-62" tabindex="-1"></a>[<span class="dv">100</span>] scattermore_1<span class="fl">.2</span>        xfun_0<span class="fl">.51</span>              statmod_1.<span class="fl">5.0</span>         </span>
<span id="cb39-63"><a href="#cb39-63" tabindex="-1"></a>[<span class="dv">103</span>] matrixStats_1.<span class="fl">5.0</span>      stringi_1.<span class="fl">8.4</span>          lazyeval_0.<span class="fl">2.2</span>        </span>
<span id="cb39-64"><a href="#cb39-64" tabindex="-1"></a>[<span class="dv">106</span>] yaml_2.<span class="fl">3.10</span>            evaluate_1.<span class="fl">0.3</span>         codetools_0<span class="fl">.2</span><span class="dv">-20</span>      </span>
<span id="cb39-65"><a href="#cb39-65" tabindex="-1"></a>[<span class="dv">109</span>] uwot_0.<span class="fl">2.3</span>             xtable_1<span class="fl">.8</span><span class="dv">-4</span>           reticulate_1.<span class="dv">41</span>.<span class="fl">0.1</span>   </span>
<span id="cb39-66"><a href="#cb39-66" tabindex="-1"></a>[<span class="dv">112</span>] munsell_0.<span class="fl">5.1</span>          processx_3.<span class="fl">8.6</span>         jquerylib_0.<span class="fl">1.4</span>       </span>
<span id="cb39-67"><a href="#cb39-67" tabindex="-1"></a>[<span class="dv">115</span>] Rcpp_1.<span class="fl">0.14</span>            globals_0.<span class="fl">16.3</span>         spatstat.random_3<span class="fl">.3</span><span class="dv">-2</span> </span>
<span id="cb39-68"><a href="#cb39-68" tabindex="-1"></a>[<span class="dv">118</span>] png_0<span class="fl">.1</span><span class="dv">-8</span>              ggrastr_1.<span class="fl">0.2</span>          spatstat.univar_3<span class="fl">.1</span><span class="dv">-2</span> </span>
<span id="cb39-69"><a href="#cb39-69" tabindex="-1"></a>[<span class="dv">121</span>] parallel_4.<span class="fl">4.0</span>         presto_1.<span class="fl">0.0</span>           dotCall64_1<span class="fl">.2</span>         </span>
<span id="cb39-70"><a href="#cb39-70" tabindex="-1"></a>[<span class="dv">124</span>] listenv_0.<span class="fl">9.1</span>          viridisLite_0.<span class="fl">4.2</span>      scales_1.<span class="fl">3.0</span>          </span>
<span id="cb39-71"><a href="#cb39-71" tabindex="-1"></a>[<span class="dv">127</span>] ggridges_0.<span class="fl">5.6</span>         rlang_1.<span class="fl">1.6</span>           </span></code></pre></div>
</div>
</div>
</div>
</div>

<div id="rmd-source-code">---
title: "Introduction to Visium Analysis"
author: "Author"
date: "2025-02" 
output:
  workflowr::wflow_html:
    toc: false
    code_folding: "hide"
editor_options: 
  chunk_output_type: console
---
 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning=FALSE, error=FALSE, collapse=T)
```

<div style = "text-align: center">

<read-time><span style="font-size: 1.3em">Please try to follow the instructions below **prior** to attending the workshop.<br><br><i>**Be aware!**</i></span><br> It is difficult to debug over zoom. <br>If you want the best feedback and experience, make sure you attend the workshop in person!</read-time>

You can download the script on the top right or from [this link](https://github.com/DrThomasOneil/digital-research-skills-network/blob/master/analysis/202504.Rmd).
 
</div><wimr>


# Introduction

# Setup{.tabset .tabset-fade}

- Create folders

- Install Packages

- Download Data

## Create Folders

```text
├── YourFolder
│     └── raw (where we can deposit raw downloaded data)
│     └── data (where we can process data and store it)
│     └── plots (where we can store plots)
```

```{r eval=T}
if(!dir.exists("raw")){dir.create("raw")}
if(!dir.exists("data")){dir.create("data")}
if(!dir.exists("plots")){dir.create("plots")}
```

<wimr>

## Install Packages

```{r, eval=F}
install.packages("Seurat")
install.packages("tidyverse")
install.packages("cowplot")

BiocManager::install("GEOquery")
```

We have a function that will let you check the setup
```{r}
source("https://github.com/DrThomasOneil/Digital-Research-Skills-Network/raw/refs/heads/main/docs/adit/checksetup.R")

checkSetup(
  cran_packages = c("Seurat", "tidyverse", "cowplot"), 
  bioc_packages=c("GEOquery") 
)

set.seed(1337)
```

<wimr>

## Download data

We will download the data directly from the [GEO](https://www.ncbi.nlm.nih.gov/geo/). We'll use [this](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE290350) dataset, which was [published in 2024](https://pmc.ncbi.nlm.nih.gov/articles/PMC11646855/). Additionally, this group have published [all of their methods and scripts for analysis](https://github.com/vildeka/Spatial_DMPA?tab=readme-ov-file).

```{r, eval=F}
# Go to the GEO page and right click on the http link under download. It should look like this

data_file <- "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE290350&format=file"

# download the data
if(!file.exists("raw/GSE290350_RAW.tar")){download.file(url=data_file, destfile = "raw/GSE290350_RAW.tar", method='curl')}

# repeat for metadata
if(!file.exists("raw/GSE290350_metadata.csv.gz")){download.file(url='https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE290350&format=file&file=GSE290350%5Fmetadata%2Ecsv%2Egz', destfile = "raw/GSE290350_metadata.csv.gz", method='curl')}

# and the supposed processed Seurat data
if(!file.exists("raw/GSE290350_seuratObj_spatial_dist.RDS")){download.file(url="https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE290350&format=file&file=GSE290350%5FseuratObj%5Fspatial%5Fdist%2ERDS", destfile = "raw/GSE290350_seuratObj_spatial_dist.RDS", method='curl')}


```

<wimr>

## Unzip and untar the raw data

```{r, eval=F}
untar("raw/GSE290350_RAW.tar", exdir = "../raw/GSE290350_RAW")
```

```{r, eval=F}
samples <- list.files("raw/GSE290350_RAW", pattern = "\\.tar\\.gz$", full.names = TRUE)

for (i in seq_along(samples)) {
  tar_path <- sub("\\.gz$", "", samples[i])

  # Unzip the .tar.gz file
  GEOquery::gunzip(samples[i], destname = tar_path, overwrite = TRUE)

  output_dir <- file.path("raw/GSE290350_RAW", sub("\\.tar$", "", basename(tar_path)))
  dir.create(output_dir, showWarnings = FALSE)

  # Extract tar to output directory
  untar(tar_path, exdir = output_dir)
}
```

<wimr>

# Pre-processing{.tabset .tabset-fade}

Here we'll load in the data. Fortunately, this data is quite neatly organised. When present, we can load the `.h5` object. Please see the `Archive` page of the [website](https://digitalresearchskills.network/0_archive.html) to find vignettes on downloading and installing GEO-sourced objects that are not as well-organised!

We will:

- Load the data

- Add the meta data

- Check the QC, filter and process the data

<div>

## Load in the data 

We'll load in the supplied metadata, and then we can load 10X data quite easily using the Load10X_Spatial function, which requires a certain format. We can load in the filtered matrix, which is filtered based on whether the initial processing detected that that spot aligned with tissue. Alternatively, you can load the raw object, and filter as you wish. 

```{r}
meta <- read.csv("raw/GSE290350_metadata.csv.gz")
```

```{r}
data <- Load10X_Spatial(
  data.dir="raw/GSE290350_RAW/GSM8811056_P004/P004",
  filename = "filtered_feature_bc_matrix.h5",
  filter.matrix = F
)
```

<wimr>

## Add meta data

The metadata provided is not found in the object itself. So we will run a small piece of code that lets you add metadata to a Seurat object. 

```{r}
data$ID <- "P004"
data <- AddMetaData(data,
                    data@meta.data %>% 
                      right_join(meta[meta$ID=="P004", ], by="ID"))
```


## QC{.tabset .tabset-fade}

There are several ways to QC transcriptomic data. First, you should check and scrutinise the [original manuscript](https://github.com/vildeka/Spatial_DMPA?tab=readme-ov-file).

We will assess: 

- number of genes per spot

- number of unique genes per spot

- mitochondrial, haemoglobin, and ribosomal percentage.

We'll also filter the genes themselves, to remove noisy/pointless genes. 

- genes that are not present in more than x number of spots

- high expressing spots.

### Genes per spot & Unique genes per spot

This is a typical `single-cell RNA` analysis metric. It should be carefully considered how this data is interpretted and filtered, as this is **not** single cell. Spots can have double the average or half the average based on the density of cells. Indeed, this is a consideration for normalisation, but this consideration is for another workshop. 

<i>**As there is <span style="color:red">no</span> gold standard**</i>, we will demonstrate how you visualise the QC metrics and then filter. 

```{r}
p1=FeatureScatter(data, "nCount_Spatial", "nFeature_Spatial")+NoLegend()
# number of genes
p2=SpatialFeaturePlot(data, "nCount_Spatial", 
                   crop=F,
                   pt.size=2,
                   alpha=c(.2,1))+
  NoLegend()
# number of unique features
p3=SpatialFeaturePlot(data, "nFeature_Spatial", 
                   crop=F,
                   pt.size=2,
                   alpha=c(.2,1))+
  NoLegend()
plot_grid(p1,plot_grid(p2,p3, nrow=2))
```

The minimum number of Features per spot (`r `min(data$nFeature_Spatial)`) and minimum number of Counts per spot (`r min(data$nCount_Spatial)`) are greater than the paper's cut off (presuming they did not use the filtered matrix). However, for the purposes of demonstrating how this is done:

```{r}
data <- subset(data, 
               subset = nCount_Spatial > 100 & nFeature_Spatial >100)
```

### Module Scoring

In this paper, they remove cells that are >15% Mt genes & > 10% Hb reads. 

```{r}
data <- PercentageFeatureSet(data, "^MT-", col.name = "percent_mito")
data <- PercentageFeatureSet(data, "^HB[^(P)]", col.name = "percent_hb")
data <- PercentageFeatureSet(data, "^RP[SL]", col.name = "percent_ribo")
```

```{r}
feature <-  c("nCount_Spatial", "nFeature_Spatial","percent_mito","percent_hb", "percent_ribo")
sapply(data@meta.data[feature], summary) %>% 
  as_tibble(rownames = "stat") %>% 
  knitr::kable(digits = 1)
```


```{r}
p1=SpatialFeaturePlot(data, "percent_mito", 
                   crop=F,
                   pt.size=2,
                   alpha=c(.2,1))+
  NoLegend()+ggtitle("Mito")
p2=SpatialFeaturePlot(data, "percent_hb", 
                   crop=F,
                   pt.size=2,
                   alpha=c(.2,1))+
  NoLegend()+ggtitle("Hb-genes")
p3=SpatialFeaturePlot(data, "percent_ribo", 
                   crop=F,
                   pt.size=2,
                   alpha=c(.2,1))+
  NoLegend()+ggtitle("Ribo")
plot_grid(p1,p2,p3)
```

Filter:

```{r}
data <- subset(data, subset = percent_mito<15 & percent_hb <10)
sapply(data@meta.data[feature], summary) %>% 
  as_tibble(rownames = "stat") %>% 
  knitr::kable(digits = 1)
```

### Filter genes

We'll first remove certain genes. MALAT1 is highly expressed in long non-coding RNA and is pretty ubiquitous. So we dont want to filter and subsequently cluster according to the expression of this gene. Similarly, Hb genes are highly expressed in rbc, which may represent contamination. We can remove these too. 

```{r}
keep_genes <- rownames(data)[!rownames(data) %in% c(grep("MALAT1", rownames(data), value=T), grep("^HB[^(P)]", rownames(data), value=T))]
data <- subset(data, features = keep_genes)
```

Low genes can create noise and take up unnecessary space in the object. Lets remove genes according to the original manuscript, whereby a gene not found in > 2 spots is removed. You can adjust this as you wish. 

```{r}
table(rowSums(data@assays$Spatial@layers$counts>=1)>=2)
keep_genes <- rowSums(data@assays$Spatial@layers$counts >= 1) >= 2
#before filtering
dim(data)
data <- subset(data, features = rownames(data)[keep_genes])
# after filtering
dim(data)
```

</div>

<hr>

We recommend taking some time to carefully consider your QC strategy. <i>**There is no universal gold standard</i>**. One common approach is to start with lenient QC, proceed with processing, and during clustering and annotation, remain aware of the relaxed filtering. If unexpected results arise, you can revisit and refine your QC. Personally, I prefer to be permissive at first, while tagging potential outliers for future consideration. For example, we initially filtered out genes not present in at least 2 spots, but I may also label genes present in fewer than 5 spots as “toQC” for later review—particularly when assessing differential expression results. Similarly, we applied a 15% mitochondrial threshold but did not filter based on ribosomal content. At a later stage, we could tighten the mitochondrial threshold to 10% and include a filter for high ribosomal spots. If unusual clusters appear, we can then assess how many of those cells would have been excluded under stricter QC.


# Data Processing{.tabset .tabset-fade}

Seurat is an efficient package with simple functions for standard processing. 

- `NormalizeData`: The default is to log-normalize *(Feature counts for each cell are divided by the total counts for that cell and multiplied by the scale.factor. This is then natural-log transformed using log1p)*. Spatial data may be better normalised with Area/cell count considered. As we don't have cell counts per spot here, we'll stick to the standard approach and that taken by the authors 

- `FindVariableFeatures`: The standard approach is 2000 top variable genes + 'vst' selection method. *(First, fits a line to the relationship of log(variance) and log(mean) using local polynomial regression (loess). Then standardizes the feature values using the observed mean and expected variance (given by the fitted line). Feature variance is then calculated on the standardized values after clipping to a maximum)*. Standard is 2000 genes

- `ScaleData`: here, we scale just the variable genes to save space. If you find yourself unable to visualise your gene of interest in subsequent visualisations, it was not in this top variable list. In that case, you can scale all data. 

- `RunPCA` & `RunUMAP`: Visualisation tools. There are a few things to consider in the UMAP function, but most important is the `dims` argument. We'll cover this below. 

<div>

## Normalisation

Again, normalisation is quite simple. 

```{r}
data <- NormalizeData(data, 
                      normalization.method = "LogNormalize",
                      verbose=F)
```

## Variable Features

We'll find the top variable features.

```{r}
data <- FindVariableFeatures(data, 
                             method = 'vst', 
                             nfeatures = 2000, 
                             verbose=F)
```

We can output/visualise them in several ways. 

```{r}
VariableFeatures(data) %>% head(50)

LabelPoints(plot = VariableFeaturePlot(data), 
            points = head(VariableFeatures(data), 50), repel = TRUE)
```

## Scale Data

Scaling the data is also easy!

```{r}
data <- ScaleData(data, 
                  features = VariableFeatures(data), #if you want to scale all features, change this to rownames(data)
        verbose=F)
```

## PCA & UMAP

Also very easy with Seurat. We'll determine the number of PCs to use for the UMAP using ElbowPlot

```{r}
data <- RunPCA(data,
               npcs = 50,
               verbose=F)

ElbowPlot(data)
```

The elbow plot drops off drastically and plateaus at ~6, meaning the variation just becomes noisy around PC6-onwards. So we'll just use the first 6. 

```{r}
data <- RunUMAP(data, 
                dims=1:6, 
                verbose=F)
```

</div><hr>

# Visualisation{.tabset .tabset-fade}

There are several visualisations we can do:

**Graph**:

- UMAP visualisation

- Cluster based expressions

**Spatial**:

- Genes overlaid on the tissue

- Module scores overlaid on the tissue

## UMAP{.tabset .tabset-fade}

### UMAP Visualisation

We'll first cluster the data.

```{r}

data <- data %>% 
  FindNeighbors(dims = 1:6, verbose=F) %>%
  FindClusters(resolution=1, verbose=F)

```

And now we can create a UMAP. 

```{r}
p1=UMAPPlot(data,
            label=T, label.box=T)+
  NoLegend()+
  NoAxes()+
  ggtitle("11 Clusters")
```

<details><summary>**Open to view more Plotting options**</summary>

Here are some additional arguments that change the default visualisations. Chop and change to see how they work:

```{r}
UMAPPlot(data, 
         pt.size=2, 
         label=T, 
         label.size=10,
         label.box=T,
         label.color="white",
         repel=T)+
  NoLegend()+
  NoAxes()+
  scale_color_manual(values= c("blue4", 'blue2', 'lightblue', 'green2', 'green4','brown', 'orange', 'red', 'red3', 'red4', 'black'))+
  scale_fill_manual(values= c("blue4", 'blue2', 'lightblue', 'green2', 'green4','brown', 'orange', 'red', 'red3', 'red4', 'black'))
```

</details>

### Featureplots

We can view the gene expression on the UMAP itself. 

```{r}
p2=FeaturePlot(data, 
            "CD3E",
            order=T,
            pt.size=2, 
            cols = c("yellow2", 'blue3'))+NoAxes()+NoLegend()
p3=FeaturePlot(data, 
            "F13A1",
            order=T,
            pt.size=2, 
            cols = c("yellow2", 'blue3'))+NoAxes()+NoLegend()
p4=FeaturePlot(data, 
            "COL1A1",
            order=T,
            pt.size=2, 
            cols = c("yellow2", 'blue3'))+NoAxes()+NoLegend()
p5=FeaturePlot(data, 
            "KRT5",
            order=T,
            pt.size=2,
            min.cutoff = 3,
            cols = c("yellow2", 'blue3'))+NoAxes()+NoLegend()
plot_grid(p1, plot_grid(p2,p3,p4,p5, ncol=2))
```

### DE and DotPlots

And we can simply assess the clusters themselves using differential analysis and expression graphs. 

```{r}


markers <- FindAllMarkers(data, 
                          logfc.threshold = 0.1,#default - increasing speeds it up, but may miss weaker signals.
                          min.pct = 0.01, #default - can be increased to ensure that the minimum expression of genes are considered. 
                          min.diff.pct = -Inf, # you can adjust this too. IF you want there to be at least a 10% difference in percentage of spots expressing the gene, you'd change this to 0.1. Can ensure that its not only the level of expression, but that you're capturing highly expressed genes
                          verbose=F #change this to true if you want to track the speed at which DE is being calculated. It is off here to save it outputting into the document 
                          ) 
top10 <- markers %>% 
  group_by(cluster) %>% 
  top_n(wt=avg_log2FC, n=10)
```

<details><summary>Heatmap</summary>

```{r}
DoHeatmap(AggregateExpression(ScaleData(data, rownames(data), verbose=F), return.seurat = T, verbose=F),
          features = top10$gene, 
          draw.lines = F)
```

<hr>

</details>

<details><summary>DotPlot</summary>

```{r}
DotPlot(data, 
        features = c("CD3E", "CLEC10A", "CD207", "KRT5", "COL1A1"), 
        cols=c("grey", "red"))
```

</details>

<details><summary>Violin</summary>

```{r}
VlnPlot(data, 
        features = c("CD3E", "CLEC10A", "CD207", "KRT5", "COL1A1"))
VlnPlot(data, 
        features = c("CD3E", "CLEC10A", "CD207", "KRT5", "COL1A1"), stack=T)
```

</details>

## Spatial{.tabset .tabset-fade}

We can do a few things here. 

### Clusters overlaid spatially

```{r}
SpatialDimPlot(data, pt.size=3)
ImageDimPlot(data, size=3)
```

### Expressions overlaid spatially

```{r}
SpatialFeaturePlot(data, 
                   features="CD3E",
                   pt.size=3)

SpatialFeaturePlot(data, 
                   features="CD3E",
                   pt.size=3)+
  scale_fill_viridis_b()

ImageFeaturePlot(data, 
                   features="CD3E",
                   size=3)
```











</div>
<div class="custom-footer">
  <div class="footer-left" style="width: 30%; display: flex; justify-content: center; align-items: center;">
    <a href="https://wimr.org.au/" target="_blank">
      <img src="assets/footer/footer1.png" alt="wimr-logo" style="max-height: 100px; width: auto;">
    </a>
   </div>
  <div class="footer-center">
    <p><strong>Digital Research Skills Network</strong><br>
      <a href="mailto:thomas.oneil@sydney.edu.au"><strong>Contact us</strong></a>
    </p>
  </div>
  <div class="footer-right" style="width: 30%; display: flex; justify-content: center; align-items: center;">
    <a href="https://www.precisiongo.com.au/" target="_blank">
      <img src="assets/footer/PGlogo.png" alt="wimr-logo" style="max-height: 100px; width: auto;">
    </a>  
  </div>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("202505_Workshop.Rmd");
  window.initializeCodeFolding("hide" === "show");
});
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
